
<html>
<head>
<meta http-equiv="CONTENT-TYPE" content="text/html; charset=utf-8" />
<title>e! Variation schema</title>

<script language="Javascript" type="text/javascript">
	// Function to show/hide the columns table
	function show_hide (param) {
		div   = document.getElementById('div_'+param);
		alink = document.getElementById('a_'+param);
		if (div.style.display=='inline') {
			div.style.display='none';
			alink.innerHTML='Show';
		}
		else {
			if (div.style.display=='none') {
				div.style.display='inline';
				alink.innerHTML='Hide';
			}
		}
	}
</script>

</head>

<body>
<h1>Ensembl Variation Schema Documentation</h1>

<h2>Introduction</h2>
<p>
This document gives a high-level description of the tables that
make up the Ensembl variation schema. Tables are listed by alphabetical order, and the purpose of each table is explained. It is intended to
allow people to familiarise themselves with the schema when
encountering it for the first time, or when they need to use some
tables that they've not used before.
</p>

<p>
This document refers to version <strong>62</strong> of the Ensembl
variation schema. 
</p>

<p>
A PDF document of the schema is available <a href="variation-database-schema.pdf">here</a>.
</p>
<br />
<hr />

<h3>List of the tables:</h3>
<table><tr><td>
 <ul>
		<li><a href="#allele"><b>allele</b></a></li>
		<li><a href="#allele_group"><b>allele_group</b></a></li>
		<li><a href="#allele_group_allele"><b>allele_group_allele</b></a></li>
		<li><a href="#associate_study"><b>associate_study</b></a></li>
		<li><a href="#attrib"><b>attrib</b></a></li>
		<li><a href="#attrib_set"><b>attrib_set</b></a></li>
		<li><a href="#attrib_type"><b>attrib_type</b></a></li>
		<li><a href="#compressed_genotype_single_bp"><b>compressed_genotype_single_bp</b></a></li>
		<li><a href="#failed_allele"><b>failed_allele</b></a></li>
		<li><a href="#failed_description"><b>failed_description</b></a></li>
		<li><a href="#failed_variation"><b>failed_variation</b></a></li>
		<li><a href="#flanking_sequence"><b>flanking_sequence</b></a></li>
		<li><a href="#httag"><b>httag</b></a></li>
		<li><a href="#individual"><b>individual</b></a></li>
		<li><a href="#individual_genotype_multiple_bp"><b>individual_genotype_multiple_bp</b></a></li>
		<li><a href="#individual_population"><b>individual_population</b></a></li>
		<li><a href="#individual_type"><b>individual_type</b></a></li>
		<li><a href="#meta"><b>meta</b></a></li>
		<li><a href="#meta_coord"><b>meta_coord</b></a></li>
		<li><a href="#phenotype"><b>phenotype</b></a></li>
	</ul></td><td><ul>
		<li><a href="#polyphen_prediction"><b>polyphen_prediction</b></a></li>
		<li><a href="#population"><b>population</b></a></li>
		<li><a href="#population_genotype"><b>population_genotype</b></a></li>
		<li><a href="#population_structure"><b>population_structure</b></a></li>
		<li><a href="#protein_info"><b>protein_info</b></a></li>
		<li><a href="#protein_position"><b>protein_position</b></a></li>
		<li><a href="#read_coverage"><b>read_coverage</b></a></li>
		<li><a href="#sample"><b>sample</b></a></li>
		<li><a href="#sample_synonym"><b>sample_synonym</b></a></li>
		<li><a href="#seq_region"><b>seq_region</b></a></li>
		<li><a href="#sift_prediction"><b>sift_prediction</b></a></li>
		<li><a href="#source"><b>source</b></a></li>
		<li><a href="#structural_variation"><b>structural_variation</b></a></li>
		<li><a href="#study"><b>study</b></a></li>
		<li><a href="#subsnp_handle"><b>subsnp_handle</b></a></li>
		<li><a href="#supporting_structural_variation"><b>supporting_structural_variation</b></a></li>
		<li><a href="#tagged_variation_feature"><b>tagged_variation_feature</b></a></li>
		<li><a href="#tmp_individual_genotype_single_bp"><b>tmp_individual_genotype_single_bp</b></a></li>
		<li><a href="#transcript_variation"><b>transcript_variation</b></a></li>
		<li><a href="#variation"><b>variation</b></a></li>
	</ul></td><td><ul>
		<li><a href="#variation_annotation"><b>variation_annotation</b></a></li>
		<li><a href="#variation_feature"><b>variation_feature</b></a></li>
		<li><a href="#variation_group"><b>variation_group</b></a></li>
		<li><a href="#variation_group_feature"><b>variation_group_feature</b></a></li>
		<li><a href="#variation_group_variation"><b>variation_group_variation</b></a></li>
		<li><a href="#variation_set"><b>variation_set</b></a></li>
		<li><a href="#variation_set_structure"><b>variation_set_structure</b></a></li>
		<li><a href="#variation_set_variation"><b>variation_set_variation</b></a></li>
		<li><a href="#variation_synonym"><b>variation_synonym</b></a></li>
		</ul>
</td></tr></table>
<hr />

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="allele" style="font-size:11pt;font-weight:bold">allele</span></td>
<td style="width:100px;text-align:right"><a id="a_allele" style="cursor:pointer;text-decoration:underline" onclick="show_hide('allele')">Show</a> columns</td>
</tr></table>
<p>This table stores information about each of a variation's alleles, along with population frequencies.<p>

	<div id="div_allele" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>allele_id</b></td><td>int(10)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
		<tr class="bg2"><td><b>variation_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#variation">variation</a> table.</td><td>key: variation_idx</td></tr>
		<tr class="bg1"><td><b>subsnp_id</b></td><td>int(15)</td><td></td><td>Foreign key references to the <a href="#subsnp_handle">subsnp_handle</a> table.</td><td>key: subsnp_idx</td></tr>
		<tr class="bg2"><td><b>allele</b></td><td>varchar(25000)</td><td></td><td>Allele found at the variation location, for the sample. e.g. "A".</td><td>key: variation_idx</td></tr>
		<tr class="bg1"><td><b>frequency</b></td><td>float</td><td></td><td>Frequency of this allele in the sample.</td><td></td></tr>
		<tr class="bg2"><td><b>sample_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#sample">sample</a> table.</td><td></td></tr>
		<tr class="bg1"><td><b>count</b></td><td>int(10)</td><td>NULL</td><td>Number of individuals in the sample where this allele is found.</td><td></td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#variation">variation</a></li>
	<li><a href="#population">population</a></li>
	<li><a href="#subsnp_handle">subsnp_handle</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="allele_group" style="font-size:11pt;font-weight:bold">allele_group</span></td>
<td style="width:100px;text-align:right"><a id="a_allele_group" style="cursor:pointer;text-decoration:underline" onclick="show_hide('allele_group')">Show</a> columns</td>
</tr></table>
<p>This table, along with allele_group_allele, represents a particular multi-marker allele of a given multi-marker variation, or haplotype. It stores an associated population frequency.<p>

	<div id="div_allele_group" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>allele_group_id</b></td><td>int(10)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
		<tr class="bg2"><td><b>variation_group_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#variation_group">variation_group</a> table.</td><td></td></tr>
		<tr class="bg1"><td><b>sample_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#population">population</a> table.</td><td></td></tr>
		<tr class="bg2"><td><b>name</b></td><td>varchar(255)</td><td></td><td>The name of this allele group.</td><td>unique</td></tr>
		<tr class="bg1"><td><b>source_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#source">source</a> table.</td><td></td></tr>
		<tr class="bg2"><td><b>frequency</b></td><td>float</td><td></td><td>The frequency of this allele_group within the referenced population.</td><td></td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#allele_group_allele">allele_group_allele</a></li>
	<li><a href="#variation_group">variation_group</a></li>
	<li><a href="#population">population</a></li>
	<li><a href="#source">source</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="allele_group_allele" style="font-size:11pt;font-weight:bold">allele_group_allele</span></td>
<td style="width:100px;text-align:right"><a id="a_allele_group_allele" style="cursor:pointer;text-decoration:underline" onclick="show_hide('allele_group_allele')">Show</a> columns</td>
</tr></table>
<p>This table represents an allele of one variation in a multi-marker variation, or haplotype. It stores a string of the allele.<p>

	<div id="div_allele_group_allele" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>allele_group_id</b></td><td>int(10)</td><td></td><td>Primary key, internal identifier.</td><td>unique<br />key: allele_idx</td></tr>
		<tr class="bg2"><td><b>allele</b></td><td>varchar(255)</td><td></td><td>Nucleotid presents in the group.</td><td></td></tr>
		<tr class="bg1"><td><b>variation_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#variation">variation</a> table.</td><td>unique<br />key: allele_idx</td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#allele_group">allele_group</a></li>
	<li><a href="#variation">variation</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="associate_study" style="font-size:11pt;font-weight:bold">associate_study</span></td>
<td style="width:100px;text-align:right"><a id="a_associate_study" style="cursor:pointer;text-decoration:underline" onclick="show_hide('associate_study')">Show</a> columns</td>
</tr></table>
<p>This table contains identifiers of associated studies (e.g. NHGRI and EGA studies with the same pubmed identifier).<p>

	<div id="div_associate_study" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>study1_id</b></td><td>int(10)</td><td></td><td>Primary key. Foreign key references to the <a href="#study">study</a> table.</td><td>primary key</td></tr>
		<tr class="bg2"><td><b>study2_id</b></td><td>int(10)</td><td></td><td>Primary key. Foreign key references to the <a href="#study">study</a> table.</td><td>primary key</td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#study">study</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="attrib" style="font-size:11pt;font-weight:bold">attrib</span></td>
<td style="width:100px;text-align:right"><a id="a_attrib" style="cursor:pointer;text-decoration:underline" onclick="show_hide('attrib')">Show</a> columns</td>
</tr></table>
<p>Defines various attributes used elsewhere in the database<p>

	<div id="div_attrib" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>attrib_id</b></td><td>INT(11)</td><td>0</td><td>Primary key</td><td>primary key</td></tr>
		<tr class="bg2"><td><b>attrib_type_id</b></td><td>SMALLINT(5)</td><td>0</td><td>Key into the <a href="#attrib_type">attrib_type</a> table, identifies the type of this attribute</td><td>key: type_val_idx</td></tr>
		<tr class="bg1"><td><b>value</b></td><td>TEXT</td><td></td><td>The value of this attribute</td><td>key: type_val_idx</td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#attrib_type">attrib_type</a></li>
	<li><a href="#attrib_set">attrib_set</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="attrib_set" style="font-size:11pt;font-weight:bold">attrib_set</span></td>
<td style="width:100px;text-align:right"><a id="a_attrib_set" style="cursor:pointer;text-decoration:underline" onclick="show_hide('attrib_set')">Show</a> columns</td>
</tr></table>
<p>Groups related attributes together<p>

	<div id="div_attrib_set" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>attrib_set_id</b></td><td>INT(11)</td><td>0</td><td>Primary key</td><td>unique key: set_idx</td></tr>
		<tr class="bg2"><td><b>attrib_id</b></td><td>INT(11)</td><td>0</td><td>Key of an attribute in this set</td><td>unique key: set_idx<br />key: attrib_idx</td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#attrib">attrib</a></li>
	<li><a href="#attrib_type">attrib_type</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="attrib_type" style="font-size:11pt;font-weight:bold">attrib_type</span></td>
<td style="width:100px;text-align:right"><a id="a_attrib_type" style="cursor:pointer;text-decoration:underline" onclick="show_hide('attrib_type')">Show</a> columns</td>
</tr></table>
<p>Defines the set of possible attribute types used in the attrib table<p>

	<div id="div_attrib_type" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>attrib_type_id</b></td><td>SMALLINT(5)</td><td>0</td><td>Primary key</td><td>primary key</td></tr>
		<tr class="bg2"><td><b>code</b></td><td>VARCHAR(20)</td><td>''</td><td>A short codename for this type (indexed, so should be used for lookups)</td><td>unique key: code_idx</td></tr>
		<tr class="bg1"><td><b>name</b></td><td>VARCHAR(255)</td><td>''</td><td>The name of this type</td><td></td></tr>
		<tr class="bg2"><td><b>description</b></td><td>TEXT</td><td></td><td>Longer description of this type</td><td></td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#attrib">attrib</a></li>
	<li><a href="#attrib_set">attrib_set</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="compressed_genotype_single_bp" style="font-size:11pt;font-weight:bold">compressed_genotype_single_bp</span></td>
<td style="width:100px;text-align:right"><a id="a_compressed_genotype_single_bp" style="cursor:pointer;text-decoration:underline" onclick="show_hide('compressed_genotype_single_bp')">Show</a> columns</td>
</tr></table>
<p>This table holds genotypes compressed using the pack() method in Perl. These genotypes are mapped to particular genomic locations rather than variation objects. The data have been compressed to reduce table size and increase the speed of the web code.<p>

	<div id="div_compressed_genotype_single_bp" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>sample_id</b></td><td>int(10)</td><td></td><td>Primary key. Foreign key references to the sample table.</td><td></td></tr>
		<tr class="bg2"><td><b>seq_region_id</b></td><td>int(10)</td><td></td><td>Foreign key references <a href="#seq_region">seq_region</a> in core db. ers to the seq_region which this variant is on, which may be a chromosome, a clone, etc...</td><td>key: pos_idx</td></tr>
		<tr class="bg1"><td><b>seq_region_start</b></td><td>int</td><td></td><td>The start position of the variation on the <a href="#seq_region">seq_region</a>.</td><td>key: pos_idx</td></tr>
		<tr class="bg2"><td><b>seq_region_end</b></td><td>int</td><td></td><td>The end position of the variation on the <a href="#seq_region">seq_region</a>.</td><td></td></tr>
		<tr class="bg1"><td><b>seq_region_strand</b></td><td>tinyint</td><td></td><td>The orientation of the variation on the <a href="#seq_region">seq_region</a>.</td><td></td></tr>
		<tr class="bg2"><td><b>genotypes</b></td><td>blob</td><td></td><td>Encoded representation of the genotype data:<br />Each row in the compressed table stores genotypes from one individual in one fixed-size region of the genome (arbitrarily defined as 100 Kb). The compressed string (using Perl's pack method) consisting of a repeating triplet of elements: a distance in base pairs from the previous genotype followed by a pair of alleles.<br />For example, a given row may have a start position of 1000, indicating the chromosomal position of the first genotype in this row. The unpacked genotypes field then may contain the following elements:<br />0, A, G, 20, C, C, 35, G, T, 320, A, A, ...<br />The first genotype has a position of 1000 + 0 = 1000 and alleles A and G.<br />The second genotype has a position of 1000 + 20 = 1020 and alleles C and C.<br />The third genotype similarly has a position of 1055 and alleles G and T, and so on.</td><td></td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#individual">individual</a></li>
	<li><a href="#seq_region">seq_region</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="failed_allele" style="font-size:11pt;font-weight:bold">failed_allele</span></td>
<td style="width:100px;text-align:right"><a id="a_failed_allele" style="cursor:pointer;text-decoration:underline" onclick="show_hide('failed_allele')">Show</a> columns</td>
</tr></table>
<p>Contains alleles that did not pass the Ensembl filters<p>

	<div id="div_failed_allele" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>failed_allele_id</b></td><td>int(11)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
		<tr class="bg2"><td><b>allele_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#allele">allele</a> table.</td><td>unique key: allele_idx</td></tr>
		<tr class="bg1"><td><b>failed_description_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#failed_description">failed_description</a> table.</td><td>unique key: allele_idx</td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#failed_description">failed_description</a></li>
	<li><a href="#allele">allele</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="failed_description" style="font-size:11pt;font-weight:bold">failed_description</span></td>
<td style="width:100px;text-align:right"><a id="a_failed_description" style="cursor:pointer;text-decoration:underline" onclick="show_hide('failed_description')">Show</a> columns</td>
</tr></table>
<p>This table contains descriptions of reasons for a variation being flagged as failed.<p>

	<div id="div_failed_description" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>failed_description_id</b></td><td>int(10)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
		<tr class="bg2"><td><b>description</b></td><td>text</td><td></td><td>Text containing the reason why the Variation has been flagged as failed. e.g. "Variation does not map to the genome".</td><td></td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#failed_variation">failed_variation</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="failed_variation" style="font-size:11pt;font-weight:bold">failed_variation</span></td>
<td style="width:100px;text-align:right"><a id="a_failed_variation" style="cursor:pointer;text-decoration:underline" onclick="show_hide('failed_variation')">Show</a> columns</td>
</tr></table>
<p>For various reasons it may be necessary to store information about a variation that has failed quality checks in the Variation pipeline. This table acts as a flag for such failures.<p>

	<div id="div_failed_variation" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>failed_variation_id</b></td><td>int(11)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
		<tr class="bg2"><td><b>variation_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#variation">variation</a> table.</td><td>unique key: variation_idx</td></tr>
		<tr class="bg1"><td><b>failed_description_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#failed_description">failed_description</a> table.</td><td>unique key: variation_idx</td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#failed_description">failed_description</a></li>
	<li><a href="#variation">variation</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="flanking_sequence" style="font-size:11pt;font-weight:bold">flanking_sequence</span></td>
<td style="width:100px;text-align:right"><a id="a_flanking_sequence" style="cursor:pointer;text-decoration:underline" onclick="show_hide('flanking_sequence')">Show</a> columns</td>
</tr></table>
<p>This table contains the upstream and downstream sequence surrounding a variation. Since each variation is defined by its flanking sequence, this table has a one-to-one relationship with the variation table.<p>

	<div id="div_flanking_sequence" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>variation_id</b></td><td>int(10)</td><td></td><td>Primary key. Foreign key references to the variation table.</td><td>primary key</td></tr>
		<tr class="bg2"><td><b>up_seq</b></td><td>text</td><td></td><td>Upstream sequence, used to initially store the sequence from the core database, and in a later process get from here the position.</td><td></td></tr>
		<tr class="bg1"><td><b>down_seq</b></td><td>text</td><td></td><td>Downstream sequence, used to initially store the sequence from the core database, and in a later process get from here the position.</td><td></td></tr>
		<tr class="bg2"><td><b>up_seq_region_start</b></td><td>int</td><td></td><td>Position of the starting of the sequence in the region.</td><td></td></tr>
		<tr class="bg1"><td><b>up_seq_region_end</b></td><td>int</td><td></td><td>Position of the end of the sequence in the region.</td><td></td></tr>
		<tr class="bg2"><td><b>down_seq_region_start</b></td><td>int</td><td></td><td>Position of the starting of the sequence in the region.</td><td></td></tr>
		<tr class="bg1"><td><b>down_seq_region_end</b></td><td>int</td><td></td><td>Position of the end of the sequence in the region.</td><td></td></tr>
		<tr class="bg2"><td><b>seq_region_id</b></td><td>int(10)</td><td></td><td>Foreign key references <a href="#seq_region">seq_region</a> in core db. Refers to the seq_region which this variant is on, which may be a chromosome or clone etc..</td><td></td></tr>
		<tr class="bg1"><td><b>seq_region_strand</b></td><td>tinyint</td><td></td><td>The orientation of the variation on the seq_region.</td><td></td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#variation">variation</a></li>
	<li><a href="#seq_region">seq_region</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="httag" style="font-size:11pt;font-weight:bold">httag</span></td>
<td style="width:100px;text-align:right"><a id="a_httag" style="cursor:pointer;text-decoration:underline" onclick="show_hide('httag')">Show</a> columns</td>
</tr></table>
<p>This table represents the equivalent of a tagged_variation_feature for multi-marker variations, representing an instance where a haplotype is tagged by or tags another marker.<p>

	<div id="div_httag" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>httag_id</b></td><td>int(10)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
		<tr class="bg2"><td><b>variation_group_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#variation_group">variation_group</a> table.</td><td>key: variation_group_idx</td></tr>
		<tr class="bg1"><td><b>name</b></td><td>varchar(255)</td><td></td><td>The name of the tag, for web purposes.</td><td></td></tr>
		<tr class="bg2"><td><b>source_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#source">source</a> table.</td><td></td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#variation_group">variation_group</a></li>
	<li><a href="#source">source</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="individual" style="font-size:11pt;font-weight:bold">individual</span></td>
<td style="width:100px;text-align:right"><a id="a_individual" style="cursor:pointer;text-decoration:underline" onclick="show_hide('individual')">Show</a> columns</td>
</tr></table>
<p>Stores information about an identifiable individual, including gender and the identifiers of the individual's parents (if known).<p>

	<div id="div_individual" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>sample_id</b></td><td>int(10)</td><td></td><td>Primary key, internal identifier. See the <a href="#sample">sample</a> table. Corresponds to the individual ID.</td><td>primary key</td></tr>
		<tr class="bg2"><td><b>gender</b></td><td>enum('Male', 'Female', 'Unknown')</td><td>'Unknown'</td><td>The sex of this individual.</td><td></td></tr>
		<tr class="bg1"><td><b>father_individual_sample_id</b></td><td>int(10)</td><td></td><td>Self referential ID, the father of this individual if known.</td><td></td></tr>
		<tr class="bg2"><td><b>mother_individual_sample_id</b></td><td>int(10)</td><td></td><td>Self referential ID, the mother of this individual if known.</td><td></td></tr>
		<tr class="bg1"><td><b>individual_type_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#individual_type">individual_type</a> table.</td><td></td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#sample">sample</a></li>
	<li><a href="#individual_type">individual_type</a></li>
	<li><a href="#individual_population">individual_population</a></li>
	<li><a href="#individual_genotype_multiple_bp">individual_genotype_multiple_bp</a></li>
	<li><a href="#compressed_genotype_single_bp">compressed_genotype_single_bp</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="individual_genotype_multiple_bp" style="font-size:11pt;font-weight:bold">individual_genotype_multiple_bp</span></td>
<td style="width:100px;text-align:right"><a id="a_individual_genotype_multiple_bp" style="cursor:pointer;text-decoration:underline" onclick="show_hide('individual_genotype_multiple_bp')">Show</a> columns</td>
</tr></table>
<p>This table holds uncompressed genotypes for given variations.<p>

	<div id="div_individual_genotype_multiple_bp" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>variation_id</b></td><td>int(10)</td><td></td><td>Primary key. Foreign key references to the <a href="#variation">variation</a> table.</td><td>key: variation_idx</td></tr>
		<tr class="bg2"><td><b>subsnp_id</b></td><td>int(15)</td><td></td><td>Foreign key references to the <a href="#subsnp_handle">subsnp_handle</a> table.</td><td>key: subsnp_idx</td></tr>
		<tr class="bg1"><td><b>allele_1</b></td><td>varchar(25000)</td><td></td><td>One of the alleles of the genotype, e.g. "TAG".</td><td></td></tr>
		<tr class="bg2"><td><b>allele_2</b></td><td>varchar(25000)</td><td></td><td>The other allele of the genotype.</td><td></td></tr>
		<tr class="bg1"><td><b>sample_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#individual">individual</a> table.</td><td>key: sample_idx</td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#individual">individual</a></li>
	<li><a href="#variation">variation</a></li>
	<li><a href="#subsnp_handle">subsnp_handle</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="individual_population" style="font-size:11pt;font-weight:bold">individual_population</span></td>
<td style="width:100px;text-align:right"><a id="a_individual_population" style="cursor:pointer;text-decoration:underline" onclick="show_hide('individual_population')">Show</a> columns</td>
</tr></table>
<p>This table resolves the many-to-many relationship between the individual and population tables; i.e. samples may belong to more than one population. Hence it is composed of rows of individual and population identifiers.<p>

	<div id="div_individual_population" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>individual_sample_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#individual">individual</a> table.</td><td>key: individual_sample_idx</td></tr>
		<tr class="bg2"><td><b>population_sample_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#population">population</a> table.</td><td>key: population_sample_idx</td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#individual">individual</a></li>
	<li><a href="#population">population</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="individual_type" style="font-size:11pt;font-weight:bold">individual_type</span></td>
<td style="width:100px;text-align:right"><a id="a_individual_type" style="cursor:pointer;text-decoration:underline" onclick="show_hide('individual_type')">Show</a> columns</td>
</tr></table>
<p>This table resolves the many-to-many relationship between the individual and population tables; i.e. samples may belong to more than one population. Hence it is composed of rows of individual and population identifiers.<p>

	<div id="div_individual_type" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>individual_type_id</b></td><td>int(0)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
		<tr class="bg2"><td><b>name</b></td><td>varchar(255)</td><td></td><td>Short name of the individual type. e.g. "fully_inbred","mutant".</td><td></td></tr>
		<tr class="bg1"><td><b>description</b></td><td>text</td><td></td><td>Long name of the individual type.</td><td></td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#individual">individual</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="meta" style="font-size:11pt;font-weight:bold">meta</span></td>
<td style="width:100px;text-align:right"><a id="a_meta" style="cursor:pointer;text-decoration:underline" onclick="show_hide('meta')">Show</a> columns</td>
</tr></table>
<p>This table stores various metadata relating to the database, generally used by the Ensembl web code.<p>

	<div id="div_meta" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>meta_id</b></td><td>INT(10)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
		<tr class="bg2"><td><b>species_id</b></td><td>INT</td><td>1</td><td>...</td><td>unique key: species_key_value_idx<br />key: species_value_idx</td></tr>
		<tr class="bg1"><td><b>meta_key</b></td><td>varchar( 40 )</td><td></td><td>Name of the meta entry, e.g. "schema_version".</td><td>unique key: species_key_value_idx</td></tr>
		<tr class="bg2"><td><b>meta_value</b></td><td>varchar( 255 )</td><td></td><td>Corresponding value of the key, e.g. "61".</td><td>unique key: species_key_value_idx<br />key: species_value_idx</td></tr>
</table>
</div>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="meta_coord" style="font-size:11pt;font-weight:bold">meta_coord</span></td>
<td style="width:100px;text-align:right"><a id="a_meta_coord" style="cursor:pointer;text-decoration:underline" onclick="show_hide('meta_coord')">Show</a> columns</td>
</tr></table>
<p>This table gives the coordinate system used by various tables in the database.<p>

	<div id="div_meta_coord" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>table_name</b></td><td>VARCHAR(40)</td><td></td><td>Name of the feature table, e.g. "variation_feature".</td><td>unique</td></tr>
		<tr class="bg2"><td><b>coord_system_id</b></td><td>INT(10)</td><td></td><td>Foreign key to core database coord_system table refers to coordinate system that features from this table can be found in.</td><td>unique</td></tr>
		<tr class="bg1"><td><b>max_length</b></td><td>INT</td><td></td><td>Maximun length of the feature. </td><td></td></tr>
</table>
</div>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="phenotype" style="font-size:11pt;font-weight:bold">phenotype</span></td>
<td style="width:100px;text-align:right"><a id="a_phenotype" style="cursor:pointer;text-decoration:underline" onclick="show_hide('phenotype')">Show</a> columns</td>
</tr></table>
<p>This table stores details of the phenotypes associated with variation annotations.<p>

	<div id="div_phenotype" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>phenotype_id</b></td><td>int(10)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
		<tr class="bg2"><td><b>name</b></td><td>varchar(50)</td><td></td><td>Phenotype short name. e.g. "CAD".</td><td>unique key: name_idx</td></tr>
		<tr class="bg1"><td><b>description</b></td><td>varchar(255)</td><td></td><td>varchar		Phenotype long name. e.g. "Coronary Artery Disease".</td><td></td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#variation_annotation">variation_annotation</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="polyphen_prediction" style="font-size:11pt;font-weight:bold">polyphen_prediction</span></td>
<td style="width:100px;text-align:right"><a id="a_polyphen_prediction" style="cursor:pointer;text-decoration:underline" onclick="show_hide('polyphen_prediction')">Show</a> columns</td>
</tr></table>
<p>Stores the PolyPhen 2 prediction for every possible amino acid substitution in the ensembl proteome<p>

	<div id="div_polyphen_prediction" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>polyphen_prediction_id</b></td><td>int(10)</td><td></td><td>Primary key</td><td>primary key</td></tr>
		<tr class="bg2"><td><b>protein_position_id</b></td><td>int(10)</td><td></td><td>Foreign key into the <a href="#protein_position">protein_position</a> table identifying the protein and position that this prediction applies to</td><td>key: pos_aa_idx</td></tr>
		<tr class="bg1"><td><b>amino_acid</b></td><td>char(1)</td><td></td><td>The substituted amino acid</td><td>key: pos_aa_idx</td></tr>
		<tr class="bg2"><td><b>prediction</b></td><td>enum('unknown', 'benign', 'possibly damaging', 'probably damaging')</td><td></td><td>The qualitative PolyPhen prediction for this substitution</td><td></td></tr>
		<tr class="bg1"><td><b>probability</b></td><td>float</td><td></td><td>The PolyPhen probability that this substitution is damaging</td><td></td></tr>
		<tr class="bg2"><td><b>compressed_result_hash</b></td><td>blob</td><td></td><td>A compressed string representation of a Perl hash with further results from PolyPhen (not released)</td><td></td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#protein_info">protein_info</a></li>
	<li><a href="#protein_position">protein_position</a></li>
	<li><a href="#sift_prediction">sift_prediction</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="population" style="font-size:11pt;font-weight:bold">population</span></td>
<td style="width:100px;text-align:right"><a id="a_population" style="cursor:pointer;text-decoration:underline" onclick="show_hide('population')">Show</a> columns</td>
</tr></table>
<p>A table consisting simply of sample_ids representing populations; all data relating to the populations are stored in separate tables (see below).<br />A population may be an ethnic group (e.g. caucasian, hispanic), assay group (e.g. 24 europeans), strain, phenotypic group (e.g. blue eyed, diabetes) etc. Populations may be composed of other populations by defining relationships in the population_structure table.<p>

	<div id="div_population" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>sample_id</b></td><td>int(10)</td><td></td><td>int	Foreign key references to the <a href="#sample">sample</a> table. Corresponds to the population ID.</td><td>primary key</td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#sample">sample</a></li>
	<li><a href="#sample_synonym">sample_synonym</a></li>
	<li><a href="#individual_population">individual_population</a></li>
	<li><a href="#population_structure">population_structure</a></li>
	<li><a href="#population_genotype">population_genotype</a></li>
	<li><a href="#allele">allele</a></li>
	<li><a href="#allele_group">allele_group</a></li>
	<li><a href="#tagged_variation_feature">tagged_variation_feature</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="population_genotype" style="font-size:11pt;font-weight:bold">population_genotype</span></td>
<td style="width:100px;text-align:right"><a id="a_population_genotype" style="cursor:pointer;text-decoration:underline" onclick="show_hide('population_genotype')">Show</a> columns</td>
</tr></table>
<p>This table stores alleles and frequencies for variations in given populations.<p>

	<div id="div_population_genotype" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>population_genotype_id</b></td><td>int(10)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
		<tr class="bg2"><td><b>variation_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#variation">variation</a> table.</td><td>key: variation_idx</td></tr>
		<tr class="bg1"><td><b>subsnp_id</b></td><td>int(15)</td><td>NULL</td><td>Foreign key references to the subsnp_handle table.</td><td>key: subsnp_idx</td></tr>
		<tr class="bg2"><td><b>allele_1</b></td><td>varchar(25000)</td><td></td><td>First allele in the genotype.</td><td></td></tr>
		<tr class="bg1"><td><b>allele_2</b></td><td>varchar(25000)</td><td></td><td>Second allele in the genotype.</td><td></td></tr>
		<tr class="bg2"><td><b>frequency</b></td><td>float</td><td></td><td>Frequency of the genotype in the population.</td><td></td></tr>
		<tr class="bg1"><td><b>sample_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#population">population</a> table.</td><td>key: sample_idx</td></tr>
		<tr class="bg2"><td><b>count</b></td><td>int(10)</td><td>NULL</td><td>Number of individuals who have this genotype, in this population.</td><td></td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#population">population</a></li>
	<li><a href="#variation">variation</a></li>
	<li><a href="#subsnp_handle">subsnp_handle</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="population_structure" style="font-size:11pt;font-weight:bold">population_structure</span></td>
<td style="width:100px;text-align:right"><a id="a_population_structure" style="cursor:pointer;text-decoration:underline" onclick="show_hide('population_structure')">Show</a> columns</td>
</tr></table>
<p>This table stores hierarchical relationships between populations by relating them as populations and sub-populations.<p>

	<div id="div_population_structure" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>super_population_sample_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the population table.</td><td>unique<br />key: sub_pop_sample_idx</td></tr>
		<tr class="bg2"><td><b>sub_population_sample_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the population table.</td><td>unique<br />key: sub_pop_sample_idx</td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#population">population</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="protein_info" style="font-size:11pt;font-weight:bold">protein_info</span></td>
<td style="width:100px;text-align:right"><a id="a_protein_info" style="cursor:pointer;text-decoration:underline" onclick="show_hide('protein_info')">Show</a> columns</td>
</tr></table>
<p>Contains information about each translation in the ensembl proteome, used by the nsSNP prediction tables<p>

	<div id="div_protein_info" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>protein_info_id</b></td><td>int(10)</td><td></td><td>Primary key</td><td>primary key</td></tr>
		<tr class="bg2"><td><b>transcript_stable_id</b></td><td>varchar(128)</td><td></td><td>The stable ID of the transcript from which this protein is translated</td><td>key: transcript_idx</td></tr>
		<tr class="bg1"><td><b>transcript_version</b></td><td>smallint</td><td></td><td>The version of the transcript</td><td>key: transcript_idx</td></tr>
		<tr class="bg2"><td><b>translation_md5</b></td><td>char(32)</td><td></td><td>A hexidecimal string representing the MD5 hash of the protein sequence</td><td></td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#protein_position">protein_position</a></li>
	<li><a href="#sift_prediction">sift_prediction</a></li>
	<li><a href="#polyphen_prediction">polyphen_prediction</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="protein_position" style="font-size:11pt;font-weight:bold">protein_position</span></td>
<td style="width:100px;text-align:right"><a id="a_protein_position" style="cursor:pointer;text-decoration:underline" onclick="show_hide('protein_position')">Show</a> columns</td>
</tr></table>
<p>Table with a row for each position in every ensembl translation, used by the nsSNP prediction tables<p>

	<div id="div_protein_position" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>protein_position_id</b></td><td>int(10)</td><td></td><td>Primary key</td><td>primary key</td></tr>
		<tr class="bg2"><td><b>protein_info_id</b></td><td>int(10)</td><td></td><td>Foreign key into the <a href="#protein_info">protein_info</a> table, identifying the relevant protein</td><td>key: pos_idx</td></tr>
		<tr class="bg1"><td><b>position</b></td><td>mediumint</td><td></td><td>The coordinate in the protein sequence</td><td>key: pos_idx</td></tr>
		<tr class="bg2"><td><b>amino_acid</b></td><td>char(1)</td><td></td><td>The amino acid at this position</td><td></td></tr>
		<tr class="bg1"><td><b>sift_median_conservation</b></td><td>float</td><td></td><td>The median conservation at this position, as calculated by SIFT</td><td></td></tr>
		<tr class="bg2"><td><b>sift_num_sequences_represented</b></td><td>smallint</td><td></td><td>The number of sequences that SIFT found at this position in its multiple alignment</td><td></td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#protein_info">protein_info</a></li>
	<li><a href="#sift_prediction">sift_prediction</a></li>
	<li><a href="#polyphen_prediction">polyphen_prediction</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="read_coverage" style="font-size:11pt;font-weight:bold">read_coverage</span></td>
<td style="width:100px;text-align:right"><a id="a_read_coverage" style="cursor:pointer;text-decoration:underline" onclick="show_hide('read_coverage')">Show</a> columns</td>
</tr></table>
<p>This table stores the read coverage in the resequencing of individuals. Each row contains an individual ID, chromosomal coordinates and a read coverage level.<p>

	<div id="div_read_coverage" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>seq_region_id</b></td><td>int(10)</td><td></td><td>Foreign key references <a href="#seq_region">seq_region</a> in core db. ers to the seq_region which this variant is on, which may be a chromosome, a clone, etc...</td><td>key: seq_region_idx</td></tr>
		<tr class="bg2"><td><b>seq_region_start</b></td><td>int</td><td></td><td>The start position of the variation on the <a href="#seq_region">seq_region</a>.</td><td>key: seq_region_idx</td></tr>
		<tr class="bg1"><td><b>seq_region_end</b></td><td>int</td><td></td><td>The end position of the variation on the <a href="#seq_region">seq_region</a>.</td><td></td></tr>
		<tr class="bg2"><td><b>level</b></td><td>tinyint</td><td></td><td>Minimum number of reads.</td><td></td></tr>
		<tr class="bg1"><td><b>sample_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#individual">individual</a> table.</td><td></td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#individual">individual</a></li>
	<li><a href="#seq_region">seq_region</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="sample" style="font-size:11pt;font-weight:bold">sample</span></td>
<td style="width:100px;text-align:right"><a id="a_sample" style="cursor:pointer;text-decoration:underline" onclick="show_hide('sample')">Show</a> columns</td>
</tr></table>
<p>Sample is used as a generic catch-all term to cover individuals, populations and strains; it contains a name and description, as well as a size if applicable to the population.<p>

	<div id="div_sample" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>sample_id</b></td><td>int(10)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
		<tr class="bg2"><td><b>name</b></td><td>varchar(255)</td><td></td><td>Name of the sample (can be an individual or a population name).</td><td>key: name_idx</td></tr>
		<tr class="bg1"><td><b>size</b></td><td>int</td><td></td><td>Number of individual in the sample.</td><td></td></tr>
		<tr class="bg2"><td><b>description</b></td><td>text</td><td></td><td>Description of the sample.</td><td></td></tr>
		<tr class="bg1"><td><b>display</b></td><td>enum('REFERENCE',<br />'DEFAULT',<br />'DISPLAYABLE',<br />'UNDISPLAYABLE',<br />'LD',<br />'MARTDISPLAYABLE')</td><td>'UNDISPLAYABLE'</td><td>Information used by the website: samples with little information are filtered from some web displays.</td><td></td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#individual">individual</a></li>
	<li><a href="#population">population</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="sample_synonym" style="font-size:11pt;font-weight:bold">sample_synonym</span></td>
<td style="width:100px;text-align:right"><a id="a_sample_synonym" style="cursor:pointer;text-decoration:underline" onclick="show_hide('sample_synonym')">Show</a> columns</td>
</tr></table>
<p>Used to store alternative names for populations when data comes from multiple sources.<p>

	<div id="div_sample_synonym" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>sample_synonym_id</b></td><td>int(10)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
		<tr class="bg2"><td><b>sample_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#sample">sample</a> table.</td><td>key: sample_idx</td></tr>
		<tr class="bg1"><td><b>source_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#source">source</a> table.</td><td></td></tr>
		<tr class="bg2"><td><b>name</b></td><td>varchar(255)</td><td></td><td>Name of the synonym (a different <b>sample_id</b>).</td><td></td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#sample">sample</a></li>
	<li><a href="#population">population</a></li>
	<li><a href="#source">source</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="seq_region" style="font-size:11pt;font-weight:bold">seq_region</span></td>
<td style="width:100px;text-align:right"><a id="a_seq_region" style="cursor:pointer;text-decoration:underline" onclick="show_hide('seq_region')">Show</a> columns</td>
</tr></table>
<p>This table stores the relationship between Ensembl's internal coordinate system identifiers and traditional chromosome names.<p>

	<div id="div_seq_region" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>seq_region_id</b></td><td>INT(10)</td><td></td><td>Primary key. Foreign key references seq_region in core db. Refers to the seq_region which this variant is on, which may be a chromosome, a clone, etc...</td><td>primary key</td></tr>
		<tr class="bg2"><td><b>name</b></td><td>VARCHAR(40)</td><td></td><td>The name of this sequence region.</td><td>unique key: name_idx</td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#variation_feature">variation_feature</a></li>
	<li><a href="#variation_group_feature">variation_group_feature</a></li>
	<li><a href="#flanking_sequence">flanking_sequence</a></li>
	<li><a href="#compressed_genotype_single_bp">compressed_genotype_single_bp</a></li>
	<li><a href="#read_coverage">read_coverage</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="sift_prediction" style="font-size:11pt;font-weight:bold">sift_prediction</span></td>
<td style="width:100px;text-align:right"><a id="a_sift_prediction" style="cursor:pointer;text-decoration:underline" onclick="show_hide('sift_prediction')">Show</a> columns</td>
</tr></table>
<p>Stores the SIFT prediction for every possible amino acid substitution in the ensembl proteome<p>

	<div id="div_sift_prediction" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>sift_prediction_id</b></td><td>int(10)</td><td></td><td>Primary key</td><td>primary key</td></tr>
		<tr class="bg2"><td><b>protein_position_id</b></td><td>int(10)</td><td></td><td>Foreign key into the <a href="#protein_position">protein_position</a> table identifying the protein and position that this prediction applies to</td><td>key: pos_aa_idx</td></tr>
		<tr class="bg1"><td><b>amino_acid</b></td><td>char(1)</td><td></td><td>The substituted amino acid</td><td>key: pos_aa_idx</td></tr>
		<tr class="bg2"><td><b>prediction</b></td><td>enum('tolerated', 'deleterious')</td><td></td><td>The qualitative SIFT prediction for this substitution</td><td></td></tr>
		<tr class="bg1"><td><b>score</b></td><td>float</td><td></td><td>The SIFT score for this substitution</td><td></td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#protein_info">protein_info</a></li>
	<li><a href="#protein_position">protein_position</a></li>
	<li><a href="#polyphen_prediction">polyphen_prediction</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="source" style="font-size:11pt;font-weight:bold">source</span></td>
<td style="width:100px;text-align:right"><a id="a_source" style="cursor:pointer;text-decoration:underline" onclick="show_hide('source')">Show</a> columns</td>
</tr></table>
<p>This table contains details of the source from which a variation is derived. Most commonly this is NCBI's dbSNP; other sources include SNPs called by Ensembl.<p>

	<div id="div_source" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>source_id</b></td><td>int(10)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
		<tr class="bg2"><td><b>name</b></td><td>varchar(255)</td><td></td><td>Name of the source. e.g. "dbSNP"</td><td></td></tr>
		<tr class="bg1"><td><b>version</b></td><td>int</td><td></td><td>Version number of the source (if available). e.g. "132"</td><td></td></tr>
		<tr class="bg2"><td><b>description</b></td><td>varchar(255)</td><td></td><td>Description of the source.</td><td></td></tr>
		<tr class="bg1"><td><b>url</b></td><td>varchar(255)</td><td></td><td>URL of the source.</td><td></td></tr>
		<tr class="bg2"><td><b>type</b></td><td>ENUM('chip')</td><td>NULL</td><td>Define the type of the source, e.g. 'chip'</td><td></td></tr>
		<tr class="bg1"><td><b>somatic_status</b></td><td>ENUM ('germline','somatic','mixed')</td><td>'germline'</td><td>Indicates if this source includes somatic or germline mutations, or a mixture</td><td></td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#variation">variation</a></li>
	<li><a href="#variation_synonym">variation_synonym</a></li>
	<li><a href="#variation_feature">variation_feature</a></li>
	<li><a href="#variation_annotation">variation_annotation</a></li>
	<li><a href="#variation_group">variation_group</a></li>
	<li><a href="#allele_group">allele_group</a></li>
	<li><a href="#sample_synonym">sample_synonym</a></li>
	<li><a href="#httag">httag</a></li>
	<li><a href="#structural_variation">structural_variation</a></li>
	<li><a href="#study">study</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="structural_variation" style="font-size:11pt;font-weight:bold">structural_variation</span></td>
<td style="width:100px;text-align:right"><a id="a_structural_variation" style="cursor:pointer;text-decoration:underline" onclick="show_hide('structural_variation')">Show</a> columns</td>
</tr></table>
<p>This table stores information about structural variation features.<p>

	<div id="div_structural_variation" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>structural_variation_id</b></td><td>int(10)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
		<tr class="bg2"><td><b>seq_region_id</b></td><td>int(10)</td><td></td><td>Foreign key references <a href="#seq_region">seq_region</a> in core db. Refers to the seq_region which this variant is on, which may be a chromosome, a clone, etc...</td><td>key: pos_idx</td></tr>
		<tr class="bg1"><td><b>seq_region_start</b></td><td>int(11)</td><td></td><td>The start position of the variation on the <a href="#seq_region">seq_region</a>.</td><td>key: pos_idx</td></tr>
		<tr class="bg2"><td><b>seq_region_end</b></td><td>int(11)</td><td></td><td>The end position of the variation on the <a href="#seq_region">seq_region</a>.</td><td>key: pos_idx</td></tr>
		<tr class="bg1"><td><b>seq_region_strand</b></td><td>tinyint(4)</td><td></td><td>The orientation of the variation on the <a href="#seq_region">seq_region</a>.</td><td></td></tr>
		<tr class="bg2"><td><b>variation_name</b></td><td>varchar(255)</td><td>NULL</td><td>The external identifier or name of the variation. e.g. "esv9549".</td><td>key: name_idx</td></tr>
		<tr class="bg1"><td><b>source_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#source">source</a> table.</td><td></td></tr>
		<tr class="bg2"><td><b>study_id</b></td><td>int(10)</td><td>NULL</td><td>Foreign key references to the <a href="#study">study</a> table.	</td><td>key: study_idx</td></tr>
		<tr class="bg1"><td><b>class</b></td><td>varchar(255)</td><td>NULL</td><td>The type of structural variation feature e.g. "SV"</td><td></td></tr>
		<tr class="bg2"><td><b>bound_start</b></td><td>int(11)</td><td>NULL</td><td>The 5'-most bound defined for the feature.</td><td></td></tr>
		<tr class="bg1"><td><b>bound_end</b></td><td>int(11)</td><td>NULL</td><td>The 3'-most bound defined for the feature.</td><td></td></tr>
		<tr class="bg2"><td><b>allele_string</b></td><td>longtext</td><td></td><td>The variant allele, where known.</td><td></td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#source">source</a></li>
	<li><a href="#study">study</a></li>
	<li><a href="#seq_region">seq_region</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="study" style="font-size:11pt;font-weight:bold">study</span></td>
<td style="width:100px;text-align:right"><a id="a_study" style="cursor:pointer;text-decoration:underline" onclick="show_hide('study')">Show</a> columns</td>
</tr></table>
<p>This table contains details of the studies. The studies information can come from internal studies (DGVa, EGA) or from external studies (Uniprot, NHGRI, ...).<p>

	<div id="div_study" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>study_id</b></td><td>int(10)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
		<tr class="bg2"><td><b>source_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#source">source</a> table.</td><td>key: source_idx</td></tr>
		<tr class="bg1"><td><b>name</b></td><td>varchar(255)</td><td></td><td>Name of the study. e.g. "EGAS00000000001"</td><td></td></tr>
		<tr class="bg2"><td><b>description</b></td><td>varchar(255)</td><td>NULL</td><td>Description of the study.</td><td></td></tr>
		<tr class="bg1"><td><b>url</b></td><td>varchar(255)</td><td>NULL</td><td>URL to find the study data (http or ftp).</td><td></td></tr>
		<tr class="bg2"><td><b>external_reference</b></td><td>varchar(255)</td><td>NULL</td><td>The pubmed/id or project name associated with this study.</td><td></td></tr>
		<tr class="bg1"><td><b>study_type</b></td><td>set('GWAS')</td><td></td><td>Displays if a study comes from a genome-wide association study or not.</td><td></td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#source">source</a></li>
	<li><a href="#variation_annotation">variation_annotation</a></li>
	<li><a href="#structural_variation">structural_variation</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="subsnp_handle" style="font-size:11pt;font-weight:bold">subsnp_handle</span></td>
<td style="width:100px;text-align:right"><a id="a_subsnp_handle" style="cursor:pointer;text-decoration:underline" onclick="show_hide('subsnp_handle')">Show</a> columns</td>
</tr></table>
<p>This table contains the SubSNP(ss) ID and the name of the submitter handle of dbSNP.<p>

	<div id="div_subsnp_handle" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>subsnp_id</b></td><td>int(11)</td><td></td><td>Primary key. It corresponds to the subsnp identifier (ssID) from dbSNP.<br />This ssID is stored in this table without the "ss" prefix. e.g. "120258606" instead of "ss120258606".</td><td>primary key</td></tr>
		<tr class="bg2"><td><b>handle</b></td><td>varchar(20)</td><td></td><td>The name of the dbSNP handler who submitted the ssID.<br />Name of the synonym (a different <b>sample_id</b>).</td><td></td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#allele">allele</a></li>
	<li><a href="#failed_variation">failed_variation</a></li>
	<li><a href="#population_genotype">population_genotype</a></li>
	<li><a href="#sample">sample</a></li>
	<li><a href="#variation_synonym">variation_synonym</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="supporting_structural_variation" style="font-size:11pt;font-weight:bold">supporting_structural_variation</span></td>
<td style="width:100px;text-align:right"><a id="a_supporting_structural_variation" style="cursor:pointer;text-decoration:underline" onclick="show_hide('supporting_structural_variation')">Show</a> columns</td>
</tr></table>
<p>This table stores the name of the supporting evidence for the structural variants (e.g. DGVa structural variants).<p>

	<div id="div_supporting_structural_variation" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>supporting_structural_variation_id</b></td><td>int(10)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
		<tr class="bg2"><td><b>name</b></td><td>varchar(255)</td><td></td><td>The identifier or name of the supporting evidence.</td><td>unique key: name_idx</td></tr>
		<tr class="bg1"><td><b>structural_variation_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#structural_variation">structural_variation</a> table.</td><td>key: structural_variation_idx</td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#structural_variation">structural_variation</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="tagged_variation_feature" style="font-size:11pt;font-weight:bold">tagged_variation_feature</span></td>
<td style="width:100px;text-align:right"><a id="a_tagged_variation_feature" style="cursor:pointer;text-decoration:underline" onclick="show_hide('tagged_variation_feature')">Show</a> columns</td>
</tr></table>
<p>This table lists variation features that are tagged by another variation feature. Tag pairs are defined as having an r<sup>2</sup> &gt; 0.99.<p>

	<div id="div_tagged_variation_feature" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>variation_feature_id</b></td><td>INT(10)</td><td></td><td>Primary key. Foreign key references to the <a href="#variation_feature">variation_feature</a> table.</td><td>primary key</td></tr>
		<tr class="bg2"><td><b>sample_id</b></td><td>INT(10)</td><td></td><td>Primary key. Foreign key references to the <a href="#sample">sample</a> table.</td><td>primary key</td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#variation_feature">variation_feature</a></li>
	<li><a href="#population">population</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="tmp_individual_genotype_single_bp" style="font-size:11pt;font-weight:bold">tmp_individual_genotype_single_bp</span></td>
<td style="width:100px;text-align:right"><a id="a_tmp_individual_genotype_single_bp" style="cursor:pointer;text-decoration:underline" onclick="show_hide('tmp_individual_genotype_single_bp')">Show</a> columns</td>
</tr></table>
<p>his table is only needed for create master schema when run healthcheck system. Needed for other species, but human, so keep it.<p>

	<div id="div_tmp_individual_genotype_single_bp" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>variation_id</b></td><td>int(10)</td><td></td><td>Primary key. Foreign key references to the <a href="#variation">variation</a> table.</td><td>key: variation_idx</td></tr>
		<tr class="bg2"><td><b>subsnp_id</b></td><td>int(15)</td><td></td><td>Foreign key references to the <a href="#subsnp_handle">subsnp_handle</a> table.</td><td>key: subsnp_idx</td></tr>
		<tr class="bg1"><td><b>allele_1</b></td><td>char(1)</td><td></td><td>One of the alleles of the genotype, e.g. "TAG".</td><td></td></tr>
		<tr class="bg2"><td><b>allele_2</b></td><td>char(1)</td><td></td><td>The other allele of the genotype.</td><td></td></tr>
		<tr class="bg1"><td><b>sample_id</b></td><td>int</td><td></td><td>Foreign key references to the <a href="#individual">individual</a> table.</td><td>key: sample_idx</td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#individual">individual</a></li>
	<li><a href="#variation">variation</a></li>
	<li><a href="#subsnp_handle">subsnp_handle</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="transcript_variation" style="font-size:11pt;font-weight:bold">transcript_variation</span></td>
<td style="width:100px;text-align:right"><a id="a_transcript_variation" style="cursor:pointer;text-decoration:underline" onclick="show_hide('transcript_variation')">Show</a> columns</td>
</tr></table>
<p>This table relates a single allele of a variation_feature to a transcript (see Core documentation). It contains the consequence of the allele e.g. intron_variant, non_synonymous_codon, stop_lost etc, along with the change in amino acid in the resulting protein if applicable.<p>

	<div id="div_transcript_variation" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>transcript_variation_id</b></td><td>int(11)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
		<tr class="bg2"><td><b>feature_stable_id</b></td><td>varchar(128)</td><td>NULL</td><td>Foreign key to core databases. Unique stable id of related transcript.</td><td>key: feature_idx<br />key: somatic_feature_idx</td></tr>
		<tr class="bg1"><td><b>variation_feature_id</b></td><td>int(11)</td><td></td><td>Foreign key references to the <a href="#variation_feature">variation_feature</a> table.</td><td>key: variation_feature_idx</td></tr>
		<tr class="bg2"><td><b>allele_string</b></td><td>text</td><td></td><td>Shows the reference sequence and variant sequence of this allele</td><td></td></tr>
		<tr class="bg1"><td><b>somatic</b></td><td>tinyint(1)</td><td>0</td><td>Flags if the associated variation is known to be somatic</td><td>key: somatic_feature_idx</td></tr>
		<tr class="bg2"><td><b>consequence_types</b></td><td>set (<br />'splice_acceptor_variant',<br />'splice_donor_variant',<br />'complex_change_in_transcript', <br />'stop_lost',<br />'coding_sequence_variant',<br />'non_synonymous_codon',<br />'stop_gained',<br />'synonymous_codon',<br />'frameshift_variant',<br />'nc_transcript_variant',<br />'mature_miRNA_variant',<br />'NMD_transcript_variant',<br />'5_prime_UTR_variant',<br />'3_prime_UTR_variant',<br />'incomplete_terminal_codon_variant',<br />'intron_variant',<br />'splice_region_variant',<br />'5KB_downstream_variant',<br />'500B_downstream_variant',<br />'5KB_upstream_variant',<br />'2KB_upstream_variant',<br />'initiator_codon_change',<br />'stop_retained_variant',<br />'inframe_codon_gain',<br />'inframe_codon_loss',<br />'pre_miRNA_variant'<br /> )</td><td></td><td>The consequence(s) of the variant allele on this transcript.</td><td>key: consequence_type_idx</td></tr>
		<tr class="bg1"><td><b>cds_start</b></td><td>int(11)</td><td></td><td>The start position of variation in cds coordinates.</td><td></td></tr>
		<tr class="bg2"><td><b>cds_end</b></td><td>int(11)</td><td></td><td>The end position of variation in cds coordinates.</td><td></td></tr>
		<tr class="bg1"><td><b>cdna_start</b></td><td>int(11)</td><td></td><td>The start position of variation in cdna coordinates.</td><td></td></tr>
		<tr class="bg2"><td><b>cdna_end</b></td><td>int(11)</td><td></td><td>The end position of variation in cdna coordinates.</td><td></td></tr>
		<tr class="bg1"><td><b>translation_start</b></td><td>int(11)</td><td></td><td>The start position of variation on peptide.</td><td></td></tr>
		<tr class="bg2"><td><b>translation_end</b></td><td>int(11)</td><td></td><td>The end position of variation on peptide.</td><td></td></tr>
		<tr class="bg1"><td><b>codon_allele_string</b></td><td>text</td><td></td><td>The reference and variant codons</td><td></td></tr>
		<tr class="bg2"><td><b>pep_allele_string</b></td><td>text</td><td></td><td>The reference and variant peptides</td><td></td></tr>
		<tr class="bg1"><td><b>hgvs_genomic</b></td><td>text</td><td></td><td>HGVS representation of this allele with respect to the genomic sequence</td><td></td></tr>
		<tr class="bg2"><td><b>hgvs_coding</b></td><td>text</td><td></td><td>HGVS representation of this allele with respect to the CDS</td><td></td></tr>
		<tr class="bg1"><td><b>hgvs_protein</b></td><td>text</td><td></td><td>HGVS representation of this allele with respect to the protein</td><td></td></tr>
		<tr class="bg2"><td><b>polyphen_prediction</b></td><td>enum('unknown', 'benign', 'possibly damaging', 'probably damaging')</td><td>NULL</td><td>The PolyPhen prediction for the effect of this allele on the protein</td><td></td></tr>
		<tr class="bg1"><td><b>sift_prediction</b></td><td>enum('tolerated', 'deleterious')</td><td>NULL</td><td>The SIFT prediction for the effect of this allele on the protein </td><td></td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#variation_feature">variation_feature</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="variation" style="font-size:11pt;font-weight:bold">variation</span></td>
<td style="width:100px;text-align:right"><a id="a_variation" style="cursor:pointer;text-decoration:underline" onclick="show_hide('variation')">Show</a> columns</td>
</tr></table>
<p>This is the schema's generic representation of a variation, defined as a genetic feature that varies between individuals of the same species.The most common type is the single nucleotide variation (SNP) though the schema also accommodates copy number variations (CNVs) and structural variations (SVs).A variation is defined by its flanking sequence rather than its mapped location on a chromosome; a variation may in fact have multiple mappings across a genome.This table stores a variation's name (commonly an ID of the form e.g. rs123456, assigned by dbSNP), along with a validation status and ancestral (or reference) allele.<p>

	<div id="div_variation" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>variation_id</b></td><td>int(10)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
		<tr class="bg2"><td><b>source_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#source">source</a> table.</td><td>key: source_idx</td></tr>
		<tr class="bg1"><td><b>name</b></td><td>varchar(255)</td><td></td><td>Name of the variation. e.g. "rs1333049".</td><td>unique</td></tr>
		<tr class="bg2"><td><b>validation_status</b></td><td>SET('cluster','freq',<br />'submitter','doublehit',<br />'hapmap','1000Genome',<br />'failed','precious')</td><td></td><td>Variant discovery method and validation from dbSNP.</td><td></td></tr>
		<tr class="bg1"><td><b>ancestral_allele</b></td><td>varchar(255)</td><td>NULL</td><td>Taken from dbSNP to show ancestral allele for the variation.</td><td></td></tr>
		<tr class="bg2"><td><b>flipped</b></td><td>tinyint(1)</td><td>NULL</td><td>This is set to 1 if the variant is flipped from the negative to the positive strand during import.</td><td></td></tr>
		<tr class="bg1"><td><b>class_attrib_id</b></td><td>int(10)</td><td>0</td><td>Class of the variation, key into the <a href="#attrib">attrib</a> table</td><td></td></tr>
		<tr class="bg2"><td><b>somatic</b></td><td>tinyint(1)</td><td>0</td><td>flags whether this variation is known to be somatic or not</td><td></td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#variation_synonym">variation_synonym</a></li>
	<li><a href="#flanking_sequence">flanking_sequence</a></li>
	<li><a href="#failed_variation">failed_variation</a></li>
	<li><a href="#variation_feature">variation_feature</a></li>
	<li><a href="#variation_group_variation">variation_group_variation</a></li>
	<li><a href="#allele">allele</a></li>
	<li><a href="#allele_group_allele">allele_group_allele</a></li>
	<li><a href="#individual_genotype_multiple_bp">individual_genotype_multiple_bp</a></li>
	<li><a href="#attrib">attrib</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="variation_annotation" style="font-size:11pt;font-weight:bold">variation_annotation</span></td>
<td style="width:100px;text-align:right"><a id="a_variation_annotation" style="cursor:pointer;text-decoration:underline" onclick="show_hide('variation_annotation')">Show</a> columns</td>
</tr></table>
<p>This table stores information linking genotypes and phenotypes. It stores various fields pertaining to the study conducted, along with the associated gene, risk allele frequency and a p-value.<p>

	<div id="div_variation_annotation" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>variation_annotation_id</b></td><td>int(10)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
		<tr class="bg2"><td><b>variation_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#variation">variation</a> table.</td><td>key: variation_idx</td></tr>
		<tr class="bg1"><td><b>phenotype_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#phenotype">phenotype</a> table.</td><td>key: phenotype_idx</td></tr>
		<tr class="bg2"><td><b>study_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#study">study</a> table.</td><td>key: study_idx</td></tr>
		<tr class="bg1"><td><b>associated_gene</b></td><td>varchar(255)</td><td>NULL</td><td>Common gene(s) name(s) associated to the variation.</td><td></td></tr>
		<tr class="bg2"><td><b>associated_variant_risk_allele</b></td><td>varchar(255)</td><td>NULL</td><td>Allele associated to the phenotype.</td><td></td></tr>
		<tr class="bg1"><td><b>variation_names</b></td><td>varchar(255)</td><td>NULL</td><td>Name of the variation. e.g. "rs1333049".</td><td></td></tr>
		<tr class="bg2"><td><b>risk_allele_freq_in_controls</b></td><td>double</td><td>NULL</td><td>Risk allele frequency.</td><td></td></tr>
		<tr class="bg1"><td><b>p_value</b></td><td>double</td><td>NULL</td><td>P value of the association phenotype/variation.</td><td></td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#variation">variation</a></li>
	<li><a href="#phenotype">phenotype</a></li>
	<li><a href="#source">source</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="variation_feature" style="font-size:11pt;font-weight:bold">variation_feature</span></td>
<td style="width:100px;text-align:right"><a id="a_variation_feature" style="cursor:pointer;text-decoration:underline" onclick="show_hide('variation_feature')">Show</a> columns</td>
</tr></table>
<p>This table represents mappings of variations to genomic locations. It stores an allele string representing the different possible alleles that are found at that locus e.g. "A/T" for a SNP, as well as a "worst case" consequence of the mutation. It also acts as part of the relationship between variations and transcripts.<p>

	<div id="div_variation_feature" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>variation_feature_id</b></td><td>int(10)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
		<tr class="bg2"><td><b>seq_region_id</b></td><td>int(10)</td><td></td><td>Foreign key references <a href="#seq_region">seq_region</a> in core db. Refers to the seq_region which this variant is on, which may be a chromosome, a clone, etc...</td><td>key: pos_idx</td></tr>
		<tr class="bg1"><td><b>seq_region_start</b></td><td>int</td><td></td><td>The start position of the variation on the <a href="#seq_region">seq_region</a>.</td><td>key: pos_idx</td></tr>
		<tr class="bg2"><td><b>seq_region_end</b></td><td>int</td><td></td><td>The end position of the variation on the <a href="#seq_region">seq_region</a>.</td><td>key: pos_idx</td></tr>
		<tr class="bg1"><td><b>seq_region_strand</b></td><td>tinyint</td><td></td><td>The orientation of the variation on the <a href="#seq_region">seq_region</a>.</td><td></td></tr>
		<tr class="bg2"><td><b>variation_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#variation">variation</a> table.</td><td>key: variation_idx</td></tr>
		<tr class="bg1"><td><b>allele_string</b></td><td>varchar(50000)</td><td></td><td>This is a denormalised string taken from the alleles in the allele table associated with this variation. The reference allele (i.e. one on the reference genome comes first).</td><td></td></tr>
		<tr class="bg2"><td><b>variation_name</b></td><td>varchar(255)</td><td></td><td>A denormalisation taken from the variation table. This is the name or identifier that is used for displaying the feature.</td><td></td></tr>
		<tr class="bg1"><td><b>map_weight</b></td><td>int</td><td></td><td>The number of times that this variation has mapped to the genome. This is a denormalisation as this particular feature is one example of a mapped location. This can be used to limit the the features that come back from a query.</td><td></td></tr>
		<tr class="bg2"><td><b>flags</b></td><td>SET('genotyped')</td><td></td><td>Flag to filter the selection of variations.</td><td></td></tr>
		<tr class="bg1"><td><b>source_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the source table.</td><td></td></tr>
		<tr class="bg2"><td><b>validation_status</b></td><td>SET(<br />'cluster',<br />'freq',<br />'submitter',<br />'doublehit',<br />'hapmap',<br />'1000Genome',<br />'precious'<br /> )</td><td></td><td>SET('cluster', 'freq', 'submitter', 'doublehit', 'hapmap', '1000Genome', 'precious')	Variant discovery method and validation from dbSNP.</td><td></td></tr>
		<tr class="bg1"><td><b>consequence_type</b></td><td>SET (<br />'intergenic_variant',<br />'splice_acceptor_variant',<br />'splice_donor_variant',<br />'complex_change_in_transcript', <br />'stop_lost',<br />'coding_sequence_variant',<br />'non_synonymous_codon',<br />'stop_gained',<br />'synonymous_codon',<br />'frameshift_variant',<br />'nc_transcript_variant',<br />'mature_miRNA_variant',<br />'NMD_transcript_variant',<br />'5_prime_UTR_variant',<br />'3_prime_UTR_variant',<br />'incomplete_terminal_codon_variant',<br />'intron_variant',<br />'splice_region_variant',<br />'5KB_downstream_variant',<br />'500B_downstream_variant',<br />'5KB_upstream_variant',<br />'2KB_upstream_variant',<br />'initiator_codon_change',<br />'stop_retained_variant',<br />'inframe_codon_gain',<br />'inframe_codon_loss',<br />'miRNA_target_site_variant',<br />'pre_miRNA_variant',<br />'regulatory_region_variant',<br />'increased_binding_affinity',<br />'decreased_binding_affinity',<br />'binding_site_variant'<br /> )</td><td>'intergenic_variant'</td><td>The SO accession(s) representing the 'worst' consequence(s) of the variation in a transcript or regulatory region</td><td></td></tr>
		<tr class="bg2"><td><b>variation_set_id</b></td><td>SET (<br />'1','2','3','4','5','6','7','8',<br />'9','10','11','12','13','14','15','16',<br />'17','18','19','20','21','22','23','24',<br />'25','26','27','28','29','30','31','32',<br />'33','34','35','36','37','38','39','40',<br />'41','42','43','44','45','46','47','48',<br />'49','50','51','52','53','54','55','56',<br />'57','58','59','60','61','62','63','64'<br /> )</td><td>''</td><td>The variation feature can belong to a <a href="#variation_set">variation_set</a>.</td><td>key: variation_set_idx</td></tr>
		<tr class="bg1"><td><b>class_attrib_id</b></td><td>int(10)</td><td>0</td><td>Class of the variation, key in the <a href="#attrib">attrib</a> table</td><td></td></tr>
		<tr class="bg2"><td><b>somatic</b></td><td>tinyint(1)</td><td>0</td><td>Flags whether this variation_feature is somatic or germline</td><td></td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#variation">variation</a></li>
	<li><a href="#tagged_variation_feature">tagged_variation_feature</a></li>
	<li><a href="#transcript_variation">transcript_variation</a></li>
	<li><a href="#seq_region">seq_region</a></li>
	<li><a href="#attrib">attrib</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="variation_group" style="font-size:11pt;font-weight:bold">variation_group</span></td>
<td style="width:100px;text-align:right"><a id="a_variation_group" style="cursor:pointer;text-decoration:underline" onclick="show_hide('variation_group')">Show</a> columns</td>
</tr></table>
<p>This table represents the equivalent of a variation for a multi-marker variation.<p>

	<div id="div_variation_group" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>variation_group_id</b></td><td>int(10)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
		<tr class="bg2"><td><b>name</b></td><td>varchar(255)</td><td></td><td>The code or name of this variation_group.</td><td>unique</td></tr>
		<tr class="bg1"><td><b>source_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#source">source</a> table.</td><td></td></tr>
		<tr class="bg2"><td><b>type</b></td><td>enum('haplotype', 'tag')</td><td></td><td>Type of the variation group.</td><td></td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#variation_group_variation">variation_group_variation</a></li>
	<li><a href="#allele_group">allele_group</a></li>
	<li><a href="#variation">variation</a></li>
	<li><a href="#source">source</a></li>
	<li><a href="#httag">httag</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="variation_group_feature" style="font-size:11pt;font-weight:bold">variation_group_feature</span></td>
<td style="width:100px;text-align:right"><a id="a_variation_group_feature" style="cursor:pointer;text-decoration:underline" onclick="show_hide('variation_group_feature')">Show</a> columns</td>
</tr></table>
<p>This table represents the equivalent of a variation_feature for multi-marker variations, mapping a haplotype to a chromosomal coordinate system.<p>

	<div id="div_variation_group_feature" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>variation_group_feature_id</b></td><td>int(10)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
		<tr class="bg2"><td><b>seq_region_id</b></td><td>int(10)</td><td></td><td>Foreign key references <a href="#seq_region">seq_region</a> in core db. Refers to the seq_region which this variant is on, which may be a chromosome, a clone, etc...</td><td>key: pos_idx</td></tr>
		<tr class="bg1"><td><b>seq_region_start</b></td><td>int</td><td></td><td>The start position of the variation on the <a href="#seq_region">seq_region</a>.</td><td>key: pos_idx</td></tr>
		<tr class="bg2"><td><b>seq_region_end</b></td><td>int</td><td></td><td>The end position of the variation on the <a href="#seq_region">seq_region</a>.</td><td></td></tr>
		<tr class="bg1"><td><b>seq_region_strand</b></td><td>tinyint</td><td></td><td>The orientation of the variation on the <a href="#seq_region">seq_region</a>.</td><td></td></tr>
		<tr class="bg2"><td><b>variation_group_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#variation_group">variation_group</a> table.</td><td>key: variation_group_idx</td></tr>
		<tr class="bg1"><td><b>variation_group_name</b></td><td>varchar(255)</td><td></td><td>Name of the variation group.</td><td></td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#variation_group">variation_group</a></li>
	<li><a href="#seq_region">seq_region</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="variation_group_variation" style="font-size:11pt;font-weight:bold">variation_group_variation</span></td>
<td style="width:100px;text-align:right"><a id="a_variation_group_variation" style="cursor:pointer;text-decoration:underline" onclick="show_hide('variation_group_variation')">Show</a> columns</td>
</tr></table>
<p>This table represents an individual variation that makes up a multi-marker variation, and resolves the many-to-many relationship between variation and variation_group.<p>

	<div id="div_variation_group_variation" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>variation_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#variation">variation</a> table.</td><td>unique<br />key: variation_idx</td></tr>
		<tr class="bg2"><td><b>variation_group_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#variation_group">variation_group</a> table.</td><td>unique<br />key: variation_idx</td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#variation_group">variation_group</a></li>
	<li><a href="#variation">variation</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="variation_set" style="font-size:11pt;font-weight:bold">variation_set</span></td>
<td style="width:100px;text-align:right"><a id="a_variation_set" style="cursor:pointer;text-decoration:underline" onclick="show_hide('variation_set')">Show</a> columns</td>
</tr></table>
<p>This table containts the name of sets and subsets of variations stored in the database. It usually represents the name of the project or subproject where a group of variations has been identified.<p>

	<div id="div_variation_set" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>variation_set_id</b></td><td>int(10)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
		<tr class="bg2"><td><b>name</b></td><td>VARCHAR(255)</td><td></td><td>Name of the set e.g. "Phenotype-associated variations".</td><td>key: name_idx</td></tr>
		<tr class="bg1"><td><b>description</b></td><td>TEXT</td><td></td><td>Description of the set.</td><td></td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#variation_set_variation">variation_set_variation</a></li>
	<li><a href="#variation_set_structure">variation_set_structure</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="variation_set_structure" style="font-size:11pt;font-weight:bold">variation_set_structure</span></td>
<td style="width:100px;text-align:right"><a id="a_variation_set_structure" style="cursor:pointer;text-decoration:underline" onclick="show_hide('variation_set_structure')">Show</a> columns</td>
</tr></table>
<p>This table stores hierarchical relationships between variation sets by relating them as variation sets and variation subsets.<p>

	<div id="div_variation_set_structure" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>variation_set_super</b></td><td>int(10)</td><td></td><td>Primary key. Foreign key references to the <a href="#variation_set">variation_set</a> table.</td><td>primary key<br />key: sub_idx</td></tr>
		<tr class="bg2"><td><b>variation_set_sub</b></td><td>int(10)</td><td></td><td>Primary key. Foreign key references to the <a href="#variation_set">variation_set</a> table.</td><td>primary key<br />key: sub_idx</td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#variation_set">variation_set</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="variation_set_variation" style="font-size:11pt;font-weight:bold">variation_set_variation</span></td>
<td style="width:100px;text-align:right"><a id="a_variation_set_variation" style="cursor:pointer;text-decoration:underline" onclick="show_hide('variation_set_variation')">Show</a> columns</td>
</tr></table>
<p>A table for mapping variations to variation_sets.<p>

	<div id="div_variation_set_variation" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>variation_id</b></td><td>int(10)</td><td></td><td>Primary key. Foreign key references to the <a href="#variation">variation</a> table.</td><td>primary key<br />key: variation_set_idx</td></tr>
		<tr class="bg2"><td><b>variation_set_id</b></td><td>int(10)</td><td></td><td>Primary key. Foreign key references to the <a href="#variation_set">variation_set</a> table.</td><td>primary key<br />key: variation_set_idx</td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#variation">variation</a></li>
	<li><a href="#variation_set">variation_set</a></li>
</ul>

<br />
<table style="border: 2px groove #CCCCCC;height:10px;background-color:#FAFAFF"><tr style="vertical-align:middle;height:10px">
<td style="width:500px;text-align:left;height:10px"><span id="variation_synonym" style="font-size:11pt;font-weight:bold">variation_synonym</span></td>
<td style="width:100px;text-align:right"><a id="a_variation_synonym" style="cursor:pointer;text-decoration:underline" onclick="show_hide('variation_synonym')">Show</a> columns</td>
</tr></table>
<p>This table allows for a variation to have multiple IDs, generally given by multiple sources.<p>

	<div id="div_variation_synonym" style="display:none">

	<table style="border:1px outset #222222">
		<tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
		<tr class="bg1"><td><b>variation_synonym_id</b></td><td>int(10)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
		<tr class="bg2"><td><b>variation_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the variation table.</td><td>key: variation_idx</td></tr>
		<tr class="bg1"><td><b>subsnp_id</b></td><td>int(15)</td><td></td><td>Foreign key references to the subsnp_handle table.</td><td>key: subsnp_idx</td></tr>
		<tr class="bg2"><td><b>source_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the source table.</td><td>unique<br />key: source_idx</td></tr>
		<tr class="bg1"><td><b>name</b></td><td>varchar(255)</td><td></td><td>Name of the synonym variation. e.g. 'rs1333049'. The corresponding variation ID of this variation is different from the one stored in the column variation_id.</td><td>unique</td></tr>
		<tr class="bg2"><td><b>moltype</b></td><td>varchar(50)</td><td></td><td>...</td><td></td></tr>
</table>
</div>
<p><b>See also:</b></p>
<ul>
	<li><a href="#source">source</a></li>
	<li><a href="#variation">variation</a></li>
	<li><a href="#subsnp_handle">subsnp_handle</a></li>
</ul>


</body>
</html>
