<html>
<head>
<title>FAQ</title>
<meta name="order" content="4" />
</head>
<body>
<div>
	
	<div style="float:right;"><img src="/img/vep_logo.png"/></div>
	
	<h1 id="top"><span style="color:#006">Variant Effect Predictor</span> <img src="/i/16/help.png"> <span style="color:#666">FAQ</span></h1>
	<hr/>
    
    <p> For any questions not covered here, please send an email to the Ensembl
    developer's mailing list, <a
    href="mailto:dev@ensembl.org">dev@ensembl.org</a> (public) or contact the
    Ensembl Helpdesk at <a
    href="mailto:helpdesk@ensembl.org">helpdesk@ensembl.org</a> (private) </p>

<h2 id="general">General questions</h2>

<p style="font-weight:bold;"> Q: Why don't I see any co-located variations when
using species X? </p>

<p> A: Ensembl only has variation databases for a subset of all Ensembl species
- see <a href="/info/genome/variation/data_description.html#source">this
  document</a> for details.</p>

<p>&nbsp;</p>

<p style="font-weight:bold;"> Q: Why has my insertion/deletion variant encoded in VCF
disappeared from the VEP output? </p>

<p> A: Ensembl treats unbalanced variants differently to VCF - your variant
hasn't disappeared, it may have just changed slightly! You can solve this by
giving your variants a unique identifier in the third column of the VCF file.
See <a href="vep_formats.html#vcf">here</a> for a full discussion.</p>

<p>&nbsp;</p>



<p style="font-weight:bold;"> Q: Why do I see so many lines of output for each variant in my input? </p>

<p> A: While it can be convenient to search for a easy, one word answer to the
question "What is the consequence of this variant?", in reality biology does not
make it this simple! Many genes have more than one transcript, so the VEP
provides a prediction for each transcript that a variant overlaps. The VEP
script can help here; the --canonical and --ccds options indicate which
transcripts are canonical and belong to the CCDS set respectively, while
--per_gene, --summary and --most_severe allow you to give a more summary level
assessment per variant.</p>

<p> Furthermore, several "compound" consequences are also possible - if, for
example, a variant falls in the final few bases of an exon, it may be considered
to affect a splicing site, in addition to possibly affecting the coding
sequence. </p>

<p> Since we cannot possibly predict the exact biology of what will happen, what
we provide is the most conservative estimate that covers all reasonable
scenarios. It is up to you, the user, to interpret this information!</p>

<hr/>

<h2 id="web">Web VEP questions</h2>

<p style="font-weight:bold;"> Q: How do I access the web version of the Variant
Effect Predictor? </p>

<p> A: You can find the web VEP on the <a href="/tools.html">Tools</a> page, or
from any species-specific page by clicking the blue "Manage your data" link
under the left-hand menu, then "Variant Effect Predictor". </p>

<p>&nbsp;</p>


<p style="font-weight:bold;"> Q: I have selected a VCF file in the file upload
field, but nothing happens when I click the blue "Next" button. Why? </p>

<p> A: Ensure that you have selected VCF as the input file format, and that your
VCF file is formatted <a
href="http://www.1000genomes.org/wiki/Analysis/Variant%20Call%20Format/vcf-variant-call-format-version-40"
target="_blank">correctly</a>.</p>

<p>&nbsp;</p>


<p style="font-weight:bold;"> Q: I uploaded a file with 1000 variants, but some
of them seem to be missing from the output? </p>

<p> A: Due to a limitation in the servers underlying Ensembl, only 750 variants
in one file can be processed at once. Consider splitting your file into smaller
chunks, or using the <a href="script/index.html">standalone perl script</a>.</p>

<p>&nbsp;</p>



<p style="font-weight:bold;"> Q: Why is the output I get for my input file
different when I use the web VEP and the VEP script? </p>

<p> A: Ensure that you are passing equivalent arguments to the script that you
are using in the web version. If you are sure this is still a problem, please
report it on the Ensembl developers mailing list, <a
href="mailto:dev@ensembl.org">dev@ensembl.org</a>.</p>
<hr/>

<h2 id="script">VEP script questions</h2>

<p style="font-weight:bold;"> Q: Why do I see the following error? </p>

<pre class="code">
Could not connect to database homo_sapiens_core_63_37 as user anonymous using [DBI:mysql:database=homo_sapiens_core_63_37;host=ensembldb.ensembl.org;port=5306] as a locator:
Unknown MySQL server host 'ensembldb.ensembl.org' (2) at $HOME/src/ensembl/modules/Bio/EnsEMBL/DBSQL/DBConnection.pm line 290.

-------------------- EXCEPTION --------------------
MSG: Could not connect to database homo_sapiens_core_63_37 as user anonymous using [DBI:mysql:database=homo_sapiens_core_63_37;host=ensembldb.ensembl.org;port=5306] as a locator:
Unknown MySQL server host 'ensembldb.ensembl.org' (2)
</pre>

<p> A: By default the VEP script is configured to connect to the public MySQL
server at ensembldb.ensembl.org. Occasionally the server may break connection
with your script, which causes this error. This can happen when the server is
busy, or due to various network issues. Consider using a <a
href="script/vep_cache.html#local">local copy of the database</a>, or the <a
href="script/vep_cache.html#cache">caching system</a>. </p>

<p>&nbsp;</p>

<p style="font-weight:bold;"> Q: Can I use the VEP script on Windows? </p>

<p> Yes - see the <a href="script/vep_download.html#windows">documentation</a> for a
few different ways to get the VEP running on Windows. </p>

<p>&nbsp;</p>

<p style="font-weight:bold;"> Q: Can I download all of the SIFT and/or PolyPhen predictions? </p>

<p> A: The Ensembl Variation database and the human VEP cache file contain
precalculated SIFT and PolyPhen predictions for every possible amino acid change
in every translated protein product in Ensembl. Since these data are huge, we
store them in a compressed format. The best approach to extract them is to use
our Perl API.</p>

<p> The format in which the data are stored in our database is described <a
href="/info/genome/variation/predicted_data.html#nsSNP_data_format">here</a></p>

<p> The simplest way to access these matrices is to use an API script to fetch a
ProteinFunctionPredictionMatrix for your protein of interest and then call its
'get_prediction' method to get the score for a particular position and amino
acid, looping over all possible amino acids for your position. There is some
detailed documentation on this class in the API documentation <a
href="http://www.ensembl.org/info/docs/Doxygen/variation-api/classBio_1_1EnsEMBL_1_1Variation_1_1ProteinFunctionPredictionMatrix.html">here</a>.</p>

<p>You would need to work out which peptide position your codon maps to, but
there are methods in the <a
href="http://www.ensembl.org/info/docs/Doxygen/variation-api/classBio_1_1EnsEMBL_1_1Variation_1_1TranscriptVariation.html">TranscriptVariationAllele</a>
class that should help you (probably translation_start and translation_end).</p>

<p>&nbsp;</p>

</div>

</body>
</html>
