<html>
<head>
<title>Compara API</title>
</head>

<body>

<h1>Compara API</h1> 


<h2>Database Description</h2>

<p>
The Ensembl Compara multi-species database stores the results of genome-wide species comparisons calculated for each data release.
The database includes:
</p>

<ul>

<li>
Comparative genomics:
  <ul>
  <li>Whole genome alignments</li>
  <li>Synteny regions</li>
  </ul>
</li>

<li>
Comparative proteomics:
  <ul>
  <li>Orthologue predictions</li>
  <li>Paralogue predictions</li>
  <li>Protein family clusters</li>
  </ul>
</li>

</ul>


<h2>Database Schema</h2>

<p>
The table layout of the database is explained in the following document:
</p>

<ul>
<li>
<a href="./compara_schema.html">Compara Schema Description</a>
</li>
</ul>


<h2>Perl API</h2>

<p>
A comprehensive Perl Application Programme Interface (API) provides efficient access to the Ensembl Compara database.
</p>

<ul>
<li>
<a href="/info/docs/api/api_installation.html">Compara Perl API Installation</a>: A step-by-step installation guide for all Ensembl Perl APIs.
</li>

<li>
<a href="/info/docs/Pdoc/ensembl-compara/index.html">Compara Perl API Documentation</a>: A complete reference to the objects and methods used in the Compara database API.
</li>

<li>
<a href="./compara_tutorial.html">Compara Perl API Tutorial</a>: An introduction to the underlying concepts of the Compara database API.
</li>

<li>
<a href="/Help/Faq?kw=compara" class="modal_link">Compara FAQs</a>
</li>


</ul>


<h2 id="support">Ensembl Software Support</h2>

<p>
Ensembl is an open project and we would like to encourage correspondence and discussions on any subject on any aspect of Ensembl.
Please see the
<a href="http://www.ensembl.org/info/about/contact/">Ensembl Contacts</a> page for suitable options getting in touch with us.
</p>

</body>
</html>
