<html>
<head>
<title>Mouse Genome Sequence Data</title>
</head>
<body>


<h2>Brief Tour of the Mouse Genome Sequence Data</h2>

<p>
  The Mouse Genome Sequencing Consortium (MGSC) seeks to develop a
  comprehensive BAC map and complete sequence of the mouse genome.  In
  the first year we have completed the physical map and generated
  about 6X whole genome shotgun coverage.  Assemblies of earlier
  stages of coverage are available and an assembly of the full WGS
  data set is underway.  The groups are now shifting to BAC clones to
  add additional shotgun data and to complete the sequence.
</p>

<p>
  The MGSC is generating whole genome shotgun sequence data, which is
  being assembled and analyzed by a variety of <a href="/Mus_musculus/credits.html">laboratories world-wide</a>.  
Here we provide links to sites that give access to the
  sequence and results of its analysis.  The sequence is being
  generated continuously, and each assembly and analysis is based on a
  particular <a
  href="http://bio.cse.psu.edu/genome/hummus/freezes.html">data
  freeze.</a>.  Ensembl has links to reference assemblies: <a
  href="ftp://ftp.ensembl.org/pub/current_mouse/data/">ftp site</a>
  and <a href="http://trace.ensembl.org/">Trace Repository</a>.  NCBI
  provides
  <a href="ftp://ftp.ncbi.nlm.nih.gov/pub/TraceDB/mus_musculus">
  centralized access </a> to sequence data and assemblies.
</p>

<h3 class="boxed">Finding the mouse contig matching a given sequence.</h3>

<p>Searches against various assemblies of the mouse sequence
  can be performed at either the
  <a href="http://www.ncbi.nlm.nih.gov/genome/seq/MmBlast.html">
  Megablast server</a> at NCBI or the
  <a href="http://www.ensembl.org/Multi/blastview?species=Mus_musculus&method=ssaha">SSAHA server </a> at Ensembl.  Information about cDNA clones from
  RIKEN is also available at <a
  href="http://fantom.gsc.riken.go.jp/db/">FANTOM-DB</a>
</p>

<h3 class="boxed">Comparisons with human genomic sequence.</h3> 
<p>To our knowledge, all
  human-mouse comparative analyses are using the August 2001 assembly
  of the human genome. Analyses include:
</p>

<ul>
 <li>
  New gene predictions by <a
  href="http://bio.math.berkeley.edu/slam/mouse/" >SLAM</a>
  (U. California Berkeley).  This is one of several projects to
  improve gene prediction by incorporating mouse-human homology.
 </li>

 <li>
  For positions of the human genome that align with the mouse, use
  the <a href="http://genome.ucsc.edu">UCSC Genome Browser</a> site.
  The August 2001 browser has translated BLAT alignments on the human
  assembly vs. the Arachne.2 mouse assembly push unassembled reads.
  The April 2001 browser has translated BLAT, and nucleotide-level
  Pattern Hunter and Exonerate alignments of the human assembly vs. 13
  million mouse reads.
 </li>

 <li>
  Percent identity plots (pips) of blastz alignments of the human
  genome against the mouse, using the <a
  href="http://www.genome.org/cgi/content/full/10/4/577">pip
  dispenser</a>.  These alignments use the October Phusion assembly
  of mouse.
 </li>

 <li>
  New Godzilla Avid alignments of the MGC v.3 Mouse Assembly against the latest Human Draft Assembly and vizualisation of this comparative analysis with the Vista Genome Browser and Vista Tracks: <a href="http://pipeline.lbl.gov">http://pipeline.lbl.gov</a>. (UC.Berkeley and Lawrence Berkeley Lab.)
 </li>
</ul>

</body>
</html>

