<html>
<head>
<meta name="order" content="99" />
<title>Projects using Ensembl</title>
</head>

<body>

<h1>Projects using Ensembl</h1>


<p>
The Ensembl project is both a source of genome sequence related data and an open
      source software system that can be used to organise any such data.
</p>

<p class="center">
<img src="/img/empowered.png" alt="'Powered by Ensembl' logo" width="98" height="20" />
</p>

<h2 id="collaborations">Collaborations</h2>

<p>
Ensembl are active collaborators in a number of projects, contributing code, 
know-how and a platform from which to distribute data.
</p>

<table class="ss tint">
<tr>
    <th>Project</th>
    <th>Description</th>
    <th>Code used/data provided</th>
</tr>

<tr class="bg2">
    <td><strong><a href="http://www.1000genomes.org/">1000 Genomes</a></strong></td>
    <td>A web browser and ftp sites are provided to access human genetic variation catalogued by the project.</td>
    <td>Web front-end derived from Ensembl webcode </td>
</tr>

<tr>
    <td><strong><a href="http://projects.ensembl.org/blueprint/">Blueprint</a></strong></td>
    <td>Epigenomics project analysing samples from healthy and diseased individuals.</td>
    <td>Track Hub available on the Ensembl browser</td>
</tr>

<tr class="bg2">
    <td><strong><a href="http://geval.sanger.ac.uk/index.html">gEVAL</a></strong></td>
    <td>This browser can be used to inspect the reference assemblies of human, mouse
          and zebrafish being created by the GRC</td>
    <td>Customised webcode, pipelines (not genebuild) and compara</td>
</tr>

<tr>
    <td><strong><a href="http://projects.ensembl.org/heroic/">HEROIC</a></strong></td>
    <td>Functional genomics data for Mouse (Mus musculus)</td>
    <td>Data viewable on Ensembl via DAS.</td>
</tr>

<tr class="bg2">
    <td><strong><a href="http://projects.ensembl.org/neandertal/">Neanderthal Genome Browser</a></strong></td>
    <td>Preliminary assembly of <i>Homo sapiens neanderthalensis</i>.</td>
    <td>Web front-end derived from Ensembl webcode, Ensembl schema databases.</td>
</tr>

<tr>
    <td><strong><a href="http://parasite.wormbase.org/index.html">WormBase Parasite</a></strong></td>
    <td>Website presenting draft genome sequences for helminths (parasitic nematodes and flatworms)</td>
    <td>Web front-end derived from Ensembl webcode, Ensembl schema databases, Ensembl Compara, REST API and (slightly modified) BioMart</td>
</tr>

<tr class="bg2">
    <td><strong><a href="http://projects.ensembl.org/nextgen/">NextGen</a></strong></td>
    <td>A collaborative project investigating biodiversity of livestock species.</td>
    <td>Variant call sets produced using the Ensembl VEP, viewable in the Ensembl browser.</td>
</tr>

<tr>
    <td><strong><a href="http://projects.ensembl.org/quantomics/">Quantomics</a></strong></td>
    <td>Large-scale project to analyse sequence and variation in livestock genomes.</td>
    <td>Sequence alignments created using the Ensembl Compara pipelines.</td>
</tr>

</table>

<h3>External projects</h3>

<p>
Our open access data and open source code mean that many projects are able to 
make use of Ensembl data and software without our active involvement. We're 
happy to list those we know about here, but if your project is 
<b><i><span style="color:#3366bb">e</span><span style="color:#cc0000">!</span></i></b>mpowered 
and you're not on this list, let us know.
</p>

<!-- ADD NEW ENTRIES AT THE TOP, TO KEEP TABLE STRIPING SIMPLE -->
<table class="ss tint">
<tr>
    <th>Project</th>
    <th>Description</th>
    <th>Code used</th>
</tr>
	
<tr>
    <td><strong><a href="http://lepbase.org">LepBase</a></strong></td>
    <td>Lepbase is the Lepidopteran genome database, providing an ensembl genome browser, a blast server, a download server and more as a set of resources for moth and butterfly genome research.</td>
    <td>Customised webcode</td>
</tr>
	
<tr class="bg2">
   <td><strong><a href="http://cancer.sanger.ac.uk/cosmic">COSMIC (Catalogue of Somatic Mutations in Cancer)</a></strong></td>
    <td>Web display of somatic sequence variant/mutation data</td>
    <td>VEP</td>
</tr>

<tr>
    <td><strong><a href="http://www.pombase.org">PomBase</a></strong></td>
    <td>A resource for <em>Schizosaccharomyces pombe</em> that includes structural and functional annotation, literature curation and access to large-scale data sets.</td>
    <td>Customised webcode, pipeline, database schema and API
</tr>

<tr class="bg2">
    <td><strong><a href="http://www.germonline.org/index.html">GermOnline</a></strong></td>
    <td>Microarray expression database focused on germline development</td>
    <td>Webcode and underlying data</td>
</tr>

<tr>
    <td><strong><a href="http://bgee.unil.ch">Bgee</a></strong></td>
    <td>Comparison of gene expression patterns between species</td>
    <td>Ensembl data and orthologues</td>
</tr>
	
<tr class="bg2">
    <td><strong><a href="http://search.cpan.org/~ggallone/Bio-Homology-InterologWalk/">InterologWalk Perl Modules</a></strong></td>
    <td>Perl modules to determine protein-protein interactions from orthology and interaction data.</td>
    <td>API, compara</td>
</tr>

<tr>
    <td><strong><a href="http://www.vectorbase.org/">VectorBase</a></strong></td>
    <td>Bioinformatic Resource Centre for Invertebrate Vectors of Human Pathogens</td>
    <td>Customised Ensembl webcode, Ensembl Compara pipeline, Ensembl annotation pipeline</td>
</tr>

<tr class="bg2">
    <td><strong><a href="http://www.gramene.org/">Gramene</a></strong></td>
    <td>A Comparative Mapping Resource for Grains</td>
    <td>Customised webcode, Ensembl Compara pipeline,  Ensembl database used for annotation, import of the Arabidopsis Ensembl database from NASC.</td>
</tr>

<tr>
    <td><strong><a href="http://www.sigenae.org">Sigenae</a></strong></td>
    <td>Sigenae EST ContigBrowser</td>
    <td></td>
</tr>

<tr class="bg2">
    <td><strong><a href="http://labs.pathology.jhu.edu/nauen/trac/">TraC: Transcript Consensus</a></strong></td>
    <td>Web-based tool for visual comparison of alternative splicing isoforms</td>
    <td>Can draw from Ensembl annotation and sequence data</td>
</tr>

<tr>
    <td><strong><a href="http://www.orthomam.univ-montp2.fr">OrthoMaM</a></strong></td>
    <td>A database of alignments and trees based on orthologous exons and CDSs for mammalian species. 
        </td>
    <td>1-to-1 orthologues extracted from Ensembl.</td>
</tr>

<tr class="bg2">
    <td><strong><a href="http://www.sanger.ac.uk/science/tools/otter">Otter</a></strong></td>
    <td>Database backend for interactive curation of annotation</td>
    <td>Otter is an extension of the Ensembl database schema</td>
</tr>

<tr>
  <td><strong><a href="http://selectome.unil.ch/">Selectome</a></strong></td>
    <td>Database of positive selection.</td>
    <td>Uses Ensembl gene trees, gene data and xrefs extracted via the APIs.</td>
</tr>

<tr class="bg2">
    <td><strong><a href="https://cellxgene.cziscience.com/">CZ CELLxGENE Discover</a></strong></td>
    <td>A free-to-use online suite of tools that help researchers discover, download, and analyze single-cell datasets from modalities that include gene expression, chromatin accessibility, DNA methylation, and spatial transcriptomics.</td>
    <td>All data in the portal have been standardized and annotated using an ontological shared vocabulary for cell and gene metadata. It uses an Ensembl gene reference across all data.</td>
</tr>

</table>

<p class="center">
<img src="/img/empowered.png" alt="'Powered by Ensembl' logo" width="98" height="20" />
</p>

<p>If you are using Ensembl code in your project, you might like to <a href="/img/empowered_lg.png">download the 'empowered' logo</a> (high-resolution PNG, suitable for use in print).</p>

<h2 id="contribs">User-contributed code</h2>

<p>Whilst we have developed a comprehensive Perl API in-house, we welcome contributions
in other programming languages from the community.</p>

<h3>Ruby</h3>

<p>A Ruby API has been developed by Jan Aerts. A small example, showing what it can do 
(including coordinate transformation and reflecting on the types of associations for a 
given class) can be found on the wiki page.</p>

<ul>
<li><a href="https://github.com/jandot/ruby-ensembl-api">repository</a></li>
<li><a href="https://github.com/jandot/ruby-ensembl-api/wiki">wiki</a></li>
</ul>

<h3>Java</h3>

<p>
An open source Java API (JEnsembl) has been developed by the
Bioinformatics Group at
<a href='http://www.roslin.ed.ac.uk/'>The Roslin Institute</a>. The code has
been designed to be version-aware (a single installation works against
current and previous releases) and is hosted, together with
<a href='http://jensembl.sourceforge.net/apidocs/'>documentation</a>,
<a href='http://jensembl.sourceforge.net/example.html'>examples</a>
and
<a href ='http://jensembl.sourceforge.net/contact.html'>contact
information for the Roslin
group</a>, at <a href='http://jensembl.sourceforge.net'>SourceForge</a>.
</p>

<p>Note that we do not support these user-contributed packages - please contact
the original developer if you have any questions or comments.</p>


</body>
</html>
