<html>
<head>
<meta name="order" content="6" />
<title>How to use BioMart</title>
</head>

<body>

<h1>How to use BioMart</h1>
<h3>Summary</h3>
<ol>
        <li><a href="#introduction">What is BioMart</a></li>
        <li><a href="#databaseselection">Select a mart database (a type of data)</a></li>
        <li><a href="#datasetselection">Select a mart dataset (a species)</a></li>
        <li><a href="#filterselection">Filter your mart query (query restriction and input data)</a></li>
        <li><a href="#attributeselection">Select mart attributes (desired output)</a></li>
        <li><a href="#resultselection">Display and retrieve your query</a></li>
</ol>

<a name="introduction"></a>
<h2>What is BioMart</h2>
<a href="/biomart/martview?VIRTUALSCHEMANAME=default&ATTRIBUTES=hsapiens_gene_ensembl.default.feature_page.ensembl_gene_id|hsapiens_gene_ensembl.default.feature_page.ensembl_transcript_id|hsapiens_gene_ensembl.default.feature_page.hgnc_id|hsapiens_gene_ensembl.default.feature_page.hgnc_symbol&FILTERS=hsapiens_gene_ensembl.default.filters.chromosome_name.1&VISIBLEPANEL=resultspanel"><img src="/img/biomart.png" style="border: 1px solid black" alt="screenshot of the biomart interface" /></a>
<p>
BioMart is an easy-to-use web-based tool that allows extraction of data without any programming knowledge or understanding of the underlying database structure.
You can navigate through the BioMart web interface using the left panel. Filters and attributes can be selected in the right panel. A summary of your choices is also displayed in the left panel.
</p>
<a name="databaseselection"></a>
<h2>1. Select a mart database (a type of data)</h2>
<p><div class="image-wrap" style=""><img src="/img/biomart_database_dropdown.png" style="border: 1px solid black" /></div></p>
<p>First, select a mart database which will correspond to the type of data you are interested in. In Ensembl, you can choose data from one of our four mart databases:</p>
<ul>
<li> Ensembl Genes: This mart contains the Ensembl gene set and allows you to retrieve Ensembl genes, transcripts and proteins as well as external references, microarrays, protein domains, structure,  sequences, variants (only variants mapped to Ensembl Transcripts) and homology data.</li>
<li> Ensembl Variation: This mart allows you to retrieve germline and somatic variants as well as germline and somatic structural variants. This mart also contains variants' phenotypes, citations, synomyms, consequences and flanking sequences; you can also retrieve Ensembl genes, transcripts, regulatory and motif features mapped to variants.</li>
<li> Ensembl Regulation: This mart allows you to retrieve regulatory features, evidence and segments, miRNA target regions, binding motifs and other regulatory regions.</li>
<li> Vega: This mart contains the Ensembl Vega gene set (manual annotation coming from Havana) and allows you to retrieve Ensembl Vega genes, transcripts and proteins as well as external references, structures, sequences and protein domains</li>
</ul>
<a name="datasetselection"></a>
<h2>2. Select a mart dataset (a species)</h2>
<p>Next, select a mart dataset which correspond to the species you are interested in and want to retrieve data from.</p>
<p><div class="image-wrap" style=""><img src="/img/biomart_dataset_dropdown.png" style="border: 1px solid black" /></div></p>
The marts are available for the following species:
<ul>
<li>Ensembl Genes: All the Ensembl species, full list on the <a href="/info/about/species.html">species page</a>.</li>
<li>Ensembl Variation: We only have a subset of our species which have variation data available, full list on the <a href="/info/genome/variation/data_description.html#sources">variation species page</a>.</li>
<li> Ensembl Regulation: We have regulation data available for human, mouse and fruitfly.</li>
<li> Vega: We have data available for human, mouse, rat, pig and zebrafish.</li>
</ul>
<a name="filterselection"></a>
<h2>3. Filter your mart query (query restriction and input data)</h2>
<p>BioMart allows you to restrict your query with information that you know, e.g: input a list of IDs, restrict to a region. You can access the filter page by clicking on the "Filters" button located on the left panel. Filters are organised into different sections, clicking on the "+/-" boxes will expand/collapse a section and display its content.</p>
<p><div class="image-wrap" style=""><img src="/img/biomart_filter_page.png" style="border: 1px solid black" /></div></p>
<p> In the following image, I have expanded the "Region" section by clicking on the "+" box as indicated by the number 1. Then, I have selected "1" in the chromosome list to restrict my query to all the genes located on human chromosome 1. The left panel now displays "Chromosome: 1" as indicated by the number 2. Clicking on the "Count" button as indicated by number 3 will display the number of Ensembl genes that the query will return next to "Dataset" as indicated by the number 4. This mean that we have 5406 Ensembl Genes located on the human chromosome 1. 
<p><div class="image-wrap" style=""><img src="/img/biomart_filter_page2.png" style="border: 1px solid black" /></div></p>
<a name="attributeselection"></a>
<h2>4. Select mart attributes (desired output)</h2>
<p> By clicking on the "Attributes" button on the left panel, you will access the mart attribute page. This page allows you to select your desired output; the default output is "Ensembl Gene ID" and "Ensembl Transcript ID" in the Ensembl Genes mart. The attributes are organised in pages that you can access by selecting the radio buttons as displayed by the red box on the image below.</p>
<p><div class="image-wrap" style=""><img src="/img/biomart_attribute_page.png" style="border: 1px solid black" /></div></p>
<a name="resultselection"></a>
<h2>5. Display and retrieve your query</h2>
<p>Clicking on the "Results" button will bring you to the mart result page. This page will, by default, show you a preview of the first 10 results of your query in HTML format. The number of results previewed and format can be changed as indicated by number 1 on the image below. You can also automatically remove all the duplicated results from your query by clicking on the "Unique results only" button as indicated by number 2. If you are happy with your query, you can use the "Export all results to" section as indicated by number 3 (select a format and click on the "Go" button) to download your results.</p>
<p><div class="image-wrap" style=""><img src="/img/biomart_result_page.png" style="border: 1px solid black" /></div></p>
<p>You can also request the results of your query to be sent to you by email. To do this, select "Compressed web file (notify by email)" in the first dropdown in the red box on the image below, then select your desired format, type your email address in the "Email notification to" box and press the "Go" button. BioMart will compile the result of your query in the background and send you a link to the compressed file by email.
<p><div class="image-wrap" style=""><img src="/img/biomart_result_page2.png" style="border: 1px solid black" /></div></p>
<p>BioMart tutorials, multiple dataset query, Perl API, RESTful and Bioconductor R package access documentations can be found on <a href="/info/data/biomart/index.html">the following page</a>. If you have any questions, you can email Ensembl helpdesk <a href="/info/about/contact/index.html">using our web form</a>.</p> 
</body>
</html>
