<html>
<head>
<meta name="order" content="2" />
<title>GFF3 File Format</title>
</head>

<body>

<h1>GFF3 File Format - Definition and supported options</h1>

<p>The GFF (General Feature Format) format consists of one line per feature, each containing 9 columns of data, plus optional track definition lines. The following documentation is based on the 
<a href="http://gmod.org/wiki/GFF3">Version 3 specifications</a>.</p>

<ul>
<li><a href="#fields">Fields</a></li>
<li><a href="#moreinfo">More information</a></li>
</ul>


<h2 id="fields">Fields</h2>

<p>The first line of a GFF3 file must be a comment that identifies the version, e.g.</p>

<pre class="code">
##gff-version 3
</pre>

<p>Fields <strong>must</strong> be tab-separated. Also, all but the final field in each feature line must contain a value; "empty" columns should be denoted with a '.'</p>

<ol>
<li><strong>seqid</strong> - name of the chromosome or scaffold; chromosome names can be given with or without the 'chr' prefix. <b>Important note</b>: the seq ID must be one used within Ensembl, i.e. a standard chromosome name or an Ensembl identifier such as a scaffold ID, without any additional content such as species or assembly. See the example GFF output below.</li>
<li><strong>source</strong> - name of the program that generated this feature, or the data source (database or project name)</li>
<li><strong>type</strong> - type of feature. Must be a term or accession from the SOFA sequence ontology</li>
<li><strong>start</strong> - Start position of the feature, with sequence numbering starting at 1.</li>
<li><strong>end</strong> - End position of the feature, with sequence numbering starting at 1.</li>
<li><strong>score</strong> - A floating point value.</li>
<li><strong>strand</strong> - defined as + (forward) or - (reverse).</li>
<li><strong>phase</strong> - One of '0', '1' or '2'. '0' indicates that the first base of the feature is the first base of a codon, '1' that the second base is the first base of a codon, and so on..</li>
<li><strong>attributes</strong> - A semicolon-separated list of tag-value pairs, providing additional information about each feature. Some of these tags are predefined, e.g. ID, Name, Alias, Parent - see the <a href="http://gmod.org/wiki/GFF3">GFF documentation</a> for more details.</li>
</ol>

<p>Note that where the attributes contain Parent identifiers, these will be used by Ensembl to display the features as joined blocks.</p>

<pre class="code">
##gff-version 3
ctg123 . mRNA            1300  9000  .  +  .  ID=mrna0001;Name=sonichedgehog
ctg123 . exon            1300  1500  .  +  .  ID=exon00001;Parent=mrna0001
ctg123 . exon            1050  1500  .  +  .  ID=exon00002;Parent=mrna0001
ctg123 . exon            3000  3902  .  +  .  ID=exon00003;Parent=mrna0001
ctg123 . exon            5000  5500  .  +  .  ID=exon00004;Parent=mrna0001
ctg123 . exon            7000  9000  .  +  .  ID=exon00005;Parent=mrna0001
</pre>

<h2 id="moreinfo">More information</h2>

<p>For more information about this file format, see the <a href="http://gmod.org/wiki/GFF3">documentation</a> on the GMOD wiki.</p>

</body>
</html>
