<html>
<head>
<meta name="order" content="1" />
<title>Large File Formats</title>
</head>

<body>

<h1>Large File Formats</h1>

<p>Ensembl supports a number of large (multi-gigabyte) file formats for datasets that are too large to upload directly to the Ensembl servers. The file remains on the remote server and is accessed via its URL, with only small chunks of data being requested by the Ensembl webcode at any one time.</p>

<p>All of these formats can be viewed in Ensembl by creating a custom data track using the 'Add your data' function, in the lefthand menu when you click "Manage Your Data".</p>

<p><img src="/img/help/large_files.png" alt="" /></p>

<p>When you select one of these formats, the form button will change to read 'Attach'.</p>

[[MOVIE::457]]

<h2 id="hubs">Track Hubs</h2>

<p>In addition to individual files, we also support track hubs (also known as data hubs) 
- collections of remote data files that uses a simple text-based configuration.<p>

<p>For how to set up a hub, see the documentation provided by UCSC, the originators
of this data service:</p>
<ul>
<li><a href="http://genome.ucsc.edu/goldenPath/help/hgTrackHubHelp.html" rel="external">Using UCSC Genome Browser Track Hubs</a></li>
<li><a href="http://hgwdev-tdreszer.cse.ucsc.edu/goldenPath/help/trackDb/trackDbHub.html" rel="external">Hub Track Database Definition</a></li>
</ul>

<p>Please note that we do not currently support all the options documented by UCSC,
though we intend to implement as many as are appropriate to our displays.</p>

<p>Note also that if you change the configuration of your hub, you will need to
shift-reload the Ensembl page to load the new file.</p>

<h2 id="bam-format">BAM format</h2>
<p>BAM is a compressed version of the SAM (Sequence Alignment/Map) binary format. BAM uses an index file to give fast access to small sections of the file..</p>
<p>Additional information about SAM/BAM is available at the <a href="http://samtools.sourceforge.net/">SAMtools development site</a>.


<h2 id="bb-format">bigBED format</h2>

<p>BigBED is an indexed format created from standard <a href="/info/website/upload/bed.html">BED</a> files using the <kbd>bedToBigBed</kbd> utility program. It allows large datasets to be processed much faster than a conventional BED file.</p> 

<h2 id="bw-format">BigWig format</h2>

<p>The BigWig format is designed for dense, continuous data that is intended to be displayed as a graph. Files can be created from WIG or BedGraph files using the appropriate utility program.</p>

<p>Ensembl currently allows the following configuration options to be set when you upload the file:</p>

<ul>
<li>Track colour (10 choices)</li>
</ul>

<h2 id="vcf-format">VCF format</h2>

<p>The VCF format is a tab delimited format for storing variant calls and and individual genotypes. It is able to store all variant calls from single nucleotide variants to large scale insertions and deletions.</p>

<p>More information on this format, which is still under development, can be obtained from the <a href="http://www.1000genomes.org/wiki/Analysis/Variant%20Call%20Format/vcf-variant-call-format-version-40">1000 Genomes Wiki</a>.</p>

<p>Please note the following:</p>
<ul>
<li>You will need to have an index file with the extension .vcf.gz.tbi in the same directory as your data file (and with the same name)</li>
<li>Ensembl does not support the https protocol - please use a <kbd>http://</kbd> or a <kbd>ftp://</kbd> url.</li>
</ul>


<p>In order to produce the indexed vcf file with the .gz.tbi extension you must follow the following steps:</p>
<ul>
<li>Compress your vcf file using <a href="http://sourceforge.net/projects/bgzip/">bgzip</a></li>
<li>Index the vcf.gz file using <a href="http://samtools.sourceforge.net/tabix.shtml">tabix</a>. Use will need to pass the option -p vcf to tabix, for example "/usr/bin/tabix -p vcf my_file.vcf.gz"</li>
</ul>

</body>
</html>
