<html>
<head>
<meta name="order" content="15" />
<title>Variation File Format</title>
</head>

<body>

<h1>Variation File Format - Definition and supported options</h1>

<ul>
<li><a href="#input">Input format</a></li>
<li><a href="#output">Output format</a></li>
</ul>

<p>
The <a href="/info/docs/variation/vep/index.html">Variant Effect Predictor</a>
tool which appears as an option when you click on <strong>Manage your
Data</strong> allows you to upload a set of variation data and predict the
effect of the variants.</p>

<p>Note that the input and output formats are completely different.</p>

<h2 id="input">Input format</h2>

<p>Data must be supplied in a simple <strong>tab-separated</strong> format,
containing five columns, all required:
</p>

<ol>
<li><strong>chromosome</strong> - just the name or number, with no 'chr' prefix</li>
<li><strong>start</strong></li>
<li><strong>end</strong></li>
<li><strong>allele</strong> - pair of alleles separated by a '/', with the reference allele first</li>
<li><strong>strand</strong> - defined as + (forward) or - (reverse).</li>
</ol>

<pre class="code">
1   881907    881906    -/C   +
5   140532    140532    T/C   +
12  1017956   1017956   T/A   +
2   946507    946507    G/C   +
14  19584687  19584687  C/T   -
19  66520     66520     G/A   +
8   150029    150029    A/T   +
</pre>

<p>An insertion is indicated by start coordinate = end coordinate +
1. For example, an insertion of 'C' between nucleotides 12600 and
12601 on the forward strand of chromosome 8 is indicated as
follows:</p>

<pre class="code">
8   12601     12600     -/C   +
</pre>
<p>A deletion is indicated by the exact nucleotide coordinates. For
example, a three base pair deletion of nucleotides 12600, 12601, and
12602 of the reverse strand of chromosome 8 will be:</p>

<pre class="code">
8   12600     12602     CGT/- -
</pre>

<!-- 5,140532,140532,T/C,+
12,1017956,1017956,T/A,+
2,946507,946507,G/C,+
14,19584687,19584687,C/T,-
19,66520,66520,G/A,+
8,150029,150029,A/T,+ -->

<p> The following input file formats are also supported: </p>

<ul>
    <li>Variant Call Format (VCF) - see <a
    href="http://www.1000genomes.org/wiki/Analysis/vcf4.0"
    target="_blank">http://www.1000genomes.org/wiki/Analysis/vcf4.0</a> for
    details.</li>
    <li>Pileup format</li>
    <li>HGVS notations - see <a
    href="http://www.hgvs.org/mutnomen/">http://www.hgvs.org/mutnomen/</a> for
    details. These must be relative to genomic or Ensembl transcript
    coordinates. It is possible, although less reliable, to <a
    href="/info/docs/variation/vep/vep_script.html#hgvs">use notations relative
    to RefSeq transcripts</a> in the VEP script.</li>
    <li>Variant identifiers - these should be e.g. dbSNP rsIDs, or any synonym
    for a variant present in the Ensembl Variation database. See <a
    href="/info/docs/variation/sources_documentation.html">here</a> for a list
    of identifier sources in Ensembl.</li>
</ul>

<p> When using the web VEP, ensure that you have the correct file format
selected from the drop-down menu. The VEP script is able to auto-detect the
format of the input file. </p>

<h2 id="output">Output format</h2>

<p>The tool predicts the consequence of this variation, the amino acid
position and change (if the variation falls within a protein) and the
identifier of known variations that occur at this position. The output
columns are:</p>

<ol>
	<li><strong>Uploaded variation</strong> - as chromosome_start_alleles</li>
	<li><strong>Location</strong> - in standard coordinate format (chr:start or chr:start-end)</li>
	<li><strong>Allele</strong> - the variant allele used to calculate the consequence</li>
	<li><strong>Gene</strong> - Ensembl stable ID of affected gene</li>
	<li><strong>Feature</strong> - Ensembl stable ID of feature</li>
	<li><strong>Feature type</strong> - type of feature. Currently one of Transcript, RegulatoryFeature, MotifFeature.</li>
	<li><strong>Consequence</strong> - consequence type of this variation</li>
	<li><strong>Relative position in cDNA</strong> - base pair position in cDNA sequence</li>
	<li><strong>Relative position in CDS</strong> - base pair position in coding sequence</li>
	<li><strong>Relative position in protein</strong> - amino acid position in protein</li>
	<li><strong>Amino acid change</strong> - only given if the variation affects the protein-coding sequence</li>
	<li><strong>Codons</strong> - the alternate codons with the variant base highlighted as bold (HTML) or upper case (text)</li>
	<li><strong>Corresponding variation</strong> - identifier of existing variation</li>
	<li><strong>Extra</strong> - this column contains extra information as key=value pairs separated by ";". The keys are as follows:
		<ul>
			<li><i>HGNC</i> - the HGNC gene identifier</li>
			<li><i>ENSP</i> - the Ensembl protein identifier of the affected transcript</li>
			<li><i>HGVSc</i> - the HGVS coding sequence name</li>
			<li><i>HGVSp</i> - the HGVS protein sequence name</li>
			<li><i>SIFT</i> - the SIFT prediction and/or score, with both given as prediction(score)</li>
			<li><i>PolyPhen</i> - the PolyPhen prediction and/or score</li>
			<li><i>Condel</i> - the Condel consensus prediction and/or score</li>
			<li><i>MOTIF_NAME</i> - the source and identifier of a transcription factor binding profile aligned at this position</li>
			<li><i>MOTIF_POS</i> - The relative position of the variation in the aligned TFBP</li>
			<li><i>HIGH_INF_POS</i> - a flag indicating if the variant falls in a high information position of a transcription factor binding profile (TFBP)</li>
			<li><i>MOTIF_SCORE_CHANGE</i> - The difference in motif score of the reference and variant sequences for the TFBP</li>
			<li><i>CANONICAL</i> - a flag indicating if the transcript is denoted as the canonical transcript for this gene</li>
			<li><i>CCDS</i> - the CCDS identifer for this transcript, where applicable</li>
			<li><i>INTRON</i> - the intron number (out of total number)</li>
			<li><i>EXON</i> - the exon number (out of total number)</li>
			<li><i>DOMAINS</i> - the source and identifer of any overlapping protein domains</li>
		</ul>
	</li>
</ol>

<p>Empty values are denoted by '-'. Further fields in the Extra column can be
added by <a href="vep_script.html#plugins">plugins</a> or using <a
href="vep_script.html#custom">custom annotations</a> in the VEP script. Output
fields can be configured using the --fields flag when running the VEP script.
</p>

<pre class="code">
11_224088_C/A    11:224088   A  ENSG00000142082  ENST00000525319  Transcript         NON_SYNONYMOUS_CODING   742  716  239  T/N  aCc/aAc  -  SIFT=deleterious(0);PolyPhen=unknown(0)
11_224088_C/A    11:224088   A  ENSG00000142082  ENST00000534381  Transcript         5_PRIME_UTR             -    -    -    -    -        -  -
11_224088_C/A    11:224088   A  ENSG00000142082  ENST00000529055  Transcript         DOWNSTREAM              -    -    -    -    -        -  -
11_224585_G/A    11:224585   A  ENSG00000142082  ENST00000529937  Transcript         INTRONIC,NMD_TRANSCRIPT -    -    -    -    -        -  HGVSc=ENST00000529937.1:c.136-346G>A
22_16084370_G/A  22:16084370 A  -                ENSR00000615113  RegulatoryFeature  REGULATORY_REGION       -    -    -    -    -        -  -
</pre>

<p> The VEP script will also add a header to the output file. This contains
information about the databases connected to, and also a key describing the
key/value pairs used in the extra column. </p>

<pre class="code">
## ENSEMBL VARIANT EFFECT PREDICTOR v2.4
## Output produced at 2012-02-20 16:09:38
## Connected to homo_sapiens_core_66_37 on ensembldb.ensembl.org
## Using API version 66, DB version 66
## Extra column keys:
## CANONICAL    : Indicates if transcript is canonical for this gene
## CCDS         : Indicates if transcript is a CCDS transcript
## HGNC         : HGNC gene identifier
## ENSP         : Ensembl protein identifer
## HGVSc        : HGVS coding sequence name
## HGVSp        : HGVS protein sequence name
## SIFT         : SIFT prediction
## PolyPhen     : PolyPhen prediction
## Condel       : Condel SIFT/PolyPhen consensus prediction
## EXON         : Exon number
## INTRON       : Intron number
## DOMAINS      : The source and identifer of any overlapping protein domains
## MOTIF_NAME   : The source and identifier of a transcription factor binding profile (TFBP) aligned at this position
## MOTIF_POS    : The relative position of the variation in the aligned TFBP
## HIGH_INF_POS : A flag indicating if the variant falls in a high information position of the TFBP
## MOTIF_SCORE_CHANGE : The difference in motif score of the reference and variant sequences for the TFBP
</pre>
</body>
</html>
