<html>

<head>
<title>Ensj FAQ</title>
</head>

<body>

<h2>Frequently asked questions</h2>

<ol>

<li><a href="#example.java">Why doesn't Example.java work?</a></li>

<li><a href="#unknown_database">Why do I get an "unkown database"
exception?</a

<li><a href="#fetch_all_overlap">Is it possible to fetch all feature_ids for features (e.g. genes)
overlapping a location?</a></li>

<li><a href="#bug">I think I've found a bug in ensj, or at least I am getting
stack traces from it, what should I do?</a></li>

<li><a href="#bug_report"/>How do I report a bug?</li>

<li><a href="#data_discrepancy">Why is the data returned via ensj different to what I see on the
web site?</a></li>

<li><a href="#version">What version of the ensj am I running?</a></li>

<li><a href="#source_code">How can I get the source code corresponding to a particular build
(release or snapshot)?</a></li>

<li><a href="#dist_ensj">Can I distribute ensj.jar or ensj-XX.X.jar with my application?</a></li>

<li><a href="#dist_mysql">Can I distribute mysql-connector-java-XXXXXX.jar with my
application?</a></li>

<li><a href="#jdts">Can I connect to ensembl data stored in MS SQL or Sybase
servers?</a></li>

<li><a href="#connection_refused">Why do I (sometimes) get a
connection refused exception when I try and retrieve data from ensembldb.ensembl.org?</a></li>

<li><a href="#orthologues">How can I get all the putative orthologues
for a gene?</a></li>

</ol>

<ol>


<li>
<a name="example.java"/><b>Why doesn't Example.java work?</b>

<p>
This is usually caused by the Example program trying to connect to a
database that has been removed from the mysql server
ensembldb.ensembl.org:3306. This bug only happens in old version of
ensj and upgrading to the <a href="download.html">latest build</a> of
ensj.jar will fix this problem.  </p> </li>


<li>
<a name="unknown_database"/><b>Why do I get an "unkown database" exception?</b>

<p>This means that the database is not available on the database
server. Perhaps the database name is misspelt or the database has been
removed. This sometimes happens if you are using an old configuration
file that refers to an obsolete database.  A list of current databases
on ensembldb.ensembl.org is available in the file
ensembldb.ensembl.org-databases.txt in the <a
href="download.html">latest build</a>. </p> </li>


<li>
<a name="fetch_all_overlap"/><b>Is it possible to fetch all feature_ids for features (e.g. genes)
overlapping a location?</b>

Yes, you can use the <code>fetchInternalIDs(Location)</code> methods
but you might want to consider using the <code>
fetchIterator(Location) </code> methods instead. These provide a very
simple interface with a fast implementation that uses relatively
little memory even for whole genome queries. Examples:

<ul>

<li><code>long[] ids =
geneAdaptor.fetchInternalIDs(location);</code>
</li> 

<li><code>
Iterator iter = geneAdaptor.fetchIterator(location);
</code>

</li>

</ul>


<li>
<a name="bug"/><b>I think I've found a bug in ensj, or at least I am getting
stack traces from it, what should I do?</b>

<p>It is possible that your bug has been fixed or is due to an
  incompatiblity between the versions of ensj.jar, mysql.jar (MySQL
  JDBC driver) or the MySQL database you are using. The first thing to
  do therefore is to make sure you are using the right jars. You can
  download the latest ensj release and a compatible mysql driver from
  the <a
  href="http://cvsweb.sanger.ac.uk/cgi-bin/cvsweb.cgi/ensj-core/snapshot/">download</a>
  section. Note that some users have experienced problems with both
  <b>older</b> and, surprisingly, <b>newer</b> mysql.jars so ensuring
  you are using the version we distribute is an important step in the
  debugging process.
</p>



<p>Secondly, several people have experienced problems using ensj that
were caused by referencing old ensj-XX.Y or mysql-XXX jars in their
CLASSPATH. 

<ul>
<li>
Running your program like this should resolve the problem:
<code>java -cp ensj-XX.Y.jar:mysql-XXX.jar:$CLASSPATH
some.package.YourClass</code>. 
</li>

<li>Add <code>System.out.println(SystemUtil.environmentDump())</code>
to the beginning of your program to discover which jars the ensj and
mysql classes are being loaded from. If the program won't compile or
run after adding this line because of a "SystemUtil.environmentDump()
is missing" error then you are using an old ensj-XX.Y.jar.  </li>
</ul>


<p>If the bug occurs when you use the latest ensj-XX.Y.jar and
mysql-ZZZ.jar then you should <a href="#bug_report">submit a bug
report</a>.</p>

</li>


<li>
<a name="bug_report"/><b>How do I report a bug?</b>

<p>To begin with, read the <a href="#bug">"I think I've found a bug in
ensj"</a> section. Once you are convinced it is a real bug in ensj you
should include these items in a <a href="contact.html"> bug
report</a>:</p> <ul>

<li>A test class with a main() method that demostrates the
bug. Include
<code>System.out.println(SystemUtil.environmentDump())</code> as the
first line of main() which will generate additional debugging information.  </li>

<li>The output generated by the test program running including all
stack traces.</li>

</ul>

</li>


<li>
<a name="data_discrepancy"/><b>Why is the data returned via ensj different to what I see on the
web site?</b>

<p>
This is usually because the ensj driver is pointing at a different
database to the website. Configuring
the driver to point at the same database as the website (the latest)
usually solves this problem.</p>
</li>

<li>
<a name="version"/><b>What version of the ensj am I running?</b>

<p>
It should be in the jar name but if not you can also run this command
to find out: <code>prompt&gt;java org.ensembl.Version</code>.  </p>
</li>

<li>
<a name="source_code"/><b>How can I get the source code corresponding to a particular build
(release or snapshot)?</b>

<p>The source code for the latest release is included in the latest
distribution. Other versions of the source code corresponding to other
builds are available from <a href="download.html#cvs">CVS</a>
using the appropriatte tag. For example, after checking out ensj-core
,<code>cvs up -r release-ensembl-25-0</code> would retrieve all the
files in the 25.0 build (release or snapshot).</p>

</li>


<li>
<a name="dist_ensj"/><b>Can I distribute ensj-XY.Z.jar with my application?</b>

<p>
Yes. Ensj is published under the LGPL so you can redistribute the jar
although you should also include a copy of the LGPL. For instance copy
<a href="http://www.ensembl.org/java/LICENSE">LICENSE</a> to
LICENSE-ensj in the same directory as your own license.
</li>


<li>
<a name="dist_mysql"/><b>Can I distribute mysql-connector-java-XXXXXX.jar with my
application?</b>

<p>
This question should really be answered by reading the MySQL AB <a
href="http://www.mysql.com/company/legal/licensing">license</a>. It
took me a little time to read through this so I'll give my short
interpretation of it here on the grounds that it might save you some
time. This interpretation might be wrong and comes with no warranty so
you should still read the original for yourself. Enough caveats, here's my
take on the license situation: <ul>

<li>If your program is GPL you
can distribute the jar.</li> 

<li>If you have another open source license you
might be able to distribute it. See <a
href="http://www.mysql.com/company/legal/licensing/foss-exception.html">
open source license exceptions</a> for a list of "compatible" open
source licenses.

<li>If you have a non open source license, e.g. commercial, you
probably need to buy a license in order to redistribute the
jar. Alternatively your end users could download the jar themselves
from the MySQLConnector J section of the <a
href="http://dev.mysql.com/downloads">MySQL download page</a>.

</ul>


<li><a name="jdts"/><b>Can I connect to ensembl data stored in MS SQL or Sybase
servers?</b>

<p>Probably. Ensj has been partially tested against an MS SQL database
loaded with ensembl data via the jDTS jdbc driver (<a href="http://jtds.sourceforge.net">http://jtds.sourceforge.net</a>). Sample
configuration properties to put in the ensembl driver config file:
<pre><code>
connection_string=jdbc:jtds:sqlserver://HOST:PORT
jdbc_driver=net.sourceforge.jtds.jdbc.Driver
</code></pre>




<li><a name="connection_refused"/><b>Why do I (sometimes) get a
"connection refused" exception when I try and retrieve data from ensembldb.ensembl.org?</b>

<p>
This happens because too many connection requests have been made from
your IP address. Ensembl's MySQL server ensembldb.ensembl.org:3306
only allows 32 connections per IP address. Note that if several users
connect to the database from behind a NAT firewall they all appear to
have the same IP address.

<li><a name="orthologues"/><b>How can I get all the putative orthologues
for a gene?</b>

<p>Putative orthologues generated by recipricol BLAST hits are stored in
the compara database and can be retrieved via a ComparaDriver. This
example shows how to print all the orthologues for the human gene
<code>ENSG00000118961</code> from
the latest ensembl databases.

<code><pre>  
    ComparaDriver compara = Registry.createDefaultUserRegistry().getGroup("compara").getComparaDriver();
  
    List gdbs = compara.getGenomeDBAdaptor().fetch();
    for (int i = 0; i < gds.size(); i++) {
      GenomeDB gdb =(GenomeDB) gdbs.get(i);
      String targetSpecies = (String) gdb.getName();
      // if we were only interested in hits in chicken...
      // List hits = compara.getMemberAdaptor().fetch("Homo sapiens", new String[]{"ENSG00000118961"}, "Gallus gallus");
      List hits = compara.getMemberAdaptor().fetch("Homo sapiens", new String[]{"ENSG00000118961"}, targetSpecies);
      System.out.print(targetSpecies);
      for (int j = 0; j < hits.size(); j++) {
        FeaturePair p = (FeaturePair) hits.get(j);
	// hitDisplayName is the orthologue gene name; begins with
	// "ENS" if ensembl predicted gene otherwise name is assigned
	// by another group.
        System.out.print("\t" +p.getHitDisplayName());
      }
      System.out.println();
    }
</pre></code>

Here is similar code in jython:

<code><pre>
    genomes = ensembl.ensembl_compara.genomeDBAdaptor.fetch()
    for speciesName in [g.name for g in genomes]:
        orthologues = ensembl.ensembl_compara.memberAdaptor.fetch("Homo Sapiens", ["ENSG00000118961"], speciesName)
        geneNames = [o.hitDisplayName for o in orthologues]
        print geneNames
</pre></code>

</ol>



</body>
</html>
