<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
<head>
  <title>EnsEMBL Compara Schema Documentation</title>
  <meta name="GENERATOR" content="Quanta Plus">
  <meta name="AUTHOR" content="Javier Herrero (jherrero@ebi.ac.uk)">
  <meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1"><style TYPE="text/css">
   TABLE { font-size: small; }
   PRE {font-size: small; }
  </style>
</head>
<body>

<h1>EnsEMBL Compara DB Schema Documentation</h1>

<p>This document refers to the schema of the EnsEMBL Compara version <strong>26.1</strong>.
This is the revision <strong>0.3</strong> of this document.</p>

<hr>
<hr>
<h2>Introduction</h2>

EnsEMBL Compara DB is divided into two parts:
<ul>
<LI>Genomic alignments
<p>This part is dedicated to DNA-DNA alignment. It includes alignment made with different alignment tools. It also
includes a list of syntenic regions.</p>
</LI>
<LI>Othologues and Protein clusters (or families)
<p>This part contains several types of data.</p>
 
<p>The first one corresponds to a list of protein orthologues. By now, for each gene, only the longest peptide is
considered for homology search. Only pairwise comparisons are taken into account. In principle, only Best Reciprocal
Hits (BRH) are considered. Nevertheless extra orthologous relationships are found taking into account the gene order
conservation of neighbour genes (RHS, Reciprocal Hit based on Synteny around BRH). Additionally, in Human vs.
Chimp relationships may be found Derived from Whole Genome Alignment (DWGA).</p>

<p>Another part is dedicated to protein clusters (or families). In this case, Metazoan entries of the SwissProt/SPTrEMBL
database are also taken into account in order to annotate the clusters.</p>

<p>A third part is planed for domain alignments but is not used by now.</p>
</LI>
</ul>

<hr>
<hr>
<h2>List of tables</h2>

<UL>
<LI><strong>General Tables</strong>
  <UL>
  <LI><A href="#meta">meta</A></LI>
  <LI><A href="#taxon">taxon</A></LI>
  <LI><A href="#genome_db">genome_db</A></LI>
  <LI><A href="#method_link">method_link</A></LI>
  <LI><A href="#method_link_species_set">method_link_species_set</A></LI>
  </UL>
</LI>
<LI><strong>Genomic Alignments Tables</strong>
  <UL>
  <LI><A href="#dnafrag">dnafrag</A></LI>
  <LI><A href="#genomic_align_block">genomic_align_block</A></LI>
  <LI><A href="#genomic_align">genomic_align</A></LI>
  <LI><A href="#genomic_align_group">genomic_align_group</A></LI>
  <LI><A href="#synteny_region">synteny_region</A></LI>
  <LI><A href="#dnafrag_region">dnafrag_region</A></LI>
  </UL>
</LI>
<LI><strong>Orthologues and Protein Clusters</strong>
  <UL>
  <LI><A href="#member">member</A></LI>
  <LI><A href="#sequence">sequence</A></LI>
  <LI><A href="#analysis">analysis</A></LI>
  <LI><A href="#analysis_description">analysis_description</A></LI>
  <LI><A href="#peptide_align_feature">peptide_align_feature</A></LI>
  <LI><A href="#homology">homology</A></LI>
  <LI><A href="#homology_member">homology_member</A></LI>
  <LI><A href="#family">family</A></LI>
  <LI><A href="#family_member">family_member</A></LI>
  <LI><A href="#domain">domain</A></LI>
  <LI><A href="#domain_member">domain_member</A></LI>
  </UL>
</LI>
</UL>

<!--
#####################################################################
#####################################################################
	GENERAL TABLES
#####################################################################
#####################################################################
 -->
<hr>
<hr>
<h2>General Tables</h2>

<!--
#####################################################################
	TABLE META
#####################################################################
 -->
<A name="meta"></A>
<h3>meta</h3>
<dl><dd>
<p>Contains configuration variables.</p>
<table border="2" cellpadding="2" bgcolor="#dcdcdc">
  <caption>meta table description</caption>
  <thead>
    <tr>
      <th>Field</th>
      <th>Type</th>
      <th>Null</th>
      <th>Key</th>
      <th>Default</th>
      <th>Extra</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>meta_id</td>
      <td>int unsigned</td>
      <td></td>
      <td>PRI</td>
      <td>NULL</td>
      <td></td>
      <td>internal unique ID</td>
    </tr>
    <tr>
      <td>meta_key</td>
      <td>varchar(40)</td>
      <td></td>
      <td>MUL</td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <td>meta_value</td>
      <td>varchar(255)</td>
      <td></td>
      <td>MUL</td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
  </tbody>
</table>
<p>Current values are:</p>
<PRE>
mysql> SELECT * FROM meta;
+---------+----------------------+------------+
| meta_id | meta_key             | meta_value |
+---------+----------------------+------------+
|       1 | max_alignment_length | 108838     |
+---------+----------------------+------------+</PRE>
<p>max_alignment_length is used in the perl API GenomicAlignAdaptor.pm to speed up the data query.</p>
</dd></dl>
<br>

<!--
#####################################################################
	TABLE TAXON
#####################################################################
 -->
<A name="taxon"></A>
<h3>taxon</h3>
<dl><dd>
<p>Contains all taxa used in this database.</p>
<table border="2" cellpadding="2" bgcolor="#dcdcdc">
  <caption>taxon table description</caption>
  <thead>
    <tr>
      <th>Field</th>
      <th>Type</th>
      <th>Null</th>
      <th>Key</th>
      <th>Default</th>
      <th>Extra</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>taxon_id</td>
      <td>int(10) unsigned</td>
      <td></td>
      <td>PRI</td>
      <td>0</td>
      <td></td>
      <td>unique ID</td>
    </tr>
    <tr>
      <td>genus</td>
      <td>varchar(50)</td>
      <td>YES</td>
      <td>MUL</td>
      <td>NULL</td>
      <td></td>
      <td>e.g. Homo</td>
    </tr>
    <tr>
      <td>species</td>
      <td>varchar(50)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>e.g. sapiens</td>
    </tr>
    <tr>
      <td>sub_species</td>
      <td>varchar(50)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <td>common_name</td>
      <td>varchar(100)</td>
      <td>YES</td>
      <td>MUL</td>
      <td>NULL</td>
      <td></td>
      <td>e.g. Human</td>
    </tr>
    <tr>
      <td>classification</td>
      <td>mediumtext</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>Full taxonomic classification</td>
    </tr>
  </tbody>
</table>
<p>E.g. the rows</p>
<PRE>
mysql> SELECT * FROM taxon WHERE taxon_id IN (9031, 9606);
+----------+--------+---------+-------------+-------------+-иии
| taxon_id | genus  | species | sub_species | common_name | 
+----------+--------+---------+-------------+-------------+-иии
|     9031 | Gallus | gallus  | NULL        | Chicken     |
|     9606 | Homo   | sapiens | NULL        | Human       | 
+----------+--------+---------+-------------+-------------+-иии

 иии-+-------------------------------------------------------------------иии
     | classification                                                           
 иии-+-------------------------------------------------------------------иии
     | gallus Gallus Phasianinae Phasianidae Galliformes Neognathae Aves
     | sapiens Homo Hominidae Catarrhini Primates Eutheria Mammalia 
 иии-+-------------------------------------------------------------------иии

 иии-------------------------------------------------------------------------+
                                                                             |
 иии-------------------------------------------------------------------------+
     Archosauria Euteleostomi Vertebrata Craniata Chordata Metazoa Eukaryota |
     Euteleostomi Vertebrata Craniata Chordata Metazoa Eukaryota             |
 иии-------------------------------------------------------------------------+
</PRE>
<P>correspond to the Chicken and the Human species.</P>
</dd></dl>
<br>

<!--
#####################################################################
	TABLE GENOME_DB
#####################################################################
 -->
<A name="genome_db"></A>
<h3>genome_db</h3>
<dl><dd>
<p>Contains information about the version of the genome assemblies used in this database.</P>
<table border="2" cellpadding="2" bgcolor="#dcdcdc">
  <caption>genome_db table description</caption>
  <thead>
    <tr>
      <th>Field</th>
      <th>Type</th>
      <th>Null</th>
      <th>Key</th>
      <th>Default</th>
      <th>Extra</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>genome_db_id</td>
      <td>int(10) unsigned</td>
      <td></td>
      <td>PRI</td>
      <td>NULL</td>
      <td>auto_increment</td>
      <td>internal unique ID</td>
    </tr>
    <tr>
      <td>taxon_id</td>
      <td>int(10) unsigned</td>
      <td></td>
      <td></td>
      <td>0</td>
      <td></td>
      <td>external reference to <A href="#taxon">taxon.taxon_id</A></td>
    </tr>
    <tr>
      <td>name</td>
      <td>varchar(40)</td>
      <td></td>
      <td>MUL</td>
      <td></td>
      <td></td>
      <td>name of the species</td>
    </tr>
    <tr>
      <td>assembly</td>
      <td>varchar(100)</td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td>assembly version of the genome</td>
    </tr>
    <tr>
      <td>assembly_default</td>
      <td>tinyint(1)</td>
      <td>YES</td>
      <td></td>
      <td>1</td>
      <td></td>
      <td>boolean value describing if this assembly is the default one or not, so that we can handle more than one
      assempbly version for a given species.</td>
    </tr>
    <tr>
      <td>genebuild</td>
      <td>varchar(100)</td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td>version of the genebuild</td>
    </tr>
    <tr>
      <td>locator</td>
      <td>varchar(255)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>used for production purposes or for user configuration in in-house installation.</td>
    </tr>
  </tbody>
</table>
<p>Eg the rows:</p>
<PRE>
mysql> SELECT * FROM genome_db WHERE genome_db_id IN (1, 11);
+--------------+----------+-------------------------+------------+-иии
| genome_db_id | taxon_id | name                    | assembly   |
+--------------+----------+-------------------------+------------+-иии
|            1 |     9606 | Homo sapiens            | NCBI34     |
|           11 |     9031 | Gallus gallus           | WASHUC1    |
+--------------+----------+-------------------------+------------+-иии

иии-+------------------+---------------+---------+
    | assembly_default | genebuild     | locator |
иии-+------------------+---------------+---------+
    |                1 | 0405Ensembl   | NULL    |
    |                1 | 0406Ensembl   | NULL    |
иии-+------------------+---------------+---------+
</PRE>
<p>correspond to the Human and Chicken genomes</p>
</dd></dl>
<br>


<!--
#####################################################################
	TABLE METHOD_LINK
#####################################################################
 -->
<A name="method_link"></A>
<h3>method_link</h3>
<dl><dd>
<p>Contains the list of alignment methods used to find links (homologies) between DNA sequences.</p>
<table border="2" cellpadding="2" bgcolor="#dcdcdc">
  <caption>method_link table description</caption>
  <thead>
    <tr>
      <th>Field</th>
      <th>Type</th>
      <th>Null</th>
      <th>Key</th>
      <th>Default</th>
      <th>Extra</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>method_link_id</td>
      <td>int(10) unsigned</td>
      <td></td>
      <td>PRI</td>
      <td>NULL</td>
      <td>auto_increment</td>
      <td>internal unique ID</td>
    </tr>
    <tr>
      <td>type</td>
      <td>varchar(50)</td>
      <td></td>
      <td>MUL</td>
      <td></td>
      <td></td>
      <td>the common name of the linking method between species.</td>
    </tr>
  </tbody>
</table>
<p>Current values are:</p>
<PRE>
mysql> SELECT * FROM method_link;
+----------------+----------------------+
| method_link_id | type                 |
+----------------+----------------------+
|              1 | BLASTZ_NET           |
|              2 | BLASTZ_NET_TIGHT     |
|              3 | BLASTZ_RECIP_NET     |
|              4 | PHUSION_BLASTN       |
|              5 | PHUSION_BLASTN_TIGHT |
|              6 | TRANSLATED_BLAT      |
|              7 | BLASTZ_GROUP         |
|              8 | BLASTZ_GROUP_TIGHT   |
|            101 | SYNTENY              |
|            201 | ENSEMBL_ORTHOLOGUES  |
|            202 | ENSEMBL_PARALOGUES   |
|            301 | FAMILY               |
+----------------+----------------------+
</PRE>
</dd></dl>
<br>


<!--
#####################################################################
	TABLE METHOD_LINK_SPECIES_SET
#####################################################################
 -->
<A name="method_link_species_set"></A>
<h3>method_link_species_set</h3>
<dl><dd>
<p>This table contains information about the comparisons stored in the database. A given
method_link_species_set_id exist for each comparison made and relates a
<a href="#method_link">method_link.method_link_id</a> with a set of
<A href="#genome_db">genome_db.genome_db_id</A>.</p>
<table border="2" cellpadding="2" bgcolor="#dcdcdc">
  <caption>method_link_species table description</caption>
  <thead>
    <tr>
      <th>Field</th>
      <th>Type</th>
      <th>Null</th>
      <th>Key</th>
      <th>Default</th>
      <th>Extra</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>method_link_species_set_id</td>
      <td>int(10) unsigned</td>
      <td></td>
      <td>MUL</td>
      <td>NULL</td>
      <td>auto_increment</td>
      <td>internal id.</td>
    </tr>
    <tr>
      <td>method_link_id</td>
      <td>int(10) unsigned</td>
      <td>YES</td>
      <td>MUL</td>
      <td>NULL</td>
      <td></td>
      <td>external reference to <A href="#method_link">method_link.method_link_id</A></td>
    </tr>
    <tr>
      <td>genome_db_id</td>
      <td>int(10) unsigned</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>external reference to <A href="#genome_db">genome_db.genome_db_id</A></td>
    </tr>
  </tbody>
</table>
<p>E.g. the rows</p>
<PRE>
mysql> SELECT * FROM method_link_species_set WHERE method_link_species_set_id = 71;
+----------------------------+----------------+--------------+
| method_link_species_set_id | method_link_id | genome_db_id |
+----------------------------+----------------+--------------+
|                         71 |              1 |            1 |
|                         71 |              1 |           11 |
+----------------------------+----------------+--------------+</PRE>
<p>mean that BLASTZ_NET (<A href="#method_link">method_link_id</A> = 1) has been used for linking all the species
of this set: Human (<A href="#genome_db">genome_db_id</A> = 1) and Chicken
(<A href="#genome_db">genome_db_id</A> = 11).</p>
</dd></dl>
<br>

<!--
#####################################################################
#####################################################################
	TABLES FOR GENOMIC ALIGNMENT
#####################################################################
#####################################################################
 -->
<hr>
<hr>
<h2>Tables For Genomic Alignments</h2>
<!--
#####################################################################
	TABLE DNAFRAG
#####################################################################
 -->
<A name="dnafrag"></A>
<h3>dnafrag</h3>
<dl><dd>
<p>This table defines the genomic sequences used in the comparative genomics analyisis. It is used by the
<A href="#genomic_align_block">genomic_align_block</A> table to define aligned sequences. It is also used by the
<A href="#dnafrag_region">dnafrag_region</A> table to define syntenic regions.</p>
<p></p>
<table border="2" cellpadding="2" bgcolor="#dcdcdc">
  <caption>dnafrag table description</caption>
  <thead>
    <tr>
      <th>Field</th>
      <th>Type</th>
      <th>Null</th>
      <th>Key</th>
      <th>Default</th>
      <th>Extra</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>dnafrag_id</td>
      <td>int(10) unsigned</td>
      <td></td>
      <td>PRI</td>
      <td>NULL</td>
      <td>auto_increment</td>
      <td>internal unique ID</td>
    </tr>
    <tr>
      <td>length</td>
      <td>int(11)</td>
      <td></td>
      <td></td>
      <td>0</td>
      <td></td>
    </tr>
    <tr>
      <td>name</td>
      <td>varchar(40)</td>
      <td></td>
      <td>MUL</td>
      <td></td>
      <td></td>
      <td>name of the DNA sequence (e.g., the name of the chromosome)</td>
    </tr>
    <tr>
      <td>genome_db_id</td>
      <td>int(10) unsigned</td>
      <td></td>
      <td></td>
      <td>0</td>
      <td></td>
      <td>external reference to <a href="#genome_db">genome_db.genome_db_id</A></td>
    </tr>
    <tr>
      <td>coord_system_name</td>
      <td>varchar(40)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>refers to the coord system in which this dnafrag has been defined</td>
    </tr>
  </tbody>
</table>
<p>E.g. the row</p>
<pre>
mysql> SELECT * FROM dnafrag WHERE dnafrag_id IN (19, 23227);
+------------+-----------+------+--------------+-------------------+
| dnafrag_id | length    | name | genome_db_id | coord_system_name |
+------------+-----------+------+--------------+-------------------+
|         19 | 105311216 | 14   |            1 | chromosome        |
|      23227 |  56310377 | 5    |           11 | chromosome        |
+------------+-----------+------+--------------+-------------------+
</pre>
<p>refer to the chromosome 14 of the Human genome (<a href="#genome_db">genome_db.genome_db_id</a> = 1 refers
to Human genome in this example) which is 105311216 nucleotides length and to the chromosome 5 of the
Chicken genome (<a href="#genome_db">genome_db.genome_db_id</a> = 11 refers to Chicken genome in this example)
which is 56310377 nucleotides length.
</p>

</dd></dl>
<br>


<!--
#####################################################################
	TABLE GENOMIC_ALIGN_BLOCK
#####################################################################
 -->
<A name="genomic_align_block"></A>
<h3>genomic_align_block</h3>
<dl><dd>
<p>This table is the key table for the genomic alignments. The software used to align the genomic blocks is
refered as an external key to the <A href="#method_link_species_set">method_link</A>
table. Nevertheless, actual aligned sequences are defined in the <a href="#genomic_align">genomic_align</a>
table.</p>
<p></p>
<table border="2" cellpadding="2" bgcolor="#dcdcdc">
  <caption>genomic_align_block table description</caption>
  <thead>
    <tr>
      <th>Field</th>
      <th>Type</th>
      <th>Null</th>
      <th>Key</th>
      <th>Default</th>
      <th>Extra</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>genomic_align_block_id</td>
      <td>bigint(20) unsigned</td>
      <td></td>
      <td>PRI</td>
      <td>NULL</td>
      <td>auto_increment</td>
      <td>internal unique ID</td>
    </tr>
    <tr>
      <td>method_link_species_set_id</td>
      <td>int(10) unsigned</td>
      <td></td>
      <td>MUL</td>
      <td>0</td>
      <td></td>
      <td>external reference to <a href="#method_link_species_set">method_link_species_set.method_link_species_set_id</A></td>
    </tr>
    <tr>
      <td>score</td>
      <td>double</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>score returned by the homology search program</td>
    </tr>
    <tr>
      <td>perc_id</td>
      <td>tinyint(3) unsigned</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>Used for pairwise comparison. Defines the percentage of identity between both sequences</td>
    </tr>
    <tr>
      <td>length</td>
      <td>int(10)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>total length of the alignment</td>
    </tr>
  </tbody>
</table>
<p>E.g. the row</p>
<pre>
mysql> SELECT * FROM genomic_align_block WHERE genomic_align_block_id = 5095912;
+------------------------+----------------------------+-------+---------+--------+
| genomic_align_block_id | method_link_species_set_id | score | perc_id | length |
+------------------------+----------------------------+-------+---------+--------+
|                5095912 |                         71 |  1102 |      60 |     53 |
+------------------------+----------------------------+-------+---------+--------+
</pre>
<p>will refer to a BLASTZ_NET alignment between human and chicken genomes
(<a href="#method_link_species_set">method_link_species_set.method_link_species_set_id</a> = 71)
with a score of 1102, an identity of 60% and a length of 53 nucleotides. The actual sequences
corresponding to this aligment are defined in the <a href="#genomic_align">genomic_align</a> table.
</p>

</dd></dl>
<br>

<!--
#####################################################################
	TABLE GENOMIC_ALIGN
#####################################################################
 -->
<A name="genomic_align"></a>
<h3>genomic_align</h3>
<dl><dd>
<p>This table contains the coordinates and all the information needed to rebuild genomic alignments. Every
entry corresponds to one of the aligned sequences. It also contains an external key to the
<A href="#method_link_species_set">method_link_species_set</A> which refers to the software and set of species
used for getting the corresponding alignment. The aligned sequence is defined by an external reference to the
<A href="#dnafrag">dnafrag</A> table, the starting and ending position within this dnafrag, the strand and a
cigar_line.</p>
<p>The original aligned sequence is not stored but it can be retrieved using the <strong>cigar_line</strong> field and the
original sequence. The cigar line defines the sequence of matches/mismatches and deletions (or gaps).
For example, this cigar line <strong>2MD3M2D2M</strong> will mean that the alignment contains 2
matches/mismatches, 1 deletion (number 1 is omitted in order to save some space), 3 matches/mismatches,
2 deletions and 2 matches/mismatches. If the original sequence is:</p>
<ul>
<LI><strong>Original sequence:</strong> AACGCTT</LI>
</ul>
<p>The aligned sequence will be:</p>
<dl><dd>
<table border="1" cellpadding="2" cellspacing="1" bgcolor="#dcffff">
  <caption>cigar line: 2MD3M2D2M</caption>
  <thead>
    <tr>
      <th>M</th>
      <th>M</th>
      <th>D</th>
      <th>M</th>
      <th>M</th>
      <th>M</th>
      <th>D</th>
      <th>D</th>
      <th>M</th>
      <th>M</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>A</td>
      <td>A</td>
      <td>-</td>
      <td>C</td>
      <td>G</td>
      <td>C</td>
      <td>-</td>
      <td>-</td>
      <td>T</td>
      <td>T</td>
    </tr>
  </tbody>
</table>
</dd></dl>
<p></p>
<table border="2" cellpadding="2" bgcolor="#dcdcdc">
  <caption>genomic_align_block table description</caption>
  <thead>
    <tr>
      <th>Field</th>
      <th>Type</th>
      <th>Null</th>
      <th>Key</th>
      <th>Default</th>
      <th>Extra</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>genomic_align_id</td>
      <td>bigint(20) unsigned</td>
      <td></td>
      <td>PRI</td>
      <td>NULL</td>
      <td>auto_increment</td>
      <td>unique internal id</td>
    </tr>
    <tr>
      <td>genomic_align_block_id</td>
      <td>bigint(10) unsigned</td>
      <td></td>
      <td>MUL</td>
      <td>0</td>
      <td></td>
      <td>external reference to <a href="#genomic_align_block">genomic_align_block.genomic_align_block_id</a></td>
    </tr>
    <tr>
      <td>method_link_species_set_id</td>
      <td>int(10) unsigned</td>
      <td></td>
      <td></td>
      <td>0</td>
      <td></td>
      <td>external reference to
      <a href="#method_link_species_set">method_link_species_set.method_link_species_set_id</a>. This
      information is redundant because it also appears in the
      <a href="#genomic_align_block">genomic_align_block</a> table but it is used to speed up the queries</td>
    </tr>
    <tr>
      <td>dnafrag_id</td>
      <td>int(10) unsigned</td>
      <td></td>
      <td>MUL</td>
      <td>0</td>
      <td></td>
      <td>external reference to <A href="#dnafrag">dnafrag.dnafrag_id</a></td>
    </tr>
    <tr>
      <td>dnafrag_start</td>
      <td>int(10)</td>
      <td></td>
      <td></td>
      <td>0</td>
      <td></td>
      <td>starting position within the dnafrag defined by dnafrag_id</td>
    </tr>
    <tr>
      <td>dnafrag_end</td>
      <td>int(10)</td>
      <td></td>
      <td></td>
      <td>0</td>
      <td></td>
      <td>ending position within the dnafrag defined by dnafrag_id</td>
    </tr>
    <tr>
      <td>dnafrag_strand</td>
      <td>tinyint(4)</td>
      <td></td>
      <td></td>
      <td>0</td>
      <td></td>
      <td>strand in the dnafrag defined by dnafrag_id</td>
    </tr>
    <tr>
      <td>cigar_line</td>
      <td>mediumtext</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>internal description of the aligned sequence</td>
    </tr>
    <tr>
      <td>level_id</td>
      <td>tinyint(2)</td>
      <td></td>
      <td></td>
      <td>0</td>
      <td></td>
      <td>level of orhologous layer. 1 corresponds to the first layer of orthologous sequences found, 2 and over
      are addiotional layers. Use for building the <A href="#synteny_region">syntenies</A> (based on level_id 
      = 1 only)</td>
    </tr>
  </tbody>
</table>
<p>E.g. the rows</p>
<PRE>
mysql> SELECT * FROM genomic_align WHERE genomic_align_block_id = 5095912;
+------------------+------------------------+----------------------------+------------+-иии
| genomic_align_id | genomic_align_block_id | method_link_species_set_id | dnafrag_id |
+------------------+------------------------+----------------------------+------------+-иии
|         10191817 |                5095912 |                         71 |      23227 |
|         10191830 |                5095912 |                         71 |         19 |
+------------------+------------------------+----------------------------+------------+-иии

 иии-+---------------+-------------+----------------+------------+----------+
     | dnafrag_start | dnafrag_end | dnafrag_strand | cigar_line | level_id |
 иии-+---------------+-------------+----------------+------------+----------+
     |      30021505 |    30021554 |              1 | 37M3D13M   |        2 |
     |      25833506 |    25833558 |             -1 | 53M        |        2 |
 иии-+---------------+-------------+----------------+------------+----------+
</PRE>
<p>correspond to both pieces of sequences included in the alignment described above (see
<a href="#genomic_align_block">genomic_align_block</a> table description). The first sequence includes the
nucleotides from 30021505 to 30021554 in the forwards strand of the chromosome 5 of the Chicken genome
(<a href="#dnafrag">dnafrag.dnafrag_id</a> = 23227). The second sequence includes the nucleotides from
25833506 to 25833558 in the backwards strand of the chromosome 14 of the Human genome
(<a href="#dnafrag">dnafrag.dnafrag_id</a> = 19).</p>
<p>The aligned sequences can be rebuild using the original sequences fetched from the corresponding core
databases and the cigar lines as explained before.</p>
</dd></dl>
<br>

<!--
#####################################################################
	TABLE GENOMIC_ALIGN_GROUP
#####################################################################
 -->
<A name="genomic_align_group"></A>
<h3>genomic_align_group</h3>
<dl><dd>
<p>This table is used to group alignments.</p>
<table border="2" cellpadding="2" bgcolor="#dcdcdc">
  <caption>genomic_align_group table description</caption>
  <thead>
    <tr>
      <th>Field</th>
      <th>Type</th>
      <th>Null</th>
      <th>Key</th>
      <th>Default</th>
      <th>Extra</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>group_id</td>
      <td>bigint(20) unsigned</td>
      <td></td>
      <td>MUL</td>
      <td>NULL</td>
      <td>auto_increment</td>
      <td>internal ID</td>
    </tr>
    <tr>
      <td>type</td>
      <td>varchar(40)</td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td>This field allow us to group genomic_aligns in several ways (types)</td>
    </tr>
    <tr>
      <td>genomic_align_id</td>
      <td>bigint(20) unsigned</td>
      <td></td>
      <td>MUL</td>
      <td>0</td>
      <td></td>
      <td>external reference to <a href="#genomic_align">genomic_align.genomic_align_id</a></td>
    </tr>
  </tbody>
</table>
<p>E.g. the rows</p>
<PRE>
mysql> SELECT * FROM genomic_align_group WHERE group_id = 64065;
+----------+---------+------------------+
| group_id | type    | genomic_align_id |
+----------+---------+------------------+
|    64065 | default |          1261387 |
|    64065 | default |          1261382 |
|    64065 | default |          1261429 |
|    64065 | default |          1261422 |
|    64065 | default |          8250806 |
|    64065 | default |          8250793 |
|    64065 | default |         10191710 |
|    64065 | default |         10191697 |
|    64065 | default |         10191750 |
|    64065 | default |         10191737 |
|    64065 | default |         10191789 |
|    64065 | default |         10191778 |
|    64065 | default |         10191830 |
|    64065 | default |         10191817 |
|    64065 | default |         10191868 |
|    64065 | default |         10191856 |
|    64065 | default |         10191909 |
|    64065 | default |         10191896 |
|    64065 | default |         11109668 |
|    64065 | default |         11109655 |
|    64065 | default |         14443871 |
|    64065 | default |         14443860 |
|    64065 | default |         19780097 |
|    64065 | default |         19780088 |
|    64065 | default |         22515872 |
|    64065 | default |         22515866 |
|    64065 | default |         23205949 |
|    64065 | default |         23205943 |
|    64065 | default |         31071852 |
|    64065 | default |         31071843 |
|    64065 | default |         54076401 |
|    64065 | default |         54076402 |
+----------+---------+------------------+
</PRE>
<p>correspond to the group of several genomic_align in which the alignment described before is included.</p>
</dd></dl>
<br>


<!--
#####################################################################
	TABLE SYNTENY_REGION
#####################################################################
 -->
<A name="synteny_region"></A>
<h3>synteny_region</h3>
<dl><dd>
<p>Contains all the syntenic relationships found and the relative orientation of both syntenic regions.</p>
<table border="2" cellpadding="2" bgcolor="#dcdcdc">
  <caption>synteny_region table description</caption>
  <thead>
    <tr>
      <th>Field</th>
      <th>Type</th>
      <th>Null</th>
      <th>Key</th>
      <th>Default</th>
      <th>Extra</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>synteny_region_id</td>
      <td>int(10) unsigned</td>
      <td></td>
      <td>PRI</td>
      <td>NULL</td>
      <td>auto_increment</td>
      <td>internal unique ID</td>
    </tr>
    <tr>
      <td>method_link_species_set_id</td>
      <td>int(10) unsigned</td>
      <td></td>
      <td></td>
      <td>0</td>
      <td></td>
      <td>external reference to
      <a href="#method_link_species_set">method_link_species_set.method_link_species_set_id</a>.</td>
    </tr>
    <tr>
      <td>rel_orientation</td>
      <td>tinyint(1)</td>
      <td></td>
      <td></td>
      <td>1</td>
      <td></td>
      <td>1 when both regions are in the same orientation, 0 otherwise.</td>
    </tr>
  </tbody>
</table>
<p>E.g. the row</p>
<PRE>mysql> SELECT * FROM synteny_region WHERE synteny_region_id = 1849;
+-------------------+-----------------+
| synteny_region_id | rel_orientation |
+-------------------+-----------------+
|              1849 |               1 |
+-------------------+-----------------+
</PRE>
<p>means that the syntenic region 1849 corresponds to a syntenic relationship where both genomic regions are in
the same orientation. See <A href="#dnafrag_region">dnafrag_region</A> table for more details.</p>
</dd></dl>
<br>

<!--
#####################################################################
	TABLE DNAFRAG_REGION
#####################################################################
 -->
<A name="dnafrag_region"></A>
<h3>dnafrag_region</h3>
<dl><dd>
<p>Contains the genomic regions corresponding to every syntenic relationship found. There are two genomic regions
for every syntenic relationship.</p>
<table border="2" cellpadding="2" bgcolor="#dcdcdc">
  <caption>dnafrag_region table description</caption>
  <thead>
    <tr>
      <th>Field</th>
      <th>Type</th>
      <th>Null</th>
      <th>Key</th>
      <th>Default</th>
      <th>Extra</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>synteny_region_id</td>
      <td>int(10) unsigned</td>
      <td></td>
      <td>PRI</td>
      <td>0</td>
      <td></td>
      <td>external reference to <A href="#synteny_region">synteny_region.synteny_region_id</A></td>
    </tr>
    <tr>
      <td>dnafrag_id</td>
      <td>int(10) unsigned</td>
      <td></td>
      <td>PRI</td>
      <td>0</td>
      <td></td>
      <td>external reference to <A href="#dnafrag">dnafrag.dnafrag_id</A></td>
    </tr>
    <tr>
      <td>dnafrag_start</td>
      <td>int(10) unsigned</td>
      <td></td>
      <td></td>
      <td>0</td>
      <td></td>
      <td>first nucleotide from this dnafrag which is in synteny</td>
    </tr>
    <tr>
      <td>dnafrag_end</td>
      <td>int(10) unsigned</td>
      <td></td>
      <td></td>
      <td>0</td>
      <td></td>
      <td>last nucleotide from this dnafrag which is in synteny</td>
    </tr>
  </tbody>
</table>
<p>E.g. the rows</p>
<PRE>
mysql> SELECT * FROM dnafrag_region WHERE synteny_region_id = 1849;
+-------------------+------------+---------------+--------------+
| synteny_region_id | dnafrag_id | dnafrag_start | dnafrag_end  |
+-------------------+------------+---------------+--------------+
|              1849 |         19 |      23939018 |     37861860 |
|              1849 |      23227 |      29431279 |     34526887 |
+-------------------+------------+---------------+--------------+
</PRE>
<p>correspond to both genomic regions of the syntenic region 1849. In this case, the first genomic region corresponds
to the sequence from 23939018 to 37861860 of the chromosome 14 of the Human genome
(<A href="#dnafrag">dnafrag_id</A> = 19 for this chromosome) and the second one corresponds to the sequence
from 29431279 to 34526887 of the chromosome 5 of the Chicken genome
(<A href="#dnafrag">dnafrag_id</A> = 23227 for this chromosome). Using the
<A href="#synteny_region">synteny_region</A> table, we know that both syntenic regions are in the same orientation.</p>
</dd></dl>
<br>


<!--
#####################################################################
#####################################################################
	TABLES FOR ORTHOLOGUES AND PROTEIN CLUSTERS
#####################################################################
#####################################################################
 -->
<hr>
<hr>
<h2>Tables For Orthologues and Protein Clusters</h2>


<!--
#####################################################################
	TABLE MEMBER
#####################################################################
 -->
<A name="member"></A>
<h3>member</h3>
<dl><dd>
<p>This table links sequences to the EnsEMBL core DB or to external DBs.</p>
<table border="2" cellpadding="2" bgcolor="#dcdcdc">
  <caption>member table description</caption>
  <thead>
    <tr>
      <th>Field</th>
      <th>Type</th>
      <th>Null</th>
      <th>Key</th>
      <th>Default</th>
      <th>Extra</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>member_id</td>
      <td>int(10) unsigned</td>
      <td></td>
      <td>PRI</td>
      <td>NULL</td>
      <td>auto_increment</td>
      <td>internal unique ID</td>
    </tr>
    <tr>
      <td>stable_id</td>
      <td>varchar(40)</td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td>EnsEMBL stable ID or external ID (for Uniprot/SWISSPROT and Uniprot/SPTREMBL)</td>
    </tr>
    <tr>
      <td>version</td>
      <td>int(10)</td>
      <td>YES</td>
      <td></td>
      <td>0</td>
      <td></td>
      <td>version of the stable ID (see EnsEMBL core DB)</td>
    </tr>
    <tr>
      <td>source_name</td>
      <td>varchar(40)</td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td>describe the source of the member (Uniprot/SWISSPROT, Uniprot/SPTREMBL, ENSEMBLGENE, ENSEMBLPEP)</td>
    </tr>
    <tr>
      <td>taxon_id</td>
      <td>int(10) unsigned</td>
      <td></td>
      <td></td>
      <td>0</td>
      <td></td>
      <td>external reference to <A href="#taxon">taxon.taxon_id</A></td>
    </tr>
    <tr>
      <td>genome_db_id</td>
      <td>int(10) unsigned</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>external reference to <A href="#genome_db">genome_db.genome_db_id</A></td>
    </tr>
    <tr>
      <td>sequence_id</td>
      <td>int(10) unsigned</td>
      <td></td>
      <td>MUL</td>
      <td>NULL</td>
      <td></td>
      <td>external reference to <A href="#sequence">sequence.sequence_id</A>.<br>
      May be 0 when the sequence is not available in the <A href="#sequence">sequence</A> table, e.g. for a gene
      instance.</td>
    </tr>
    <tr>
      <td>gene_member_id</td>
      <td>int(10) unsigned</td>
      <td></td>
      <td>MUL</td>
      <td>NULL</td>
      <td></td>
      <td>external reference to <A href="#member">member.memebr_id</A> to allow linkage from peptides to genes.</td>
    </tr>
    <tr>
      <td>description</td>
      <td>varchar(255)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>the description of the protein</td>
    </tr>
    <tr>
      <td>chr_name</td>
      <td>varchar(40)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>chromosome where this sequence is located</td>
    </tr>
    <tr>
      <td>chr_start</td>
      <td>int(10)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>first nucleotide of this chromosome which corresponds to this member</td>
    </tr>
    <tr>
      <td>chr_end</td>
      <td>int(10)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>last nucleotide of this chromosome which corresponds to this member</td>
    </tr>
    <tr>
      <td>chr_strand</td>
      <td>tinyint(1)</td>
      <td></td>
      <td></td>
      <td>0</td>
      <td></td>
      <td>strand of the chromosome in which the member is</td>
    </tr>
  </tbody>
</table>
<p>E.g. the row</p>
<PRE>
mysql> SELECT * FROM member WHERE member_id = 1945;
+-----------+--------------------+---------+-------------+----------+--------------+-------------+-иии
| member_id | stable_id          | version | source_name | taxon_id | genome_db_id | sequence_id | 
+-----------+--------------------+---------+-------------+----------+--------------+-------------+-иии
|      1945 | ENSGALP00000012979 |       1 | ENSEMBLPEP  |     9031 |           11 |        1121 | 
+-----------+--------------------+---------+-------------+----------+--------------+-------------+-иии

 иии-+----------------+-иии
     | gene_member_id | 
 иии-+----------------+-иии
     |           1932 | 
 иии-+----------------+-иии
     
 иии-+---------------------------------------------------------------------------------------+-иии
     | description                                                                           | 
 иии-+---------------------------------------------------------------------------------------+-иии
     | Transcript:ENSGALT00000012994 Gene:ENSGALG00000008004 Chr:1 Start:8903413 End:8906606 | 
 иии-+---------------------------------------------------------------------------------------+-иии
     
 иии-+----------+-----------+---------+------------+
     | chr_name | chr_start | chr_end | chr_strand |
 иии-+----------+-----------+---------+------------+
     | 1        |   8903413 | 8906606 |         -1 |
 иии-+----------+-----------+---------+------------+


</PRE>
<p>refers to the chicken (<A href="#taxon">taxon_id</A> = 9031 or <a href="#genome_db">genome_db_id</a> = 11)
gene ENSGALP00000012979 which is located in the chromosome 1 (from 8903413 to 8906606, in the reverse strand).
This gene is described as "Transcript:ENSGALT00000012994 Gene:ENSGALG00000008004 Chr:1 Start:8903413 End:8906606"
and the sequence can be found in the <A href="#sequence">sequence</A> table.</p>
</dd></dl>
<br>

<!--
#####################################################################
	TABLE SEQUENCE
#####################################################################
 -->
<A name="sequence"></A>
<h3>sequence</h3>
<dl><dd>
<p>This table contains the protein sequences present in the member table used in the protein alignment part of the
EnsEMBL Compara DB.</p>
<table border="2" cellpadding="2" bgcolor="#dcdcdc">
  <caption>sequence table description</caption>
  <thead>
    <tr>
      <th>Field</th>
      <th>Type</th>
      <th>Null</th>
      <th>Key</th>
      <th>Default</th>
      <th>Extra</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>sequence_id</td>
      <td>int(10) unsigned</td>
      <td></td>
      <td>PRI</td>
      <td>NULL</td>
      <td>auto_increment</td>
      <td>internal unique ID</td>
    </tr>
    <tr>
      <td>sequence</td>
      <td>longtext</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>the sequence</td>
    </tr>
    <tr>
      <td>length</td>
      <td>int(10)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>the length of the sequence</td>
    </tr>
  </tbody>
</table>
<p>E.g. the row</p>
<PRE>
mysql> select * from sequence where sequence_id = 1121;
+-------------+-------------------------------------------+--------+
| sequence_id | sequence                                  | length |
+-------------+-------------------------------------------+--------+
|        1121 | LPNTRGYTQVWECSLAVLIAMVCMTLVGWGLIWLFSVTASV |     41 |
+-------------+-------------------------------------------+--------+
</PRE>
<p>contains a 41 aminoacids long sequence.</p>
</dd></dl>
<br>

<!--
#####################################################################
	TABLE ANALYSIS
#####################################################################
 -->
<A name="analysis"></A>
<h3>analysis</h3>
<dl><dd>
<p>This table is mainly used for production purposes.</p>
<table border="2" cellpadding="2" bgcolor="#dcdcdc">
  <caption>analysis table description</caption>
  <thead>
    <tr>
      <th>Field</th>
      <th>Type</th>
      <th>Null</th>
      <th>Key</th>
      <th>Default</th>
      <th>Extra</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>analysis_id</td>
      <td>int(10) unsigned</td>
      <td></td>
      <td>PRI</td>
      <td>NULL</td>
      <td>auto_increment</td>
      <td>internal unique ID</td>
    </tr>
    <tr>
      <td>created</td>
      <td>datetime</td>
      <td></td>
      <td></td>
      <td>0000-00-00 00:00:00</td>
      <td></td>
      <td>date to distinguish newer and older versions off the same analysis. Not well maintained so far.</td>
    </tr>
    <tr>
      <td>logic_name</td>
      <td>varchar(40)</td>
      <td></td>
      <td>UNI</td>
      <td></td>
      <td></td>
      <td>string to identify the analysis. Used mainly inside pipeline.</td>
    </tr>
    <tr>
      <td>db</td>
      <td>varchar(120)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td rowspan="2">db should be a database name, db version the version of that db</td>
    </tr>
    <tr>
      <td>db_version</td>
      <td>varchar(40)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
    </tr>
    <tr>
      <td>db_file</td>
      <td>varchar(120)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>the file system location of that database, probably wiser to generate from just db and configurations</td>
    </tr>
    <tr>
      <td>program</td>
      <td>varchar(80)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td rowspan="3">The binary used to create a feature. Similar semantic to above</td>
    </tr>
    <tr>
      <td>program_version</td>
      <td>varchar(40)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
    </tr>
    <tr>
      <td>program_file</td>
      <td>varchar(80)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
    </tr>
    <tr>
      <td>parameters</td>
      <td>varchar(255)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>a parameter string which is processed by the perl module</td>
    </tr>
    <tr>
      <td>module</td>
      <td>varchar(80)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td rowspan="2">Perl module names (RunnableDBS usually) executing this analysis</td>
    </tr>
    <tr>
      <td>module_version</td>
      <td>varchar(40)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
    </tr>
    <tr>
      <td>gff_source</td>
      <td>varchar(40)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td rowspan="2">how to make a gff dump from features with this analysis</td>
    </tr>
    <tr>
      <td>gff_feature</td>
      <td>varchar(40)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
    </tr>
  </tbody>
</table>
</dd></dl>
<br>

<!--
#####################################################################
	TABLE ANALYSIS_DESCRIPTION
#####################################################################
 -->
<A name="analysis_description"></A>
<h3>analysis_description</h3>
<dl><dd>
<p>This table has been added to comply with core Bio::EnsEMBL::DBSQL::AnalysisAdaptor requirements.</p>
<table border="2" cellpadding="2" bgcolor="#dcdcdc">
  <caption>analysis_description table description</caption>
  <thead>
    <tr>
      <th>Field</th>
      <th>Type</th>
      <th>Null</th>
      <th>Key</th>
      <th>Default</th>
      <th>Extra</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>analysis_id</td>
      <td>int(10) unsigned</td>
      <td></td>
      <td>MUL</td>
      <td>0</td>
      <td></td>
      <td>external reference to <a href="#analysis">analysis.analysis_id</a></td>
    </tr>
    <tr>
      <td>description</td>
      <td>text</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <td>display_label</td>
      <td>varchar(255)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td></td>
    </tr>
  </tbody>
</table>
</dd></dl>
<br>

<!--
#####################################################################
	TABLE PEPTIDE_ALIGN_FEATURE
#####################################################################
 -->
<A name="peptide_align_feature"></A>
<h3>
      peptide_align_feature
    </h3>
<dl><dd>
<p>This tables stores the raw HSP local alignment results of peptide to peptide alignments returned by a
BLAST run it is translated from a FeaturePair object.</p>
<table border="2" cellpadding="2" bgcolor="#dcdcdc">
  <caption>peptide_align_feature table description</caption>
  <thead>
    <tr>
      <th>Field</th>
      <th>Type</th>
      <th>Null</th>
      <th>Key</th>
      <th>Default</th>
      <th>Extra</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>peptide_align_feature_id</td>
      <td>int(10) unsigned</td>
      <td></td>
      <td>PRI</td>
      <td>NULL</td>
      <td>auto_increment</td>
      <td>internal unique ID</td>
    </tr>
    <tr>
      <td>qmember_id</td>
      <td>int(10) unsigned</td>
      <td></td>
      <td>MUL</td>
      <td>0</td>
      <td></td>
      <td>external reference to <A HREF="#member">member.member_id</A> for the query peptide</td>
    </tr>
    <tr>
      <td>hmember_id</td>
      <td>int(10) unsigned</td>
      <td></td>
      <td>MUL</td>
      <td>0</td>
      <td></td>
      <td>external reference to <A HREF="#member">member.member_id</A> for the hit peptide</td>
    </tr>
    <tr>
      <td>qgenome_db_id</td>
      <td>int(10) unsigned</td>
      <td></td>
      <td></td>
      <td>0</td>
      <td></td>
      <td>external reference to <A href="#genome_db">genome_db.genome_db_id</A> for the query peptide (for query optimization)</td>
    </tr>
    <tr>
      <td>hgenome_db_id</td>
      <td>int(10) unsigned</td>
      <td></td>
      <td></td>
      <td>0</td>
      <td></td>
      <td>external reference to <A href="#genome_db">genome_db.genome_db_id</A> for the hit peptide (for query optimization)</td>
    </tr>
    <tr>
      <td>analysis_id</td>
      <td>int(10) unsigned</td>
      <td></td>
      <td></td>
      <td>0</td>
      <td></td>
      <td>external reference to <A href="#analysis">analysis.analyis_id</A></td>
    </tr>
    <tr>
      <td>qstart</td>
      <td>int(10)</td>
      <td></td>
      <td></td>
      <td>0</td>
      <td></td>
      <td>starting position in the query peptide sequence</td>
    </tr>
    <tr>
      <td>qend</td>
      <td>int(10)</td>
      <td></td>
      <td></td>
      <td>0</td>
      <td></td>
      <td>ending  position in the query peptide sequence</td>
    </tr>
    <tr>
      <td>hstart</td>
      <td>int(10)</td>
      <td></td>
      <td></td>
      <td>0</td>
      <td></td>
      <td>starting position in the hit peptide sequence</td>
    </tr>
    <tr>
      <td>hend</td>
      <td>int(10)</td>
      <td></td>
      <td></td>
      <td>0</td>
      <td></td>
      <td>ending  position in the hit peptide sequence</td>
    </tr>
    <tr>
      <td>score</td>
      <td>double(16,4)</td>
      <td></td>
      <td></td>
      <td>0.0000</td>
      <td></td>
      <td>blast score for this HSP</td>
    </tr>
    <tr>
      <td>evalue</td>
      <td>varchar(20)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>blast evalue for this HSP</td>
    </tr>
    <tr>
      <td>align_length</td>
      <td>int(10)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>alignment length of HSP</td>
    </tr>
    <tr>
      <td>identical_matches</td>
      <td>int(10)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>blast HSP match score</td>
    </tr>
    <tr>
      <td>perc_ident</td>
      <td>int(10)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>percent identical matches in the HSP length</td>
    </tr>
    <tr>
      <td>positive_matches</td>
      <td>int(10)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>blast HSP positive score</td>
    </tr>
    <tr>
      <td>perc_pos</td>
      <td>int(10)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>precent positive matches in the HSP length</td>
    </tr>
    <tr>
      <td>hit_rank</td>
      <td>int(10)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>rank in blast result</td>
    </tr>
    <tr>
      <td>cigar_line</td>
      <td>mediumtext</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>cigar string coding the actual alignment</td>
    </tr>
  </tbody>
</table>
<p>E.g. the rows:</p>
<PRE>
mysql> SELECT * FROM peptide_align_feature WHERE qmember_id = 442105;
+--------------------------+------------+------------+---------------+---------------+-иии
| peptide_align_feature_id | qmember_id | hmember_id | qgenome_db_id | hgenome_db_id |
+--------------------------+------------+------------+---------------+---------------+-иии
|                        1 |     442105 |     248885 |             3 |            12 |
|                        2 |     442105 |      86297 |             3 |            12 |
|                        3 |     442105 |     215369 |             3 |            12 |
|                        4 |     442105 |      67917 |             3 |            12 |
|                        5 |     442105 |     182642 |             3 |            12 |
|                        6 |     442105 |     212528 |             3 |            12 |
|                        7 |     442105 |     215342 |             3 |            12 |
|                        8 |     442105 |     260556 |             3 |             5 |
+--------------------------+------------+------------+---------------+---------------+-иии

 иии-+-------------+--------+------+--------+------+----------+---------+--------------+-иии
     | analysis_id | qstart | qend | hstart | hend | score    | evalue  | align_length |
 иии-+-------------+--------+------+--------+------+----------+---------+--------------+-иии
     |          14 |      3 |  240 |     38 |  276 | 276.0000 | 1.1e-25 |          243 |
     |          14 |     32 |  244 |      8 |  220 | 224.0000 | 3.5e-20 |          219 |
     |          14 |     30 |  213 |      8 |  198 | 212.0000 | 6.6e-19 |          194 |
     |          14 |     32 |  244 |      2 |  219 | 176.0000 | 2.2e-14 |          224 |
     |          14 |     36 |  227 |      1 |  201 | 161.0000 | 5.2e-13 |          205 |
     |          14 |     30 |  219 |      2 |  208 | 158.0000 | 8.5e-12 |          214 |
     |          14 |     29 |  214 |      7 |  201 | 158.0000 | 9.8e-12 |          201 |
     |          15 |     32 |  242 |      8 |  224 | 249.0000 | 1.2e-22 |          220 |
 иии-+-------------+--------+------+--------+------+----------+---------+--------------+-иии

 иии-+-------------------+------------+------------------+----------+----------+-иии
     | identical_matches | perc_ident | positive_matches | perc_pos | hit_rank |
 иии-+-------------------+------------+------------------+----------+----------+-иии
     |                80 |         32 |              124 |       51 |        1 |
     |                66 |         30 |              112 |       51 |        2 |
     |                63 |         32 |              104 |       53 |        3 |
     |                60 |         26 |              107 |       47 |        4 |
     |                61 |         29 |               99 |       48 |        5 |
     |                61 |         28 |               98 |       45 |        6 |
     |                57 |         28 |               97 |       48 |        7 |
     |                73 |         33 |              109 |       49 |        1 |
 иии-+-------------------+------------+------------------+----------+----------+-иии

 иии-+---------------------------------------------+
     | cigar_line                                  |
 иии-+---------------------------------------------+
     | 23M2I34MD42MI13MI51MD43M3D28M               |
     | 31M2D9M3I27M2D15MI56MD39MI7MD20MI3M         |
     | 27M2D7MD23M2D20M2I16MI33M2D9MD8M2D38M       |
     | 40M3D8M3D3M2D24M2I8MI33MD21MD43MD16M3I11M   |
     | 33MD5M2D8M5D20MI24MI29M4D17MD42M2I10M       |
     | 25M3D17M2D21M6I57M13D18MD3M4DMI36MD5M       |
     | 30MD7M3D9MI31M2I5M2D17M3I22M3D13M4D29M2D17M |
     | 33M2D7M2D20M4D9MI16MI55MD65MI3M             |
 иии-+---------------------------------------------+
</PRE>
<p>corresponds to all the hits found for the rat peptide defined by the <a href="#member">member.member_id</a>
442105.</p>
</dd></dl>
<br>

<!--
#####################################################################
	TABLE PROTEIN_TREE_NODE
#####################################################################
 -->
<A name="protein_tree_node"></A>
<h3>
      protein_tree_node
    </h3>
<dl><dd>
<p>This tables stores the phylogenetic structure of the genetrees.</p>
<table border="2" cellpadding="2" bgcolor="#dcdcdc">
  <caption>protein_tree_node description</caption>
  <thead>
    <tr>
      <th>Field</th>
      <th>Type</th>
      <th>Null</th>
      <th>Key</th>
      <th>Default</th>
      <th>Extra</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>node_id</td>
      <td>int(10) unsigned</td>
      <td>NO</td>
      <td>PRI</td>
      <td>NULL</td>
      <td>auto_increment</td>
      <td>node unique ID</td>
    </tr>
    <tr>
      <td>parent_id</td>
      <td>int(10) unsigned</td>
      <td>NO</td>
      <td>MUL</td>
      <td></td>
      <td></td>
      <td>parent node id</td>
    </tr>
    <tr>
      <td>root_id</td>
      <td>int(10) unsigned</td>
      <td>NO</td>
      <td>MUL</td>
      <td></td>
      <td></td>
      <td>cluster node id (usually 1) where this genetree is attached</td>
    </tr>
    <tr>
      <td>left_index</td>
      <td>int(10)</td>
      <td>NO</td>
      <td>MUL</td>
      <td></td>
      <td></td>
      <td>left index of the tree-traversal binary search</td>
    </tr>
    <tr>
      <td>right_index</td>
      <td>int(10)</td>
      <td>NO</td>
      <td>MUL</td>
      <td></td>
      <td></td>
      <td>right index of the tree-traversal binary search</td>
    </tr>
    <tr>
      <td>distance_to_parent</td>
      <td>double</td>
      <td>NO</td>
      <td></td>
      <td>1</td>
      <td></td>
      <td>branch length between this node and parent node</td>
    </tr>
  </tbody>
</table>
<p>E.g. the rows:</p>
<PRE>
mysql> select * from protein_tree_node where node_id=10;
+---------+-----------+---------+------------+-------------+--------------------+
| node_id | parent_id | root_id | left_index | right_index | distance_to_parent |
+---------+-----------+---------+------------+-------------+--------------------+
|      10 |    707637 |       1 |       7645 |        7646 |           0.001496 |
+---------+-----------+---------+------------+-------------+--------------------+

</PRE>
<p>corresponds to the node_id=10 with a branch length of 0.001496. This the right_index-left_index=1 value indicates that is an external node (leaf)</p>
</dd></dl>
<br>

<!--
#####################################################################
	TABLE PROTEIN_TREE_MEMBER
#####################################################################
 -->
<A name="protein_tree_member"></A>
<h3>
      protein_tree_member
    </h3>
<dl><dd>
<p>This tables stores the relationship between the <A href="#protein_tree_node">protein_tree_node</A> and the <A href="#member">member</A> tables.</p>
<table border="2" cellpadding="2" bgcolor="#dcdcdc">
  <caption>protein_tree_member description</caption>
  <thead>
    <tr>
      <th>Field</th>
      <th>Type</th>
      <th>Null</th>
      <th>Key</th>
      <th>Default</th>
      <th>Extra</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>node_id</td>
      <td>int(10) unsigned</td>
      <td>NO</td>
      <td>PRI</td>
      <td></td>
      <td></td>
      <td>external reference to <A href="#protein_tree_node">protein_tree_node.node_id</A></td>
    </tr>
    <tr>
      <td>member_id</td>
      <td>int(10) unsigned</td>
      <td>NO</td>
      <td>MUL</td>
      <td></td>
      <td></td>
      <td>external reference to <A href="#member">member.member_id</A></td>
    </tr>
    <tr>
      <td>method_link_species_set_id</td>
      <td>int(10) unsigned</td>
      <td>NO</td>
      <td></td>
      <td></td>
      <td></td>
      <td>method_link_species_set_id</td>
    </tr> 
    <tr>
      <td>cigar_line</td>
      <td>mediumtext</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>cigar string coding the alignment</td>
    </tr>
    <tr>
      <td>cigar_start</td>
      <td>int(10)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>defines the first alignment aminoacid</td>
    </tr>
    <tr>
      <td>cigar_end</td>
      <td>int(10)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>defines the last alignment aminoacid</td>
    </tr>
  </tbody>
</table>
<p>E.g. the rows:</p>
<PRE>
mysql> select * from protein_tree_member where node_id=10;
+---------+-----------+----------------------------+------------------------------+-------------+-----------+
| node_id | member_id | method_link_species_set_id | cigar_line                   | cigar_start | cigar_end |
+---------+-----------+----------------------------+------------------------------+-------------+-----------+
|      10 |    416753 |                      40046 | 106D53MD9M4D4MD42MD47MD47M8D |        NULL |      NULL |
+---------+-----------+----------------------------+------------------------------+-------------+-----------+

</PRE>
<p>corresponds to the node_id=10.</p>
<p><STRONG>N.B.</STRONG>The cigar_start and cigar_end fields will be NULL in multiple sequence alignments (only used in pairwise alignments).</p>
</dd></dl>
<br>

<!--
#####################################################################
	TABLE PROTEIN_TREE_TAG
#####################################################################
 -->
<A name="protein_tree_tag"></A>
<h3>
      protein_tree_tag
    </h3>
<dl><dd>
<p>This tables stores various tags for nodes in <A href="#protein_tree_node">protein_tree_node</A>.</p>
<table border="2" cellpadding="2" bgcolor="#dcdcdc">
  <caption>protein_tree_tag description</caption>
  <thead>
    <tr>
      <th>Field</th>
      <th>Type</th>
      <th>Null</th>
      <th>Key</th>
      <th>Default</th>
      <th>Extra</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>node_id</td>
      <td>int(10) unsigned</td>
      <td>NO</td>
      <td>MUL</td>
      <td></td>
      <td></td>
      <td>node unique ID</td>
    </tr>
    <tr>
      <td>tag</td>
      <td>varchar(50)</td>
      <td>YES</td>
      <td>MUL</td>
      <td>NULL</td>
      <td></td>
      <td>tag</td>
    </tr>
    <tr>
      <td>value</td>
      <td>mediumtext</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>value</td>
    </tr>
  </tbody>
</table>
<p>E.g. the rows:</p>
<PRE>
mysql> select * from protein_tree_tag where node_id=707631;
+---------+----------------------------+------------------+
| node_id | tag                        | value            |
+---------+----------------------------+------------------+
|  707631 | Bootstrap                  | 70               |
|  707631 | Duplication                | 0                |
|  707631 | Sitewise_dNdS_runtime_msec | 58504.8859863281 |
|  707631 | Sitewise_dNdS_subroot_id   | 8                |
|  707631 | taxon_alias                | Primates         |
|  707631 | taxon_id                   | 9443             |
|  707631 | taxon_name                 | Primates         |
+---------+----------------------------+------------------+

</PRE>
<p>corresponds to the node_id=707631 and shows all the tags associated with this internal node.</p>
</dd></dl>
<br>

<!--
#####################################################################
	TABLE NCBI_TAXA_NODE
#####################################################################
 -->
<A name="ncbi_taxa_node"></A>
<h3>
      ncbi_taxa_node
    </h3>
<dl><dd>
<p>This tables stores the phylogenetic structure of the NCBI taxonomy DB.</p>
<table border="2" cellpadding="2" bgcolor="#dcdcdc">
  <caption>ncbi_taxa_node description</caption>
  <thead>
    <tr>
      <th>Field</th>
      <th>Type</th>
      <th>Null</th>
      <th>Key</th>
      <th>Default</th>
      <th>Extra</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>taxon_id</td>
      <td>int(10) unsigned</td>
      <td>NO</td>
      <td>PRI</td>
      <td></td>
      <td></td>
      <td>taxon ID</td>
    </tr>
    <tr>
      <td>parent_id</td>
      <td>int(10) unsigned</td>
      <td>NO</td>
      <td>MUL</td>
      <td></td>
      <td></td>
      <td>parent taxon id</td>
    </tr>
    <tr>
      <td>rank</td>
      <td>char(32)</td>
      <td>NO</td>
      <td>MUL</td>
      <td></td>
      <td></td>
      <td>rank in the tree of life (family, genus, species, ...)</td>
    </tr>
    <tr>
      <td>genbank_hidden_flag</td>
      <td>tynyint(1)</td>
      <td>NO</td>
      <td></td>
      <td>0</td>
      <td></td>
      <td>genbank flag</td>
    </tr>
    <tr>
      <td>left_index</td>
      <td>int(10)</td>
      <td>NO</td>
      <td>MUL</td>
      <td></td>
      <td></td>
      <td>left index of the tree-traversal binary search</td>
    </tr>
    <tr>
      <td>right_index</td>
      <td>int(10)</td>
      <td>NO</td>
      <td>MUL</td>
      <td></td>
      <td></td>
      <td>right index of the tree-traversal binary search</td>
    </tr>
    <tr>
      <td>root_id</td>
      <td>int(10)</td>
      <td>NO</td>
      <td></td>
      <td>1</td>
      <td></td>
      <td>cluster node id (usually 1) where this taxonomy node is attached</td>
    </tr>
  </tbody>
</table>
<p>E.g. the rows:</p>
<PRE>
mysql> select * from ncbi_taxa_node limit 10;
+----------+-----------+--------------+---------------------+------------+-------------+---------+
| taxon_id | parent_id | rank         | genbank_hidden_flag | left_index | right_index | root_id |
+----------+-----------+--------------+---------------------+------------+-------------+---------+
|        1 |         0 | no rank      |                   0 |          1 |      891068 |       1 |
|        2 |    131567 | superkingdom |                   0 |       7665 |      234070 |       1 |
|        6 |    335928 | genus        |                   0 |      97634 |       97669 |       1 |
|        7 |         6 | species      |                   1 |      97635 |       97638 |       1 |
|        9 |     32199 | species      |                   1 |     123547 |      123678 |       1 |
|       10 |    135621 | genus        |                   0 |     129052 |      129111 |       1 |
|       11 |        10 | species      |                   1 |     129075 |      129076 |       1 |
|       13 |    203488 | genus        |                   0 |      50174 |       50197 |       1 |
|       14 |        13 | species      |                   1 |      50179 |       50182 |       1 |
|       16 |     32011 | genus        |                   0 |      70484 |       70531 |       1 |
+----------+-----------+--------------+---------------------+------------+-------------+---------+

</PRE>
<p>correspond to some taxon_ids in the NCBI taxonomy DB</p>
</dd></dl>
<br>

<!--
#####################################################################
	TABLE NCBI_TAXA_NAME
#####################################################################
 -->
<A name="ncbi_taxa_name"></A>
<h3>
      ncbi_taxa_name
    </h3>
<dl><dd>
<p>This tables stores the names of the NCBI taxonomy entries.</p>
<table border="2" cellpadding="2" bgcolor="#dcdcdc">
  <caption>ncbi_taxa_name description</caption>
  <thead>
    <tr>
      <th>Field</th>
      <th>Type</th>
      <th>Null</th>
      <th>Key</th>
      <th>Default</th>
      <th>Extra</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>taxon_id</td>
      <td>int(10) unsigned</td>
      <td>NO</td>
      <td>PRI</td>
      <td></td>
      <td></td>
      <td>external reference to <A href="#ncbi_taxa_node">ncbi_taxa_node.taxon_id</A></td>
    </tr>
    <tr>
      <td>name</td>
      <td>varchar(255)</td>
      <td>YES</td>
      <td>MUL</td>
      <td>NULL</td>
      <td></td>
      <td>Name of the taxon</td>
    </tr>
    <tr>
      <td>name_class</td>
      <td>varchar(50)</td>
      <td>YES</td>
      <td>MUL</td>
      <td>NULL</td>
      <td></td>
      <td>class of name</td>
    </tr>
  </tbody>
</table>
<p>E.g. the rows:</p>
<PRE>
mysql> select * from ncbi_taxa_name where taxon_id=9606;
+----------+--------------+---------------------+
| taxon_id | name         | name_class          |
+----------+--------------+---------------------+
|     9606 | Homo sapiens | scientific name     |
|     9606 | human        | genbank common name |
|     9606 | man          | common name         |
|     9606 | Human        | ensembl alias name  |
|     9606 | human        | ensembl common name |
+----------+--------------+---------------------+

</PRE>
<p>correspond to the human taxon_id names in the NCBI taxonomy DB</p>
</dd></dl>
<br>


<!--
#####################################################################
	TABLE HOMOLOGY
#####################################################################
 -->
<A name="homology"></A>
<h3>homology</h3>
<dl><dd>
<p>Contains all the genomic homologies found. There are two <A href="#homology_member">homology_member</A> entries
for each homology entry for now, but both the schema and the API can handle more than just pairwise relations.</p>
<table border="2" cellpadding="2" bgcolor="#dcdcdc">
  <caption>homology table description</caption>
  <thead>
    <tr>
      <th>Field</th>
      <th>Type</th>
      <th>Null</th>
      <th>Key</th>
      <th>Default</th>
      <th>Extra</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>homology_id</td>
      <td>int(10) unsigned</td>
      <td></td>
      <td>PRI</td>
      <td>NULL</td>
      <td>auto_increment</td>
      <td>internal unique ID</td>
    </tr>
    <tr>
      <td>stable_id</td>
      <td>varchar(40)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>stable ID of the pairwise homology relationship</td>
    </tr>
    <tr>
      <td>method_link_species_set_id</td>
      <td>int(10) unsigned</td>
      <td></td>
      <td>MUL</td>
      <td></td>
      <td></td>
      <td>external reference to <A href="#method_link_species_set">method_link_species_set.method_link_species_set_id</A></td>
    </tr>
    <tr>
      <td>description</td>
      <td>varchar(40)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>describes the type of homology found:
      <ul><LI><strong>ortholog_one2one</strong> for one to one orthologs</LI>
      <LI><strong>apparent_ortholog_one2one</strong> for apparent one to one orthologs</LI>
      <LI><strong>ortholog_one2many</strong> for one to many orthologs</LI> 
      <LI><strong>ortholog_many2many</strong> for many to many orthologs</LI> 
      <LI><strong>between_species_paralog</strong> for between species paralogs</LI> 
      </LI>
      </ul></td>
    </tr>
    <tr>
      <td>subtype</td>
      <td>varchar(40)</td>
      <td>NO</td>
      <td></td>
      <td></td>
      <td></td>
      <td>The subtype defines the taxonomic level of the relationship found (taxon_name).
      </ul></td>
    </tr>
    <tr>
      <td>dn</td>
      <td>float(10,5)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>number of nonsynonymous substitutions per nonsynonymous site</td>
    </tr>
    <tr>
      <td>ds</td>
      <td>float(10,5)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>number of synonymous substitutions per synonymous site</td>
    </tr>
    <tr>
      <td>n</td>
      <td>float(10,1)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>number of nonsynonymous substitutions</td>
    </tr>
    <tr>
      <td>s</td>
      <td>float(10,1)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>number of synonymous substitutions</td>
    </tr>
    <tr>
      <td>lnl</td>
      <td>float(10,3)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>maximum likelihood test value</td>
    </tr>
    <tr>
      <td>threshold_on_ds</td>
      <td>float(10,5)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>used by the EnsEMBL Web Browser to decide whether or not to display dN/DS ratio</td>
    </tr>
    <tr>
      <td>ancestor_node_id</td>
      <td>int(10) unsigned</td>
      <td>NO</td>
      <td></td>
      <td></td>
      <td></td>
      <td>ancestor node_id of the relationship</td>
    </tr>
    <tr>
      <td>tree_node_id</td>
      <td>int(10) unsigned</td>
      <td>NO</td>
      <td>MUL</td>
      <td></td>
      <td></td>
      <td>root node_id for the genetree where this relationship lies</td>
    </tr>
  </tbody>
</table>
<p><strong>dN</strong>, <strong>dS</strong>, <strong>N</strong>, <strong>S</strong> and <strong>lnL</strong> are
statistical values given by the <strong>codeml</strong> program of the
<A href="http://abacus.gene.ucl.ac.uk/software/paml.html">Phylogenetic Analysis by Maximum Likelihood (PAML)</A>
package.</p>
<p>E.g. the row</p>
<PRE>
mysql> select * from homology where homology_id=79;
+-------------+-----------+----------------------------+------------------+-...
| homology_id | stable_id | method_link_species_set_id | description      | 
+-------------+-----------+----------------------------+------------------+-...
|          79 | NULL      |                      22467 | ortholog_one2one | 
+-------------+-----------+----------------------------+------------------+-...

...+------------+---------+---------+------+------+----------+-----------------+------------------+--------------+
   | subtype    | dn      | ds      | n    | s    | lnl      | threshold_on_ds | ancestor_node_id | tree_node_id |
...+------------+---------+---------+------+------+----------+-----------------+------------------+--------------+
   | Catarrhini | 0.14610 | 0.21260 | 82.1 | 37.9 | -222.752 |         0.17120 |              284 |          284 |
...+------------+---------+---------+------+------+----------+-----------------+------------------+--------------+

</PRE>
<p>defines a one to one orthology at the taxonomic level Catarrhini (Apes and old world monkeys).</p>
<p><STRONG>N.B.</STRONG>At the moment there are no stable ids for homologies.</p>
</dd></dl>
<br>

<!--
#####################################################################
	TABLE HOMOLOGY_MEMBER
#####################################################################
 -->
<A name="homology_member"></A>
<h3>homology_member</h3>
<dl><dd>
<p>Contains the sequences corresponding to every genomic homology relationship found.
There are two homology_member entries for each pairwise <A href="#homology">homology</A> entry. As written in the
homology table section, both schema and API can deal with more than pairwise relationships.</p>
<p>The original alignment is not stored but it can be retrieved using the <strong>cigar_line</strong> field and the
original sequences. The cigar line defines the sequence of matches or mismatches and deletions in the alignment.</p>
<ul>
<LI><PRE><strong>First peptide sequence:</strong>  SERCQVVVISIGPISVLSMILDFY</PRE></LI>
<LI><PRE><strong>Second peptide sequence:</strong> SDRCQVLVISILSMIGLDFY</PRE></LI>
<LI><PRE><strong>First corresponding cigar line:</strong>  20MD4M</PRE></LI>
<LI><PRE><strong>Second corresponding cigar line:</strong> 11M5D9M</PRE></LI>
</ul>
<p>The alignment will be:</p>
<dl><dd>
<table border="1" cellpadding="2" cellspacing="1" bgcolor="#dcffff">
  <caption>Example of alignment reconstruction</caption>
  <tbody>
    <tr>
      <th>First peptide cigar line</th>
      <td>M</td>
      <td>M</td>
      <td>M</td>
      <td>M</td>
      <td>M</td>
      <td>M</td>
      <td>M</td>
      <td>M</td>
      <td>M</td>
      <td>M</td>
      <td>M</td>
      <td>M</td>
      <td>M</td>
      <td>M</td>
      <td>M</td>
      <td>M</td>
      <td>M</td>
      <td>M</td>
      <td>M</td>
      <td>M</td>
      <td>D</td>
      <td>M</td>
      <td>M</td>
      <td>M</td>
      <td>M</td>
    </tr>
    <tr>
      <th>First aligned peptide</th>
      <td>S</td>
      <td>E</td>
      <td>R</td>
      <td>C</td>
      <td>Q</td>
      <td>V</td>
      <td>V</td>
      <td>V</td>
      <td>I</td>
      <td>S</td>
      <td>I</td>
      <td>G</td>
      <td>P</td>
      <td>I</td>
      <td>S</td>
      <td>V</td>
      <td>L</td>
      <td>S</td>
      <td>M</td>
      <td>I</td>
      <td>-</td>
      <td>L</td>
      <td>D</td>
      <td>F</td>
      <td>Y</td>
    </tr>
    <tr>
      <th>Second aligned peptide</th>
      <td>S</td>
      <td>D</td>
      <td>R</td>
      <td>C</td>
      <td>Q</td>
      <td>V</td>
      <td>L</td>
      <td>V</td>
      <td>I</td>
      <td>S</td>
      <td>I</td>
      <td>-</td>
      <td>-</td>
      <td>-</td>
      <td>-</td>
      <td>-</td>
      <td>L</td>
      <td>S</td>
      <td>M</td>
      <td>I</td>
      <td>G</td>
      <td>L</td>
      <td>D</td>
      <td>F</td>
      <td>Y</td>
    </tr>
    <tr>
      <th>Second peptide cigar line</th>
      <td>M</td>
      <td>M</td>
      <td>M</td>
      <td>M</td>
      <td>M</td>
      <td>M</td>
      <td>M</td>
      <td>M</td>
      <td>M</td>
      <td>M</td>
      <td>M</td>
      <td>D</td>
      <td>D</td>
      <td>D</td>
      <td>D</td>
      <td>D</td>
      <td>M</td>
      <td>M</td>
      <td>M</td>
      <td>M</td>
      <td>M</td>
      <td>M</td>
      <td>M</td>
      <td>M</td>
      <td>M</td>
    </tr>
  </tbody>
</table>
</dd></dl>
<p></p>
<table border="2" cellpadding="2" bgcolor="#dcdcdc">
  <caption>homology_member table description</caption>
  <thead>
    <tr>
      <th>Field</th>
      <th>Type</th>
      <th>Null</th>
      <th>Key</th>
      <th>Default</th>
      <th>Extra</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>homology_id</td>
      <td>int(10) unsigned</td>
      <td></td>
      <td>PRI</td>
      <td>0</td>
      <td></td>
      <td>external reference to <A href="#homology">homology.homology_id</A></td>
    </tr>
    <tr>
      <td>member_id</td>
      <td>int(10) unsigned</td>
      <td></td>
      <td>PRI</td>
      <td>0</td>
      <td></td>
      <td>external reference to <A href="#member">member.member_id</A>.It refers to the corresponding gene
      (ENSEMBL_GENE).</td>
    </tr>
    <tr>
      <td>peptide_member_id</td>
      <td>int(10) unsigned</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>external reference to <A href="#member">member.member_id</A>.It refers to the peptide/protein
      (ENSEMBL_PEP).</td>
    </tr>
    <tr>
      <td>peptide_align_feature_id</td>
      <td>int(10) unsigned</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>external reference to <A href="#peptide_align_feature">peptide_align_feature.peptide_align_feature_id</A></td>
    </tr>
    <tr>
      <td>cigar_line</td>
      <td>mediumtext</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>an internal description of the alignment. It contains mathces/mismatches (M) and delations (D)and refers to
      the corresponding peptide_member_id sequence.</td>
    </tr>
    <tr>
      <td>cigar_start</td>
      <td>int(10)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>defines the first aligned aminoacid</td>
    </tr>
    <tr>
      <td>cigar_end</td>
      <td>int(10)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>defines the last aligned aminoacid</td>
    </tr>
    <tr>
      <td>perc_cov</td>
      <td>int(10)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>defines the percentage of the peptide which has been aligned</td>
    </tr>
    <tr>
      <td>perc_id</td>
      <td>int(10)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>defines the percentage of identity between both homologues</td>
    </tr>
    <tr>
      <td>perc_pos</td>
      <td>int(10)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>defines the percentage of positivity (similarity) between both homologues</td>
    </tr>
  </tbody>
</table>
<p>E.g. the rows</p>
<PRE>
mysql> select * from homology_member where homology_id = 296648;
+-------------+-----------+-------------------+--------------------------+------------+-иии
| homology_id | member_id | peptide_member_id | peptide_align_feature_id | cigar_line |
+-------------+-----------+-------------------+--------------------------+------------+-иии
|      296648 |       685 |               722 |                     NULL | 77M        |
|      296648 |    406456 |            406463 |                     NULL | 77M        |
+-------------+-----------+-------------------+--------------------------+------------+-иии

 иии-+-------------+-----------+----------+---------+----------+
     | cigar_start | cigar_end | perc_cov | perc_id | perc_pos |
 иии-+-------------+-----------+----------+---------+----------+
     |           1 |        77 |       30 |      79 |       87 |
     |          91 |       167 |       38 |      79 |       87 |
 иии-+-------------+-----------+----------+---------+----------+
</PRE>
<p>refer to the two homologue sequences defined by the <A href="#homology">homology.homology_id</A> 296648. The gene
corresponding to the first sequence can be retrieved using the <A href="#member">member.member_id</A> 685 and the
corresponding peptide using the <A href="#member">member.member_id</A> 722. Gene and peptide sequence of
the second homologue can retrieved in the same way.</p>
<p><STRONG>N.B.</STRONG>At the moment the peptide_align_feature_ids are NULL because they are only indirectly used
in the homology prediction but left there for compatibility with the old BRH system.</p>
</dd></dl>
<br>

<!--
#####################################################################
	TABLE SITEWISE_ALN
#####################################################################
 -->
<A name="sitewise_aln"></A>
<h3>sitewise_aln</h3>
<dl><dd>
<p>Contains the site wise dN/dS values for the codon alignments.</p>
<table border="2" cellpadding="2" bgcolor="#dcdcdc">
  <caption>sitewise_aln table description</caption>
  <thead>
    <tr>
      <th>Field</th>
      <th>Type</th>
      <th>Null</th>
      <th>Key</th>
      <th>Default</th>
      <th>Extra</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>sitewise_id</td>
      <td>int(10) unsigned</td>
      <td>NO</td>
      <td>PRI</td>
      <td>NULL</td>
      <td>auto_increment</td>
      <td>internal unique ID</td>
    </tr>
    <tr>
      <td>aln_position</td>
      <td>int(10) unsigned</td>
      <td>NO</td>
      <td>MUL</td>
      <td></td>
      <td></td>
      <td>alignment position of the codon/aminoacid</td>
    </tr>
    <tr>
      <td>node_id</td>
      <td>int(10) unsigned</td>
      <td>NO</td>
      <td>MUL</td>
      <td></td>
      <td></td>
      <td>root node_id for the subtree used in the analysis. External reference to <A href="#protein_tree_node">protein_tree_node.node_id</A></td>
    </tr>
    <tr>
      <td>tree_node_id</td>
      <td>int(10) unsigned</td>
      <td>NO</td>
      <td></td>
      <td></td>
      <td></td>
      <td>root node_id for the tree. External reference to <A href="#protein_tree_node">protein_tree_node.node_id</A></td>
    </tr>
    <tr>
      <td>omega</td>
      <td>float(10,5)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>Omega (dn/ds) value</td>
    </tr>
    <tr>
      <td>omega_lower</td>
      <td>float(10,5)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>Lower bound of support interval (confidence interval) for omega.</td>
    </tr>
    <tr>
      <td>omega_upper</td>
      <td>float(10,5)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>Upper bound of support interval (confidence interval) for omega.</td>
    </tr>
<!--     <tr> -->
<!--       <td>optimal</td> -->
<!--       <td>float(10,5)</td> -->
<!--       <td>YES</td> -->
<!--       <td></td> -->
<!--       <td>NULL</td> -->
<!--       <td></td> -->
<!--       <td>(minus) Log-probability of observing site given that it was evolving at the optimal value of omega.</td> -->
<!--     </tr> -->
    <tr>
      <td>threshold_on_branch_ds</td>
      <td>float(10,5)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>Synonymous length used in the analysis at which a branch is considered to be subject to saturation of synonymous mutation.</td>
    </tr>
    <tr>
      <td>type</td>
      <td>varchar(10)</td>
      <td>NO</td>
      <td></td>
      <td></td>
      <td></td>
      <td>describes the type of site found:
      <ul><LI><strong>positive1</strong> for positively selected site at uncorrected 95% probability</LI>
      <LI><strong>positive2</strong> for positively selected site at uncorrected 99% probability</LI>
      <LI><strong>positive3</strong> for positively selected site at corrected 95% probability</LI>
      <LI><strong>positive4</strong> for positively selected site at corrected 99% probability</LI>
      <LI><strong>negative1</strong> for negatively selected site at uncorrected 95% probability</LI>
      <LI><strong>negative2</strong> for negatively selected site at uncorrected 99% probability</LI>
      <LI><strong>negative3</strong> for negatively selected site at corrected 95% probability</LI>
      <LI><strong>negative4</strong> for negatively selected site at corrected 99% probability</LI>
      <LI><strong>constant</strong> for site with no changes</LI>
      <LI><strong>all_gaps</strong> for site with only gaps (due to the use of subalignments)</LI>
      <LI><strong>single_character</strong> for site with only 1 site and the remaining positions being gaps</LI>
      <LI><strong>synonymous</strong> for site with only synonymous changes</LI>
      <LI><strong>default</strong> for site that are none of the above</LI>
      </LI>
      </ul></td>
    </tr>
  </tbody>
</table>
<p>E.g. the row</p>
<PRE>
mysql> select * from sitewise_aln where sitewise_id=1;
+-------------+--------------+---------+--------------+-...
| sitewise_id | aln_position | node_id | tree_node_id |
+-------------+--------------+---------+--------------+-...
|           1 |           31 |    1827 |         1827 |
+-------------+--------------+---------+--------------+-...

...+---------+-------------+-------------+------------------------+-----------+
   | omega   | omega_lower | omega_upper | threshold_on_branch_ds | type      |
...+---------+-------------+-------------+------------------------+-----------+
   | 9.23330 |     3.22020 |    23.21050 |                1.50000 | positive3 |
...+---------+-------------+-------------+------------------------+-----------+

</PRE>
<p>defines a positively selected position with an omega value omega=9.23330 with a confidence interval of (3.2,23.2) 
calculated using a threshold_on_branch_ds=1.5 for the genetree with root 1827 (in this case a full genetree, not a subalignment as node_id=tree_node_id).</p>
</dd></dl>
<br>


<!--
#####################################################################
	TABLE FAMILY
#####################################################################
 -->
<A name="family"></a>
<h3>family</h3>
<dl><dd>
<p>Contains all the group homologies found. There are several <A href="#family_member">family_member</A> entries
for each family entry.</p>
<table border="2" cellpadding="2" bgcolor="#dcdcdc">
  <caption>family table description</caption>
  <thead>
    <tr>
      <th>Field</th>
      <th>Type</th>
      <th>Null</th>
      <th>Key</th>
      <th>Default</th>
      <th>Extra</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>family_id</td>
      <td>int(10) unsigned</td>
      <td></td>
      <td>PRI</td>
      <td>NULL</td>
      <td>auto_increment</td>
      <td>internal unique ID</td>
    </tr>
    <tr>
      <td>stable_id</td>
      <td>varchar(40)</td>
      <td></td>
      <td>UNI</td>
      <td></td>
      <td></td>
      <td>stable family ID</td>
    </tr>
    <tr>
      <td>method_link_species_set_id</td>
      <td>int(10) unsigned</td>
      <td></td>
      <td></td>
      <td>0</td>
      <td></td>
      <td>external reference to <A href="#method_link_species_set">method_link_species_set.method_link_species_set_id</A></td>
    </tr>
    <tr>
      <td>description</td>
      <td>varchar(255)</td>
      <td>YES</td>
      <td>MUL</td>
      <td>NULL</td>
      <td></td>
      <td>description of the family as found using the Longest Common String (LCS) of the descriptions of the
      member proteins.</td>
    </tr>
    <tr>
      <td>description_score</td>
      <td>double</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>Scores the accuracy of the annotation (max. 100)</td>
    </tr>
  </tbody>
</table>
<p>E.g. the row</p>
<PRE>
mysql> select * from family where family_id=4000;
+-----------+-------------------+----------------------------+-...
| family_id | stable_id         | method_link_species_set_id |
+-----------+-------------------+----------------------------+-...
|      4000 | fam51v00000004000 |                      30017 |
+-----------+-------------------+----------------------------+-...

...+--------------------------------------------------------------------+-------------------+
   | description                                                        | description_score |
...+--------------------------------------------------------------------+-------------------+
   | NUCLEAR PORE COMPLEX NUP205 NUCLEOPORIN NUP205.205 KDA NUCLEOPORIN |               100 |
...+--------------------------------------------------------------------+-------------------+
</PRE>
<p>defines a family homology found which stable ID is fam51v00000004000 and the description of this family is
"NUCLEAR PORE COMPLEX NUP205 NUCLEOPORIN NUP205.205 KDA NUCLEOPORIN" scored with a 100.</p>
</dd></dl>
<br>

<!--
#####################################################################
	TABLE FAMILY_MEMBER
#####################################################################
 -->
<A name="family_member"></A>
<h3>family_member</h3>
<dl><dd>
<p>Contains the proteins corresponding to protein family relationship found.
There are several family_member entries for each <A href="#family">family</A> entry.</p>
<table border="2" cellpadding="2" bgcolor="#dcdcdc">
  <caption>family_member table description</caption>
  <thead>
    <tr>
      <th>Field</th>
      <th>Type</th>
      <th>Null</th>
      <th>Key</th>
      <th>Default</th>
      <th>Extra</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>family_id</td>
      <td>int(10) unsigned</td>
      <td></td>
      <td>PRI</td>
      <td>0</td>
      <td></td>
      <td>external reference to <A href="#family">family.family_id</A></td>
    </tr>
    <tr>
      <td>member_id</td>
      <td>int(10) unsigned</td>
      <td></td>
      <td>PRI</td>
      <td>0</td>
      <td></td>
      <td>external reference to <A href="#member">member.member_id</A></td>
    </tr>
    <tr>
      <td>cigar_line</td>
      <td>mediumtext</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td>internal description of the multiple alignment (see <A href="#homology_member">homology_member</A>
      table)</td>
    </tr>
  </tbody>
</table>
<p>E.g. the rows</p>
<PRE>
mysql> SELECT * FROM family_member WHERE family_id = 13252;
+-----------+-----------+------------+
| family_id | member_id | cigar_line |
+-----------+-----------+------------+
|     13252 |     69013 | NULL       |
|     13252 |     69028 | 26D348M    |
|     13252 |    217503 | NULL       |
|     13252 |    217511 | 374M       |
|     13252 |    823691 | 26D348M    |
+-----------+-----------+------------+
</PRE>
<p>refer to the five members of the protein family 13252. The proteins can be retieved using the
<A href="#member">member_id</A>s. The multiple alignment can be restored using the cigar_lines.</p>
</dd></dl>
<br>

<!--
#####################################################################
	TABLE DOMAIN
#####################################################################
 -->
<A name="domain"></A>
<h3>domain</h3>
<dl><dd>
<p><span style="color: red">Not used by now</span></p>
<table border="2" cellpadding="2" bgcolor="#dcdcdc">
  <caption>domain table description</caption>
  <thead>
    <tr>
      <th>Field</th>
      <th>Type</th>
      <th>Null</th>
      <th>Key</th>
      <th>Default</th>
      <th>Extra</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>domain_id</td>
      <td>int(10) unsigned</td>
      <td></td>
      <td>PRI</td>
      <td>NULL</td>
      <td>auto_increment</td>
      <td>internal unique ID</td>
    </tr>
    <tr>
      <td>stable_id</td>
      <td>varchar(40)</td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <td>method_link_species_set</td>
      <td>int(10) unsigned</td>
      <td></td>
      <td></td>
      <td>0</td>
      <td></td>
      <td>external reference to <A href="#method_link_species_set">method_link_species_set.method_link_species_set_id</A></td>
    </tr>
    <tr>
      <td>description</td>
      <td>varchar(255)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td></td>
    </tr>
  </tbody>
</table>
</dd></dl>
<br>

<!--
#####################################################################
	TABLE DOMAIN_MEMBER
#####################################################################
 -->
<A name="domain_member"></A>
<h3>domain_member</h3>
<dl><dd>
<p><span style="color: red">Not used by now</span></p>
<table border="2" cellpadding="2" bgcolor="#dcdcdc">
  <caption>domain_member table description</caption>
  <thead>
    <tr>
      <th>Field</th>
      <th>Type</th>
      <th>Null</th>
      <th>Key</th>
      <th>Default</th>
      <th>Extra</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>domain_id</td>
      <td>int(10) unsigned</td>
      <td></td>
      <td>MUL</td>
      <td>0</td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <td>member_id</td>
      <td>int(10) unsigned</td>
      <td></td>
      <td>MUL</td>
      <td>0</td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <td>member_start</td>
      <td>int(10)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <td>member_end</td>
      <td>int(10)</td>
      <td>YES</td>
      <td></td>
      <td>NULL</td>
      <td></td>
      <td></td>
    </tr>
  </tbody>
</table>
</dd></dl>
<br>

<!--
#####################################################################
#####################################################################
	FOOTER
#####################################################################
#####################################################################
 -->

<hr>
<hr>
<p>
<a href="http://validator.w3.org/check?uri=referer"><img border="0"
 src="http://www.w3.org/Icons/valid-html401"
 alt="Valid HTML 4.01!" height="31" width="88"></A>
</p>
</body>
</html>
