<html>
<head>
<meta name="author" content="webteam" />
<meta name="navigation" content="Data" />
<meta name="update-period" content="2005-05-27" />
<title>Downloads</title>
</head>

<body>

<h2>Downloads</h2>

<h3 class="boxed">Introduction</h3>

<p>
All data sets generated by the Ensembl project are freely available to download
from the
<a href="ftp://ftp.ensembl.org/pub/" target="external">ftp.ensembl.org</a>
site.
Please see also the <a href="../about/disclaimer.html">disclaimer</a>.
</p>

<p>
<strong>Please note:</strong> Ensembl supports downloading of many more
correlation tables via the highly customisable
<a href="http://www.ebi.ac.uk/biomart/">BioMart</a> data mining tool.
You may find exploring this web-based data mining tool easier than extracting
information from our normalised database dumps.
</p>

<p>
Additionally, Ensembl would like to encourage users to directly extract
information from our databases via SQL rather than downloading huge flat files.
We offer a public MySQL interface at ensembldb.ensembl.org that accepts SQL
queries as user '<kbd>anonymous</kbd>'.
<a href="http://dev.mysql.com/downloads/" target="external">Client programs</a>
for accessing this interface are available via
<a href="http://www.mysql.com/" target="external">MySQL</a>.
</p>

<p>
It is also possible to
<a href="../software/index.html">install</a> the Ensembl API code locally and
configure it to access databases on ensembldb.ensembl.org.
Users can develop their own analysis scripts to access Ensembl's object
orientated representation of biological objects.
This is much easier than querying SQL and avoids downloading huge databases.
</p>


<h3 class="boxed">The Ensembl FTP Site</h3>

<p>
Ensembl provides sequence databases of gene, transcript and protein
predictions.
These sequences are suitable for a local installation of a sequence similarity
search system.
<a href="http://www.mysql.com/" target="external">MySQL</a> database table
dumps are available for <strong>all</strong> databases underlying the Ensembl
system.
These text format dumps can be imported into relational database management
systems which would enable
<a href="../software/website/installation/index.html">installation</a> of a
complete Ensembl mirror site.
</p>

<p>
There are five types of data dumps for each species on the FTP site:
</p>

<ul>

<li>
<p>
<strong>FASTA</strong> -
FASTA sequence databases of Ensembl gene, transcript and protein model
predictions. 
Since the 
<a href="http://www.ebi.ac.uk/help/formats.html#fasta" target="external">FASTA format</a>
does not permit sequence annotation, these database files are mainly intended
for use with local sequence similarity search algorithms.
Please see the <a href="./ftp_files.html">FASTA files</a> document for a more
detailed description of the header line format and the file naming conventions.
</p>

<ul>

<li>
<p>
<strong>DNA</strong> -
Masked and unmasked genome sequences associated with the assembly
(contigs, chromosomes etc.).
</p>
</li>

<li>
<p>
<strong>cDNA</strong> -
cDNA sequences for Ensembl or <i>ab initio</i> predicted genes.
</p>
</li>

<li>
<p>
<strong>Peptides</strong> -
Protein sequences for Ensembl or <i>ab initio</i> predicted genes.
</p>
</li>

<li>
<p>
<strong>RNA</strong> -
Non-coding RNA gene preditions.
</p>
</li>

</ul>

</li>


<li>
<p>
<strong>Flatfile</strong> - 
Flat files allow more extensive sequence annotation by means of feature tables
and contain thus the genome sequence as annotated by the automated Ensembl
<a href="./docs/genome_annotation.html">genome annotation pipeline</a>.
Each nucleotide sequence record in a flat file represents a 1Mb slice of the
genome sequence.
Flat files are broken into chunks of 1000 sequence records for easier
downloading.
</p>

<ul>

<li>
<p>
<strong>EMBL</strong> -
Ensembl database dumps in 
<a href="http://www.ebi.ac.uk/embl/" target="external">EMBL</a> nucleotide
sequence 
<a href="http://www.ebi.ac.uk/embl/Documentation/User_manual/usrman.html" target="external">database format</a>
</p>
</li>

<li>
<p>
<strong>GenBank</strong> -
Ensembl database dumps in 
<a href="http://www.ncbi.nlm.nih.gov/Genbank/" target="external">GenBank</a>
nucleotide sequence
<a href="http://www.ncbi.nlm.nih.gov/Sitemap/samplerecord.html" target="external">database format</a>
</p>
</li>

</ul>

</li>


<li>
<p>
<b>MySQL</b> -
Database table dumps in text format, as well as SQL table definition files
should be portable to any SQL database.
Generally, the FTP directory tree contains one one directory per database.
For more information about these databases and their associated Application
Programming Interfaces (or APIs) see the
<a href="../software/index.html">software</a> section.
</p>
</li>


<li>
<p>
<b>GTF</b> -
Gene sets for each species. These files include annotations of both coding
and non-coding genes. This file format is described <a href="http://mblab.wustl.edu/GTF2.html">here</a>.
</p>
</li>


<li>
<p>
<b>EMF flatfile dumps</b> -
Alignments of resequencing data are available for several species as Ensembl
Multi Format (EMF) flatfile dumps. The accompanying README file describes the file
format.
</p>

<p>
Also, the same format is used to dump whole-genome multiple alignments as
well as gene-based multiple alignments and phylogentic trees used to infer
Ensembl orthologues and paralogues. These files are available in the ensembl_compara
database which will be found in the <a href="ftp://ftp.ensembl.org/pub/current_multi_species/">multi_species directory</a>.
</p>
</li>

</ul>

<p>
These additional documents explain the FTP directory
<a href="./ftp_structure.html">structure</a> and the 
<a href="./ftp_files.html">FASTA file</a> naming and header conventions used on this site.
</p>

[[INCLUDE::/info/data/download_links.inc]]

</body>
</html>
