<html>
<head>
<title>Tutorial</title>
<meta name="order" content="1" />
<script type="text/javascript">

	// Function to show/hide divs
	function show_hide (param) {
		div   = document.getElementById('div_'+param);
		alink = document.getElementById('a_'+param);
		if (div.style.display=='inline') {
			div.style.display='none';
			alink.innerHTML='Show';
		}
		else {
			if (div.style.display=='none') {
				div.style.display='inline';
				alink.innerHTML='Hide';
			}
		}
	}
</script>
</head>

<body>
<div>
	
	
	<div style="float:right"><img src="/img/vep_logo.png"/></div>
	
	<h1 id="top"><span style="color:#006">Variant Effect Predictor</span> <img src="/i/16/star.png"> <span style="color:#666">Tutorial</span></h1>
    
    
    <hr/>
    <h1 id="tutorial"> Basic tutorial </h1>
    
    <p> <b>NOTE:</b> If you're using a UNIX or Mac system, you can dive straight
    into this tutorial by opening your favourite terminal application. If you're
    on Windows you might like to have a look at the <img src="/i/16/windows.png"
    height="12px"> <a href="vep_download.html#windows">guide for Windows
    users</a> before starting. </p>
    
    <p> Have you downloaded the VEP yet? Click this link to <img src="/i/16/download.png" height="12px"> <a
    href="http://cvs.sanger.ac.uk/cgi-bin/viewvc.cgi/ensembl-tools/scripts/variant_effect_predictor.tar.gz?view=tar&root=ensembl&pathrev=branch-ensembl-[[SPECIESDEFS::ENSEMBL_VERSION]]">download</a> if you haven't already. If you prefer to do things on
    the terminal, you can use curl to grab and unpack the VEP archive: </p>
    
    <pre class="code"><b>&gt; curl "http://cvs.sanger.ac.uk/cgi-bin/viewvc.cgi/ensembl-tools/scripts/variant_effect_predictor.tar.gz?view=tar&root=ensembl&pathrev=branch-ensembl-[[SPECIESDEFS::ENSEMBL_VERSION]]" | tar xz
&gt; cd variant_effect_predictor</b></pre>
    
    <p> If you clicked the download link, let's find the file and unpack it ourselves. Typically the file will be in your <b>Downloads</b> folder: </p>
    
    <pre class="code"><b>&gt; cd ~/Downloads
&gt; tar xfz variant_effect_predictor.tar.gz
&gt; cd variant_effect_predictor</b></pre>
    
    <p> The VEP uses "cache files" or a remote database to read genomic data.
    Using cache files gives the best performance - let's set one up using the
    installer: </p>
    
    <pre class="code"><b>&gt; perl INSTALL.pl</b>
Hello! This installer is configured to install v[[SPECIESDEFS::ENSEMBL_VERSION]] of the Ensembl API for use by the VEP.
It will not affect any existing installations of the Ensembl API that you may have.

It will also download and install cache files from Ensembl's FTP server.

Checking for installed versions of the Ensembl API...done
It looks like you already have v[[SPECIESDEFS::ENSEMBL_VERSION]] of the API installed.
You shouldn't need to install the API

Skip to the next step (n) to install cache files

Do you want to continue installing the API (y/n)?</pre>
    
    <p> If you haven't yet installed the API, type "y" followed by enter,
    otherwise type "n" (perhaps if you ran the installer before). At the next
    prompt, type "y" to install cache files</p>
    
    <pre class="code">Do you want to continue installing the API (y/n)? n
 - skipping API installation

The VEP can either connect to remote or local databases, or use local cache files.
Cache files will be stored in /nfs/users/nfs_w/wm2/.vep
Do you want to install any cache files (y/n)? y

Downloading list of available cache files
The following species/files are available; which do you want (can specify multiple separated by spaces):
1 : ailuropoda_melanoleuca_vep_[[SPECIESDEFS::ENSEMBL_VERSION]].tar.gz
2 : anolis_carolinensis_vep_[[SPECIESDEFS::ENSEMBL_VERSION]].tar.gz
3 : bos_taurus_vep_[[SPECIESDEFS::ENSEMBL_VERSION]].tar.gz
...
24 : homo_sapiens_vep_[[SPECIESDEFS::ENSEMBL_VERSION]].tar.gz
...

?</pre>
    <p> Type "24" (or the relevant number for homo_sapiens) to install the cache
    for human. This will take a little while to download and unpack! By default
    the VEP assumes you are working in human; it's easy to switch to any other
    species using <a href="vep_options.html#opt_species">--species
    [species]</a>. </p>
    
    <pre class="code">? 24
 - downloading ftp://ftp.ensembl.org/pub/release-[[SPECIESDEFS::ENSEMBL_VERSION]]/variation/VEP/homo_sapiens_vep_[[SPECIESDEFS::ENSEMBL_VERSION]].tar.gz
 - unpacking homo_sapiens_vep_[[SPECIESDEFS::ENSEMBL_VERSION]].tar.gz

Success</pre>
    
    <p> By default the VEP installs cache files in a folder in your home area
    (<b>$HOME/.vep</b>); you can easily change this using the <b>-d</b> flag
    when running the install script. Have a look at the <a
    href="vep_download.html#installer">installer documentation</a> for more
    details. </p>
    
    <p> The VEP needs some input containing variant positions to run. In their
    most basic form, this should just be a chromosomal location and a pair of
    alleles (reference and alternate). The VEP can also use common formats such
    as VCF and pileup as input; these are commonly produced by variant calling
    packages. Have a look at the <a href="../vep_formats.html#input">Data
    formats</a> page for more information. </p>
    
    <p> We can now use our cache file to run the VEP on the supplied example
    file <b>example.vcf</b>, which is a VCF file containing variants from the
    1000 Genomes Project exon-sequencing pilot: </p>
    
    <pre class="code"><b>&gt; perl variant_effect_predictor.pl -i example.vcf --cache</b>
2013-07-31 09:17:54 - Read existing cache info
2013-07-31 09:17:54 - Starting...
ERROR: Output file variant_effect_output.txt already exists. Specify a different output file
with --output_file or overwrite existing file with --force_overwrite</pre>
    
    <p> You may see this error message if you've already run the script once.
    The VEP tries not to trample over your existing files unless you tell it to.
    So let's tell it to using <a
    href="vep_options.html#opt_force_overwrite">--force_overwrite</a> </p>
    
    <pre class="code"><b>&gt; perl variant_effect_predictor.pl -i example.vcf --cache --force_overwrite</b>
2013-07-31 09:19:30 - Read existing cache info
2013-07-31 09:19:30 - Starting...
2013-07-31 09:19:30 - Detected format of input file as vcf
2013-07-31 09:19:30 - Read 173 variants into buffer
2013-07-31 09:19:30 - Reading transcript data from cache and/or database
[================================================================================]  [ 100% ]
2013-07-31 09:19:31 - Retrieved 3037 transcripts (0 mem, 3106 cached, 0 DB, 69 duplicates)
2013-07-31 09:19:31 - Analyzing chromosome 21
2013-07-31 09:19:31 - Analyzing variants
[================================================================================]  [ 100% ]
2013-07-31 09:19:32 - Calculating consequences
[================================================================================]  [ 100% ]
2013-07-31 09:19:32 - Analyzing chromosome 22
2013-07-31 09:19:32 - Analyzing variants
[================================================================================]  [ 100% ]
2013-07-31 09:19:33 - Calculating consequences
[================================================================================]  [ 100% ]
2013-07-31 09:19:34 - Processed 173 total variants (43 vars/sec, 43 vars/sec total)
2013-07-31 09:19:34 - Wrote stats summary to variant_effect_output.txt_summary.html
2013-07-31 09:19:34 - Finished!</pre>
    
    <p> Note that the VEP has told us it retrieved 3037 transcripts from the
    cache - this means the cache we downloaded is working OK. </p>
    
    <p> The VEP produces status output as it runs to give you an idea of the
    progress it is making. You can silence this output using <a href="vep_options.html#opt_quiet">--quiet</a>,
    though note that errors and warnings will still be printed out. </p>
    
    <p> By default the VEP writes to a file named "variant_effect_output.txt" -
    you can change this file name using <a
    href="vep_options.html#opt_output_file">-o</a>. Let's have a look at the
    output that the script generated. </p>
    
    <pre class="code"><b>&gt; head variant_effect_output.txt</b>
## ENSEMBL VARIANT EFFECT PREDICTOR v[[SPECIESDEFS::ENSEMBL_VERSION]]
## Output produced at 2013-07-31 09:19:30
## Connected to homo_sapiens_core_[[SPECIESDEFS::ENSEMBL_VERSION]]_37 on ensembldb.ensembl.org
## Using cache in /nfs/users/nfs_w/wm2/.vep/homo_sapiens/[[SPECIESDEFS::ENSEMBL_VERSION]]
## Using API version [[SPECIESDEFS::ENSEMBL_VERSION]], DB version [[SPECIESDEFS::ENSEMBL_VERSION]]
## Extra column keys:
## DISTANCE : Shortest distance from variant to transcript
<span style="color:#006">#Uploaded_variation  Location      Allele  Gene            Feature         Feature_type    Consequence ...</span>
rs116645811          21:26960070   A       ENSG00000260583 ENST00000567517 Transcript      upstream_gene_variant ...
rs116645811          21:26960070   A       ENSG00000154719 ENST00000352957 Transcript      intron_variant ...</pre>
    
    <p> The lines starting with "#" are header or meta information lines. The
    final one of these (highlighted in blue above) gives the column names for
    the data that follows. To see more information about the VEP's output
    format, see the <a href="../vep_formats.html#output">Data formats</a> page.
    </p>
    
    <p> We can see two lines of output here, both for the uploaded variant named
    rs116645811. In many cases, a variant will fall in more than one transcript.
    Typically this is where a single gene has multiple splicing variants;
    however, as in this case, it can also be that the variant overlaps
    transcripts from two different genes. Here our variant has a consequence for
    the genes ENSG00000260583 and ENSG00000154719. </p>
    
    <p> In the consequence column, we can see two different terms,
    upstream_gene_variant and intron_variant. These terms are standardised terms
    for describing the effects of sequence variants on genomic features,
    produced by the <a rel="external" href="http://www.sequenceontology.org/">Sequence Ontology
    (SO)</a>. See our <a
    href="/info/docs/variation/predicted_data.html#consequences">predicted
    data</a> page for a guide to the consequence types that the VEP and Ensembl
    uses. </p>
    
    <p> Let's try something a little more interesting. SIFT is an algorithm for
    predicting whether a given change in a protein sequence will be deleterious
    to the function of that protein. The VEP can give SIFT predictions for any
    of the missense variants that it predicts. To do this, simply add <a
    href="vep_options.html#opt_sift">--sift b</a> (the b means we want
    <b>b</b>oth the prediction and the score):</p>
    
    <pre class="code"><b>&gt; perl variant_effect_predictor.pl -i example.vcf --cache --force_overwrite --sift b</b></pre>
    
    <p> SIFT calls variants either "deleterious" or "tolerated". We can use the
    VEP's <a href="vep_filter.html">filtering script</a> to find only those that
    SIFT considers deleterious: </p>
    
    <pre class="code"><b>&gt; perl filter_vep.pl -i variant_effect_output.txt -filter "SIFT is deleterious" | head -n15</b>
...
## SIFT : SIFT prediction
#Uploaded_variation     Location        Allele  Gene            Feature         ...  Extra
rs114053718             21:34029195     G       ENSG00000159082 ENST00000382499 ...  SIFT=deleterious(0)
rs114053718             21:34029195     G       ENSG00000159082 ENST00000382491 ...  SIFT=deleterious(0)
rs114053718             21:34029195     G       ENSG00000159082 ENST00000322229 ...  SIFT=deleterious(0)
rs114053718             21:34029195     G       ENSG00000159082 ENST00000433931 ...  SIFT=deleterious(0)</pre>
    
    <p> Note that the SIFT score appears in the "Extra" column, as a key/value
    pair. This column can contain multiple key/value pairs depending on the
    options you give to the VEP script. See the <a
    href="../vep_formats.html#output">Data formats</a> page for more information on
    the fields in the Extra column. </p>
    
    <p>You can also configure how the VEP writes its output using the <a
    href="vep_options.html#opt_fields">--fields</a> flag. </p>
    
    <p> You'll also see that we have multiple results for the same gene,
    ENSG00000159082. Let's say we're only interested in what is considered the
    canonical transcript for this gene (<a
    href="vep_options.html#opt_canonical">--canonical</a>), and that we want to
    know what the commonly used gene symbol from HGNC is for this gene (<a
    href="vep_options.html#opt_symbol">--symbol</a>). We can also use a UNIX
    pipe to pass the output from the VEP directly into the filtering script:
    </p>
    
    <pre class="code"><b>&gt; perl variant_effect_predictor.pl -i example.vcf --cache --force_overwrite --sift b --canonical --symbol \
--fields Uploaded_variation,SYMBOL,CANONICAL,SIFT -o STDOUT | \
perl filter_vep.pl -format vep -filter "CANONICAL is YES and SIFT is deleterious"</b>
...
#Uploaded_variation  SYMBOL    CANONICAL  SIFT
rs114053718          SYNJ1     YES        deleterious(0)
rs2254562            SYNJ1     YES        deleterious(0.04)
rs4986958            IFNGR2    YES        deleterious(0.03)
rs16994704           PIGP      YES        deleterious(0.01)
...
rs115264708          PHF21B    YES        deleterious(0.04)</pre>
    
    <p> So now we can see all of the variants that have a deleterious effect on
    canonical transcripts, and the symbol for their genes. Nice! </p>
    
    <hr/>
    <h3> Over to you! </h3>
    
    <p> This has been just a short introduction to the capabilities of the VEP -
    have a look through some more of the <a href="vep_options.html">options</a>,
    see them all on the command line using <a href="vep_options.html#opt_help">--help</a>, or try using the
    shortcut <a href="vep_options.html#opt_everything">--everything</a> which switches on almost all available output
    fields! Try out the different options in the <a
    href="vep_filter.html">filtering script</a>, and if you're feeling
    adventurous why not use some of your <a href="vep_custom.html">own data to
    annotate your variants</a> or have a go with a <a
    href="vep_plugins.html">plugin</a> or two. </p>
	
	</div>
</div>

</body>
</html>
