<html>
<head>
<title>Running the script</title>
<meta name="order" content="3" />
</head>

<body>
    
<style>
tr:nth-child(odd) {background-color: #f0f0f0;}
</style>

<div>
	
	<div style="float:right"><img src="/img/vep_logo.png"/></div>
	
	<h1 id="top"><span style="color:#006">Variant Effect Predictor</span> <img src="/i/16/setting.png"> <span style="color:#666">Running the VEP</span></h1>
	<hr/>
	
	<p> The VEP script is run on the command line as follows: </p>
	
	<pre class="code"> perl variant_effect_predictor.pl [options] </pre>
	
	<p> where [options] represent a set of flags and options to the script.
	These can be listed using the flag <a href="#opt_help">--help</a>: </p>
	
	<pre class="code"> perl variant_effect_predictor.pl --help </pre>
	
    <p> Users should download a cache file for their species of interest, using
    either the <a href="vep_download.html#installer">installer script</a> or by
    following the <a href="vep_cache.html#pre">documentation</a>, and run the VEP with either
    the <a href="#opt_cache">--cache</a> or <a href="#opt_offline">--offline</a>
    option. </p>
	
	<p> For smaller input files, it is possible for the VEP to connect to
	Ensembl's public database servers in place of the cache; to enable this, use
	<a href="#opt_database">--database</a> </p>
	
	<p> Most users will need to use only a few of the options described below;
	for most the following command will be enough to get started with: </p>
	
	<pre class="code"> perl variant_effect_predictor.pl --cache -i input.txt -o output.txt </pre>
	
	<p> where input.txt contains data in one of the compatible input formats,
	and output.txt is the output file created by the script. See <a
	href="../vep_formats.html">Data Formats</a> for more detail on input and output
	formats.</p>
	
    <p> Options can be passed as the full string (e.g. <a
    href="vep_options.html#opt_format">--format</a>), or as the shortest unique
    string among the options (e.g. <a
    href="vep_options.html#opt_form">--form</a> for <a
    href="vep_options.html#opt_format">--format</a>, since there is another
    option <a
    href="vep_options.html#opt_force_overwrite">--force_overwrite</a>). </p>
	
    <p> Options can also be read from a configuration file - either passively
    stored as $HOME/.vep/vep.ini, or actively using <a
    href="#opt_config">--config</a>. </p>

	
	<hr/>
	<h2 id="basic">Basic options</h2>
	
	<table class="ss">
		<tr>
			<th width="20%">Flag</th>
			<th width="10%">Alternate</th>
			<th>Description</th>
		</tr>
		<tr id="opt_help">
			<td><pre>--help</pre></td>
			<td>&nbsp;</td>
			<td>Display help message and quit</td>
		</tr>
		<tr id="opt_verbose">
			<td><pre>--verbose</pre></td>
			<td><pre>-v</pre></td>
			<td>
				Output longer status messages as the script runs. This option
				can be used to generate the basis of a configuration file - see
				--config below. <i>Not used by default</i>
			</td>
		</tr>
		<tr id="opt_quiet">
			<td><pre>--quiet</pre></td>
			<td><pre>-q</pre></td>
			<td>Suppress status and warning messages. <i>Not used by
			default</i></td>
		</tr>
		<tr id="opt_no_progress">
			<td><pre>--no_progress</pre></td>
			<td>&nbsp;</td>
			<td>Don't show progress bars. <i>Progress bars shown by
			default</i></td>
		</tr>
		<tr id="opt_config">
			<td><pre>--config [filename]</pre></td>
			<td>&nbsp;</td>
			<td>
				Load configuration options from a config file. The config file
				should consist of whitespace-separated pairs of option names and
				settings e.g.: 
	<pre>output_file   my_output.txt
species       mus_musculus
format        vcf
host          useastdb.ensembl.org</pre>
                A config file can also be implicitly read; save the file as
                $HOME/.vep/vep.ini (or equivalent directory if using
                <a href="vep_options.html#opt_dir">--dir</a>). Any options in this file will be overridden by
                those specified in a config file using <a href="vep_options.html#opt_config">--config</a>, and in
                turn by any options manually specified on the command line. You
                can create a quick version file of this by setting the flags as
                normal and running the script in verbose (<b>-v</b>) mode. This
                will output lines that can be copied to a config file that can
                be loaded in on the next run using <a href="vep_options.html#opt_config">--config</a>. <i>Not used
                by default</i>
			</td>
		</tr>
		<tr id="opt_everything">
			<td><pre>--everything</pre></td>
			<td>&nbsp;</td>
			<td>Shortcut flag to switch on all of the following:
				<p>
					<a href="#opt_sift">--sift b</a>,
					<a href="#opt_polyphen">--polyphen b</a>,
					<a href="#opt_ccds">--ccds</a>,
					<a href="#opt_hgvs">--hgvs</a>,
					<a href="#opt_symbol">--symbol</a>,
					<a href="#opt_numbers">--numbers</a>,
					<a href="#opt_domains">--domains</a>,
					<a href="#opt_regulatory">--regulatory</a>,
					<a href="#opt_canonical">--canonical</a>,
					<a href="#opt_protein">--protein</a>,
                    <a href="#opt_biotype">--biotype</a>,
					<a href="#opt_gmaf">--gmaf</a>,
                    <a href="#opt_maf_1kg">--maf_1kg</a>,
                    <a href="#opt_maf_esp">--maf_esp</a>,
                    <a href="#opt_pubmed">--pubmed</a>
				</ul>
			</td>
		</tr>
		<tr id="opt_fork">
			<td><pre>--fork [num_forks]</pre></td>
			<td>&nbsp;</td>
			<td>
				Enable <a href="#forking">forking</a>, using the specified
				number of forks. Forking can dramatically improve the runtime of
				the script. <i>Not used by default</i>
			</td>
		</tr>
	</table>
	
	
	
	
	<hr/>
	
	<h2 id="input">Input options</h2>
	
	<table class="ss">
		<tr>
			<th width="20%">Flag</th>
			<th width="10%">Alternate</th>
			<th>Description</th>
		</tr>
		<tr id="opt_species">
			<td><pre>--species [species]</pre></td>
			<td>&nbsp;</td>
			<td>
				Species for your data. This can be the latin name e.g.
				"homo_sapiens" or any Ensembl alias e.g. "mouse". Specifying the
				latin name can speed up initial database connection as the
				registry does not have to load all available database aliases on
				the server. <i>Default = "homo_sapiens"</i>
			</td>
		</tr>
		<tr id="opt_input_file">
			<td><pre>--input_file [filename]</pre></td>
			<td><pre>-i</pre></td>
			<td>Input file name. If not specified, the script will attempt to read from STDIN.</td>
		</tr>
		<tr id="opt_format">
			<td><pre>--format [format]</pre></td>
			<td>&nbsp;</td>
			<td>
                <a href="../vep_formats.html#input">Input file format</a> - one of
                "ensembl", "vcf", "pileup", "hgvs", "id", "vep". By default, the
                script auto-detects the input file format. Using this option you
                can force the script to read the input file as Ensembl, VCF,
                pileup or HGVS format, a list of variant identifiers (e.g. rsIDs
                from dbSNP), or the output from the VEP (e.g. to add custom
                annotation to an existing results
				file using <a href="vep_options.html#opt_custom">--custom</a>). <i>Auto-detects format by default</i>
			</td>
		</tr>
		<tr id="opt_output_file">
			<td><pre>--output_file [filename]</pre></td>
			<td><pre>-o</pre></td>
			<td>
				Output file name. The script can write to STDOUT by specifying
				STDOUT as the output file name - this will force quiet mode.
				<i>Default = "variant_effect_output.txt"</i>
			</td>
		</tr>
		<tr id="opt_force_overwrite">
			<td><pre>--force_overwrite</pre></td>
			<td><pre>--force</pre></td>
			<td>
				By default, the script will fail with an error if the output
				file already exists. You can force the overwrite of the existing
				file by using this flag. <i>Not used by default</i>
			</td>
		</tr>
		<tr id="opt_stats_file">
			<td><pre>--stats_file [filename]</pre></td>
			<td>&nbsp;</td>
			<td>
                Summary stats file name. This is an <a
                href="../vep_formats.html#stats">HTML file</a> containing a summary
                of the VEP run - the file name must end ".htm" or ".html".
                <i>Default = "variant_effect_output.txt_summary.html"</i>
			</td>
		</tr>
		<tr id="opt_no_stats">
			<td><pre>--no_stats</pre></td>
			<td>&nbsp;</td>
			<td>
				Don't generate a stats file.
			</td>
		</tr>
		<tr id="opt_stats_text">
			<td><pre>--stats_text</pre></td>
			<td>&nbsp;</td>
			<td>
				Generate a plain text stats file in place of the HTML.
			</td>
		</tr>
		<tr id="opt_html">
			<td><pre>--html</pre></td>
			<td>&nbsp;</td>
			<td>
				Generate an additional HTML version of the output file
				containing hyperlinks to Ensembl and other resources. File name
				of this file is [output_file].html
			</td>
		</tr>
	</table>
	
	
	
	
	<hr/>
    
    <h2 id="cacheopt">Cache options</h2>
	
	<table class="ss">
		<tr>
			<th width="20%">Flag</th>
			<th width="10%">Alternate</th>
			<th>Description</th>
		</tr>
		<tr id="opt_cache">
			<td><pre>--cache</pre></td>
			<td>&nbsp;</td>
			<td>
				Enables use of the <a href="vep_cache.html#cache">cache</a>.
			</td>
		</tr>
		<tr id="opt_dir">
			<td><pre>--dir [directory]</pre></td>
			<td>&nbsp;</td>
			<td>
				Specify the base cache/plugin directory to use. <i>Default =
				"$HOME/.vep/"</i>
			</td>
		</tr>
		<tr id="opt_dir_cache">
			<td><pre>--dir_cache [directory]</pre></td>
			<td>&nbsp;</td>
			<td>
				Specify the cache directory to use. <i>Default =
				"$HOME/.vep/"</i>
			</td>
		</tr>
		<tr id="opt_dir_plugins">
			<td><pre>--dir_plugins [directory]</pre></td>
			<td>&nbsp;</td>
			<td>
				Specify the plugin directory to use. <i>Default =
				"$HOME/.vep/"</i>
			</td>
		</tr>
		<tr id="opt_offline">
			<td><pre>--offline</pre></td>
			<td>&nbsp;</td>
			<td>
                Enable <a href="vep_cache.html#offline">offline mode</a>. No
                database connections will be made, and only a complete cache
                (either downloaded or built using <a
                href="#opt_build">--build</a>) can be used for this mode. <i>Not
                used by default</i>
			</td>
		</tr>
		<tr id="opt_fasta">
			<td><pre>--fasta [file|dir]</pre></td>
			<td>&nbsp;</td>
			<td>
                Specify a FASTA file or a directory containing FASTA files to
                use to look up reference sequence. The first time you run the
                script with this parameter an index will be built which can take
                a few minutes. This is required if fetching HGVS annotations (<a
                href="#opt_hgvs">--hgvs</a>) or checking reference sequences (<a
                href="#opt_check_ref">--check_ref</a>) in offline mode (<a
                href="#opt_offline">--offline</a>), and optional with some
                performance increase in cache mode (<a
                href="#opt_cache">--cache</a>). See <a
                href="vep_cache.html#fasta">documentation</a> for more details.
                <i>Not used by default</i>
			</td>
		</tr>
	</table>
	  
	
	
	<hr/>
	
	
	
	<h2 id="output">Output options</h2>
	
	<table class="ss">
		<tr>
			<th width="20%">Flag</th>
			<th width="10%">Alternate</th>
			<th>Description</th>
		</tr>
		<tr id="opt_sift">
			<td><pre>--sift [p<b>|</b>s<b>|</b>b]</pre></td>
			<td>&nbsp;</td>
			<td>
				<span style="color:red;">Species limited</span>
				<a rel="external" href="http://sift.jcvi.org/" target="_blank">SIFT</a>
				predicts whether an amino acid substitution affects protein
				function based on sequence homology and the physical properties
				of amino acids. The VEP can output the <b>p</b>rediction term,
				<b>s</b>core or <b>b</b>oth. <i>Not used by default</i>
			</td>
		</tr>
		<tr id="opt_polyphen">
			<td><pre>--polyphen [p<b>|</b>s<b>|</b>b]</pre></td>
			<td><pre>--poly</pre></td>
			<td>
				<span style="color:red;">Human only</span>
				<a rel="external" href="http://genetics.bwh.harvard.edu/pph2/"
				target="_blank">PolyPhen</a> is a tool which predicts possible
				impact of an amino acid substitution on the structure and
				function of a human protein using straightforward physical and
				comparative considerations. The VEP can output the
				<b>p</b>rediction term, <b>s</b>core or <b>b</b>oth. <i>Not used
				by default</i>
			</td>
		</tr>
		<tr id="opt_regulatory">
			<td><pre>--regulatory</pre></td>
			<td>&nbsp;</td>
			<td>
				Look for overlaps with regulatory regions. The script can also
				call if a variant falls in a high information position within a
				transcription factor binding site. Output lines have a Feature
				type of RegulatoryFeature or MotifFeature. <i>Not used by
				default</i>
			</td>
		</tr>
		<tr id="opt_cell_type">
			<td><pre>--cell_type</pre></td>
			<td>&nbsp;</td>
			<td>
				Report only regulatory regions that are found in the given cell
				type(s). Can be a single cell type or a comma-separated list.
				The functional type in each cell type is reported under
				CELL_TYPE in the output. To retrieve a list of cell types, use
				<a href="vep_options.html#opt_cell_type">--cell_type list</a>. <i>Not used by default</i>
			</td>
		</tr>
		<tr id="opt_custom">
			<td><pre>--custom [filename]</pre></td>
			<td>&nbsp;</td>
			<td>
				Add custom annotation to the output. Files must be tabix indexed
				or in the bigWig format. Multiple files can be specified by
				supplying the <a href="vep_options.html#opt_custom">--custom</a> flag multiple times. <a
                href="vep_custom.html">See here</a> for full details. <i>Not
                used by default</i>
			</td>
		</tr>
		<tr id="opt_plugin">
			<td><pre>--plugin [plugin name]</pre></td>
			<td>&nbsp;</td>
			<td>
                Use named plugin. Plugin modules should be installed in the
                Plugins subdirectory of the VEP cache directory (defaults to
                $HOME/.vep/). Multiple plugins can be used by supplying the
                <a href="vep_options.html#opt_plugin">--plugin</a> flag multiple times. See <a
                href="vep_plugins.html">plugin documentation</a>. <i>Not used by
                default</i>
			</td>
		</tr>
		<tr id="opt_individual">
			<td><pre>--individual [all|ind list]</pre></td>
			<td>&nbsp;</td>
			<td>
				Consider only alternate alleles present in the genotypes of the
				specified individual(s). May be a single individual, a
				comma-separated list or "all" to assess all individuals
				separately. Individual variant combinations homozygous for the
				given reference allele will not be reported. Each individual and
				variant combination is given on a separate line of output. Only
				works with VCF files containing individual genotype data;
				individual IDs are taken from column headers. <i>Not used by
				default</i>
			</td>
		</tr>
		<tr id="opt_phased">
			<td><pre>--phased</pre></td>
			<td>&nbsp;</td>
			<td>
				Force VCF genotypes to be interpreted as phased. For use with
				plugins that depend on phased data. <i>Not used by
				default</i>
			</td>
		</tr>
        <tr id="opt_allele_number">
            <td><pre>--allele_number</pre></td>
			<td>&nbsp;</td>
			<td>
                Identify allele number from VCF input, where 1 = first ALT
                allele, 2 = second ALT allele etc. <i>Not used by default</i>
			</td>
		</tr>
        <tr id="opt_total_length">
			<td><pre>--total_length</pre></td>
			<td>&nbsp;</td>
			<td>
                Give cDNA, CDS and protein positions as Position/Length. <i>Not
                used by default</i>
			</td>
		</tr>
		<tr id="opt_numbers">
			<td><pre>--numbers</pre></td>
			<td>&nbsp;</td>
			<td>
				Adds affected exon and intron numbering to to output. Format is
				Number/Total. <i>Not used by default</i>
			</td>
		</tr>
		<tr id="opt_domains">
			<td><pre>--domains</pre></td>
			<td>&nbsp;</td>
			<td>
				Adds names of overlapping protein domains to output. <i>Not used
				by default</i>
			</td>
		</tr>
		<tr id="opt_no_escape">
			<td><pre>--no_escape</pre></td>
			<td>&nbsp;</td>
			<td>
				Don't URI escape HGVS strings. <i>Default = escape</i>
			</td>
		</tr>
		<tr id="opt_terms">
			<td><pre>--terms [ensembl<b>|</b>so]</pre></td>
			<td><pre>-t</pre></td>
			<td>
				The type of consequence terms to output. The Ensembl terms are
				described <a
				href="/info/docs/variation/predicted_data.html#consequences"
				target="_blank">here</a>. The <a
				rel="external" href="http://www.sequenceontology.org/" target="_blank">Sequence
				Ontology</a> is a joint effort by genome annotation centres to
				standardise descriptions of biological sequences. <i>Default =
				"SO"</i>
			</td>
		</tr>
	</table>
	  
	
	
	<hr/>
	
	
	
	<h2 id="ident">Identifiers</h2>
	
	<table class="ss">
		<tr>
			<th width="20%">Flag</th>
			<th width="10%">Alternate</th>
			<th>Description</th>
		</tr>
		<tr id="opt_hgvs">
			<td><pre>--hgvs</pre></td>
			<td>&nbsp;</td>
			<td>
				Add <a rel="external" href="http://www.hgvs.org/mutnomen/"
				target="_blank">HGVS</a> nomenclature based on Ensembl stable
				identifiers to the output. Both coding and protein sequence
				names are added where appropriate. Currently it is not possible
				to generate HGVS identifiers from the cache; a database
				connection must be made. <i>Not used by default</i>
			</td>
		</tr>
		<tr id="opt_protein">
			<td><pre>--protein</pre></td>
			<td>&nbsp;</td>
			<td>
				Add the Ensembl protein identifier to the output where
				appropriate. <i>Not used by default</i>
			</td>
		</tr>
		<tr id="opt_symbol">
			<td><pre>--symbol</pre></td>
			<td>&nbsp;</td>
			<td>
				Adds the gene symbol (e.g. HGNC) (where available) to the output.
				<i>Not used by default</i>
			</td>
		</tr>
		<tr id="opt_ccds">
			<td><pre>--ccds</pre></td>
			<td>&nbsp;</td>
			<td>
				Adds the CCDS transcript identifer (where available) to the
				output. <i>Not used by default</i>
			</td>
		</tr>
		<tr id="opt_canonical">
			<td><pre>--canonical</pre></td>
			<td>&nbsp;</td>
			<td>
				Adds a flag indicating if the transcript is the canonical
				transcript for the gene. <i>Not used by default</i>
			</td>
		</tr>
		<tr id="opt_biotype">
			<td><pre>--biotype</pre></td>
			<td>&nbsp;</td>
			<td>
				Adds the biotype of the transcript. <i>Not used by default</i>
			</td>
		</tr>
		<tr id="opt_xref_refseq">
			<td><pre>--xref_refseq</pre></td>
			<td>&nbsp;</td>
			<td>
				Output aligned RefSeq mRNA identifier for transcript. NB:
				theRefSeq and Ensembl transcripts aligned in this way MAY NOT,
				AND FREQUENTLY WILL NOT, match exactly in sequence, exon
				structure and protein product. <i>Not used by default</i>
			</td>
		</tr>
	</table>
	  
	
	
	<hr/>
	
	
	
	<h2 id="existing">Co-located variants</h2>
	
	<table class="ss">
		<tr>
			<th width="20%">Flag</th>
			<th width="10%">Alternate</th>
			<th>Description</th>
		</tr>
		<tr id="opt_check_existing">
			<td><pre>--check_existing</pre></td>
			<td>&nbsp;</td>
			<td>
                Checks for the existence of known variants that are co-located
                with your input. By default the alleles are not compared - to do
                so, use <a href="#opt_check_alleles">--check_alleles</a>. <i>Not
                used by default</i>
			</td>
		</tr>
		<tr id="opt_check_alleles">
			<td><pre>--check_alleles</pre></td>
			<td>&nbsp;</td>
			<td>
				When checking for existing variants, only report a co-located
				variant if none of the alleles supplied are novel. For example,
				if the user input has alleles A/G, and an existing co-located
				variant has alleles A/C, the co-located variant will not be
				reported.<br/><br/>Strand is also taken into account - in the
				same example, if the user input has alleles T/G but on the
				negative strand, then the co-located variant <b>will</b> be
				reported since its alleles match the reverse complement of user
				input. <i>Not used by default</i>
			</td>
		</tr>
		<tr id="opt_check_svs">
			<td><pre>--check_svs</pre></td>
			<td>&nbsp;</td>
			<td>
				Checks for the existence of structural variants that overlap your
				input. Currently requires database access. <i>Not used by
				default</i>
			</td>
		</tr>
		<tr id="opt_gmaf">
			<td><pre>--gmaf</pre></td>
			<td>&nbsp;</td>
			<td>
				Add the global minor allele frequency (MAF) from 1000 Genomes
				Phase 1 data for any existing variant to the output. <i>Not used
				by default</i>
			</td>
		</tr>
		<tr id="opt_maf_1kg">
			<td><pre>--maf_1kg</pre></td>
			<td>&nbsp;</td>
			<td>
                Add MAF from continental populations (AFR,AMR,ASN,EUR) of 1000
                Genomes Phase 1 to the output. Must be used with <a
                href="#opt_cache">--cache</a> <i>Not used by default</i>
			</td>
		</tr>
		<tr id="opt_maf_esp">
			<td><pre>--maf_esp</pre></td>
			<td>&nbsp;</td>
			<td>
				Include MAF from NHLBI-ESP populations. Must be used with
                <a href="#opt_cache">--cache</a> <i>Not used	by default</i>
			</td>
		</tr>
		<tr id="opt_pubmed">
			<td><pre>--pubmed</pre></td>
			<td>&nbsp;</td>
			<td>
                Report Pubmed IDs for publications that cite existing variant.
                Must be used with <a href="#opt_cache">--cache</a>. <i>Not used
                by default</i>
			</td>
		</tr>
		<tr id="opt_failed">
			<td><pre>--failed</pre></td>
			<td>&nbsp;</td>
			<td>
				When checking for co-located variants, by default the script
				will exclude variants that have been flagged as failed. Set this
				flag to include such variants. <i>Exclude by default</i>
			</td>
		</tr>
	</table>
	  
	
	
	<hr/>
	
	
	
	<h2 id="format">Data format options</h2>
	
	<table class="ss">
		<tr>
			<th width="20%">Flag</th>
			<th width="10%">Alternate</th>
			<th>Description</th>
		</tr>
		<tr id="opt_vcf">
			<td><pre>--vcf</pre></td>
			<td>&nbsp;</td>
			<td>
                Writes output in VCF format. Consequences are added in the INFO
                field of the VCF file, using the key "CSQ". Data fields are
                encoded separated by "&#124;"; the order of fields is written in
                the VCF header. Output fields can be selected by using <a
                href="#opt_fields">--fields</a>.<br/><br/> If the input format
                was VCF, the file will remain unchanged save for the addition of
                the CSQ field (unless using any filtering).<br/><br/> Custom
                data added with <a href="#opt_custom">--custom</a> are added as
                separate fields, using the key specified for each data
                file.<br/><br/> Commas in fields are
                replaced with ampersands (&amp;) to preserve VCF
                format.<br/><br/><i>Not used by default</i>
			</td>
		</tr>
		<tr id="opt_gvf">
			<td><pre>--gvf</pre></td>
			<td>&nbsp;</td>
			<td>
				Writes output in GVF format. <i>Not used by default</i>
			</td>
		</tr>
		<tr id="opt_original">
			<td><pre>--original</pre></td>
			<td>&nbsp;</td>
			<td>
                Writes output as a filtered set of the input. Must be used with
                <a href="#opt_filter">--filter</a>. Input lines are unchanged -
                consequences are calculated but not written to the output.
                <i>Not used by default</i>
			</td>
		</tr>
		<tr id="opt_fields">
			<td><pre>--fields [list]</pre></td>
			<td>&nbsp;</td>
			<td>
                Configure the output format using a comma separated list of
                fields. Fields may be those present in the default <a
                href="../vep_formats.html#output">output columns</a>, or any of
                those that appear in the Extra column (including those added by
                plugins or custom annotations). Output remains tab-delimited.
                <i>Not used by default</i>
			</td>
		</tr>
		<tr id="opt_convert">
			<td><pre>--convert [format]</pre></td>
			<td>&nbsp;</td>
			<td>
                Converts the input file to the specified format (one of
                "ensembl", "vcf", "pileup"). See <a
                href="vep_other.html#convert">documentation</a> for more
                details. Converted output is written to the file specified with
                <a href="#opt_output_file">--output_file</a>. No consequence
                prediction is carried out. <i>Not used by default</i>
			</td>
		</tr>
	</table>
	
	
	
	<hr/>
	
	<h2 id="filt">Filtering and QC options</h2>
    
    <p> <b>NOTE:</b> The filtering options here filter your results
    <b>before</b> they are written to your output file. Using the VEP's <a
    href="vep_filter.html">filtering script</a>, it is possible to filter your
    results <b>after</b> the VEP has run. This way you can retain all of the
    results and run multiple filter sets on the same results to find different
    data of interest. </p>
	
	<table class="ss">
		<tr>
			<th width="20%">Flag</th>
			<th width="10%">Alternate</th>
			<th>Description</th>
		</tr>
		<tr id="opt_check_ref">
			<td><pre>--check_ref</pre></td>
			<td>&nbsp;</td>
			<td>
				Force the script to check the supplied reference allele against
				the sequence stored in the Ensembl Core database. Lines that do
				not match are skipped. <i>Not used by default</i>
			</td>
		</tr>
		<tr id="opt_coding_only">
			<td><pre>--coding_only</pre></td>
			<td>&nbsp;</td>
			<td>
				Only return consequences that fall in the coding regions of
				transcripts. <i>Not used by default</i>
			</td>
		</tr>
		<tr id="opt_chr">
			<td><pre>--chr [list]</pre></td>
			<td>&nbsp;</td>
			<td>
				Select a subset of chromosomes to analyse from your file. Any
				data not on this chromosome in the input will be skipped. The
				list can be comma separated, with "-" characters representing an
				interval. For example, to include chromsomes 1, 2, 3, 10 and X
				you could use <a href="vep_options.html#opt_chr">--chr 1-3,10,X</a> <i>Not used by
				default</i>
			</td>
		</tr>
		<tr id="opt_no_intergenic">
			<td><pre>--no_intergenic</pre></td>
			<td>&nbsp;</td>
			<td>
				Do not include intergenic consequences in the output. <i>Not
				used by default</i>
			</td>
		</tr>
		<tr id="opt_most_severe">
			<td><pre>--most_severe</pre></td>
			<td>&nbsp;</td>
			<td>
				Output only the most severe consequence per variation.
				Transcript-specific columns will be left blank. <i>Not used by
				default</i>
			</td>
		</tr>
		<tr id="opt_summary">
			<td><pre>--summary</pre></td>
			<td>&nbsp;</td>
			<td>
				Output only a comma-separated list of all observed consequences
				per variation. Transcript-specific columns will be left blank.
				<i>Not used by default</i>
			</td>
		</tr>
		<tr id="opt_per_gene">
			<td><pre>--per_gene</pre></td>
			<td>&nbsp;</td>
			<td>
				Output only the most severe consequence per gene. The transcript
				selected is arbitrary if more than one has the same predicted
				consequence. <i>Not used by default</i>
			</td>
		</tr>
		<tr id="opt_filter_common">
			<td><pre>--filter_common</pre></td>
			<td>&nbsp;</td>
			<td>
				Shortcut flag for the filters below - this will exclude variants
				that have a co-located existing variant with global MAF &gt;
				0.01 (1%). May be modified using any of the following freq_*
				filters. For human, this can be used in offline mode for the
				following populations: 1KG_ALL, 1KG_AFR, 1KG_AMR, 1KG_ASN,
				1KG_EUR. <i>Not used by default</i>
			</td>
		</tr>
		<tr id="opt_check_frequency">
			<td><pre>--check_frequency</pre></td>
			<td>&nbsp;</td>
			<td>
				Turns on frequency filtering. Use this to include or exclude
				variants based on the frequency of co-located existing variants
                in the Ensembl Variation database. You must also specify all of
                the <a href="vep_options.html#opt_freq">--freq</a> flags below. Using this option requires a
                database connection - while it can be used with --cache, the
                database will still be accessed to retrieve frequency data.
                Frequencies used in filtering are added to the output under the
                FREQS key in the Extra field. <i>Not used by default</i>
			</td>
		</tr>
		<tr id="opt_freq_pop">
			<td><pre>--freq_pop [pop]</pre></td>
			<td>&nbsp;</td>
			<td>
				Name of the population to use in frequency filter. This can be
				the name of the population as it appears on the Ensembl website
				(suitable for most species), or in the following short form for
				human. 1000 genomes populations are currently pilot 1 (low
				coverage).
				<p>
					<table class="ss">
						<tr>
							<th>Example value for --freq_pop</th><th>Description</th>
						</tr>
						<tr>
							<td>1kg_all</td><td>1000 genomes combined population (global)</td>
						</tr>
						<tr>
							<td>1kg_afr</td><td>1000 genomes combined African populations (also amr, asn, eur)</td>
						</tr>
						<tr>
							<td>1kg_chb</td><td>1000 genomes CHB population</td>
						</tr>
						<tr>
							<td>hapmap_yri</td><td>HapMap YRI population</td>
						</tr>
						<tr>
							<td>1kg</td><td>Any 1000 genomes phase 1 population</td>
						</tr>
						<tr>
							<td>ceu</td><td>Any of HapMap or 1000 genomes CEU populations</td>
						</tr>
						<tr>
							<td>any</td><td>Any HapMap or 1000 genomes population</td>
						</tr>
					</table>
				</p>
			</td>
		</tr>
		<tr id="opt_freq_freq">
			<td><pre>--freq_freq [freq]</pre></td>
			<td>&nbsp;</td>
			<td>
				Minor allele frequency to use for filtering. Must be a float
				value between 0 and 0.5
			</td>
		</tr>
		<tr id="opt_freq_gt_lt">
			<td><pre>--freq_gt_lt [gt<b>|</b>lt]</pre></td>
			<td>&nbsp;</td>
			<td>
				Specify whether the frequency of the co-located variant must be
				greater than (<b>gt</b>) or less than (<b>lt</b>) the value
				specified with <a href="#opt_freq_freq">--freq_freq</a>
			</td>
		</tr>
		<tr id="opt_freq_filter">
			<td><pre>--freq_filter [exclude<b>|</b>include]</pre></td>
			<td>&nbsp;</td>
			<td>
				Specify whether to <b>exclude</b> or <b>include</b> only
				variants that pass the frequency filter
			</td>
		</tr>    
		<tr id="opt_filter">
			<td><pre>--filter [filters]</pre></td>
			<td>&nbsp;</td>
			<td>
				Filter the output on consequence type. Multiple allowed types
				can be specified, separated by commas. SO terms should ideally
				be used, although Ensembl and NCBI types are also allowed.
				Consequence types can be excluded by adding "no_" to the start
				of the filter name. Shortcuts to common groupings are available:
				<p>
					<table class="ss">
						<tr>
							<th>Shortcut name</th><th>Description</th>
						</tr>
						<tr>
							<td>upstream</td><td>Any upstream variant</td>
						</tr>
						<tr>
							<td>downstream</td><td>Any downstream variant</td>
						</tr>
						<tr>
							<td>utr</td><td>Any UTR variant</td>
						</tr>
						<tr>
							<td>splice</td><td>Any splicing region variants</td>
						</tr>
						<tr>
							<td>coding</td><td>Any variant that falls in the coding region of a transcript</td>
						</tr>
						<tr>
							<td>coding_change</td><td>Any variant that causes a coding change in the transcript</td>
						</tr>
						<tr>
							<td>regulatory</td><td>Any variant that falls in a regulatory or binding motif feature</td>
						</tr>
					</table>
				</p> To reproduce a filtered version of the input file, add the flag
				<a href="#opt_original">--original</a><br/>
				<i>Not used by default</i>
			</td>
		</tr>
		<tr id="opt_allow_non_variant">
			<td><pre>--allow_non_variant</pre></td>
			<td>&nbsp;</td>
			<td>
				When using VCF format as input and output, by default the VEP
				will skip non-variant lines of input (where the ALT allele is
				null). Enabling this option the lines will be printed in the VCF
				output with no consequence data added.
			</td>
		</tr>
	</table>
	
	
	<hr/>
	
	<h2 id="dbopt">Database options</h2>
	
	<table class="ss">
		<tr>
			<th width="20%">Flag</th>
			<th width="10%">Alternate</th>
			<th>Description</th>
		</tr>
		<tr id="opt_database">
			<td><pre>--database</pre></td>
			<td>&nbsp;</td>
			<td>Enable the VEP to use local or remote databases.</td>
		</tr>
		<tr id="opt_host">
			<td><pre>--host [hostname]</pre></td>
			<td>&nbsp;</td>
			<td>
				Manually define the database host to connect to. Users in the US
				may find connection and transfer speeds quicker using our East
				coast mirror, useastdb.ensembl.org. <i>Default =
				"ensembldb.ensembl.org"</i>
			</td>
		</tr>
		<tr id="opt_user">
			<td><pre>--user [username]</pre></td>
			<td><pre>-u</pre></td>
			<td>Manually define the database username. <i>Default =
			"anonymous"</i></td>
		</tr>
		<tr id="opt_password">
			<td><pre>--password [password]</pre></td>
			<td><pre>--pass</pre></td>
			<td>Manually define the database password. <i>Not used by
			default</i></td>
		</tr>
		<tr id="opt_port">
			<td><pre>--port [number]</pre></td>
			<td>&nbsp;</td>
			<td>Manually define the database port. <i>Default = 5306</i></td>
		</tr>
		<tr id="opt_genomes">
			<td><pre>--genomes</pre></td>
			<td>&nbsp;</td>
			<td>
				Override the default connection settings with those for the
				Ensembl Genomes public MySQL server. Required when using any of
				the <a rel="external" href="http://www.ensemblgenomes.org/"
				target="_blank">Ensembl Genomes</a> species. <i>Not used by
				default</i>
			</td>
		</tr>
		<tr id="opt_refseq">
			<td><pre>--refseq</pre></td>
			<td>&nbsp;</td>
			<td>
				Instead of using the core database, use the otherfeatures
				database to retrieve transcripts. This database contains
				transcript objects corresponding to RefSeq transcripts, along
				with CCDS and Ensembl ESTs. Consequence output will be given
				relative to these transcripts in place of the default Ensembl
				transcripts. See <a
				href="/info/docs/api/core/index.html#est">here</a> for more
				details on the contents of the otherfeatures database. The
				otherfeatures database currently only exists for human and
				mouse. <i>Not used by default</i>
			</td>
		</tr>
		<tr id="opt_db_version">
			<td><pre>--db_version [number]</pre></td>
			<td><pre>--db</pre></td>
			<td>
				Force the script to connect to a specific version of the Ensembl
				databases. Not recommended as there will usually be conflicts
				between software and database versions. <i>Not used by
				default</i>
			</td>
		</tr>
		<tr id="opt_registry">
			<td><pre>--registry [filename]</pre></td>
			<td>&nbsp;</td>
			<td>
				Defining a registry file overwrites other connection settings
				and uses those found in the specified registry file to connect.
				<i>Not used by default</i>
			</td>
		</tr>
	</table>
	
	
	
	<hr/>
	
	
	
	<h2 id="advopt">Advanced options</h2>
	
	<table class="ss">
		<tr>
			<th width="20%">Flag</th>
			<th width="10%">Alternate</th>
			<th>Description</th>
		</tr>
		<tr id="opt_no_whole_genome">
			<td><pre>--no_whole_genome</pre></td>
			<td>&nbsp;</td>
			<td>
				Force the script to run in non-whole-genome mode. This was the
				original default mode for the VEP script, but has now been
				superceded by whole-genome mode, which is the default. In this
				mode, variants are analysed one at a time, with no caching of
				transcript data. <i>Not used by default</i>
			</td>
		</tr>
		<tr id="opt_buffer_size">
			<td><pre>--buffer_size [number]</pre></td>
			<td>&nbsp;</td>
			<td>
				Sets the internal buffer size, corresponding to the number of
				variations that are read in to memory simultaneously. Set this
				lower to use less memory at the expense of longer run time, and
				higher to use more memory with a faster run time. <i>Default =
				5000</i>
			</td>
		</tr>
		<tr id="opt_write_cache">
			<td><pre>--write_cache</pre></td>
			<td>&nbsp;</td>
			<td>
				Enable writing to the cache. <i>Not used by default</i>
			</td>
		</tr>
		<tr id="opt_build">
			<td><pre>--build [all<b>|</b>list]</pre></td>
			<td>&nbsp;</td>
			<td>
                Build a complete cache for the selected species from the
                database. Either specify a list of chromosomes (see <a
                href="#opt_chr">--chr</a> for how to do this), or use
                <pre>--build all</pre> to build for all top-level chromosomes.
                <span style="color:red;">WARNING: Do not use this flag when
                connected to one of the public databases - please instead
                download a pre-built cache or build against a local database.
                </span><i>Not used by default</i>
			</td>
		</tr>
		<tr id="opt_compress">
			<td><pre>--compress [command]</pre></td>
			<td>&nbsp;</td>
			<td>
				By default the VEP uses the utility zcat to decompress cached
				files. On some systems zcat may not be installed or may
				misbehave; by specifying one of <a href="vep_options.html#opt_compress">--compress gzcat</a> or
				<a href="vep_options.html#opt_compress">--compress "gzip -dc"</a> you may be able to bypass these
				problems. <i>Not used by default</i>
			</td>
		</tr>
		<tr id="opt_skip_db_check">
			<td><pre>--skip_db_check</pre></td>
			<td>&nbsp;</td>
			<td>
				<span style="color:red;">ADVANCED</span> Force the script to use
				a cache built from a different host than specified with --host.
				Only use this if you are sure the two hosts are compatible (e.g.
				ensembldb.ensembl.org can be considered compatible with
				useastdb.ensembl.org as the data is mirrored between the two).
				<i>Not used by default</i>
			</td>
		</tr>
		<tr id="opt_cache_region_size">
			<td><pre>--cache_region_size [size]</pre></td>
			<td>&nbsp;</td>
			<td>
				<span style="color:red;">ADVANCED</span> The size in base-pairs
				of the region covered by one file in the cache. By default this
				is 1MB, which produces approximately ~500 files maximum per
				sub-directory in human. Reducing this can reduce the amount of
				memory and decrease the run-time when you use a cache built this
				way. Note that you must specify the same --cache_region_size
				when both building/writing to the cache and reading from it.
				<i>Not used by default</i>
			</td>
		</tr>
	</table>
	
	
	
	<hr/>
	
	
	<h2 id="examples">Examples</h2>
	
	<ul>
		<li>
			Read input from STDIN, output to STDOUT
	<pre class="code">perl variant_effect_predictor.pl -cache -o stdout</pre>
		</li>
		<li>
			Add regulatory region consequences
	<pre class="code">perl variant_effect_predictor.pl -cache -i variants.txt -regulatory</pre>
		</li>
		<li>
			Input file variants.vcf.txt, input file format VCF, add gene symbol
			identifiers, output SO consequence terms
	<pre class="code">perl variant_effect_predictor.pl -cache -i variants.vcf.txt -format vcf -symbol -t so</pre>
		</li>
		<li>
			Filter out common variants based on 1000 genomes data
	<pre class="code">perl variant_effect_predictor.pl -cache -i variants.txt -filter_common</pre>
		</li>
		<li>
			Force overwrite of output file variants_output.txt, check for existing
			co-located variants, output only coding sequence consequences, output
			HGVS names
	<pre class="code">perl variant_effect_predictor.pl -cache -i variants.txt -o variants_output.txt -force -check_existing -coding_only -hgvs</pre>
		</li>
		<li>
			Specify DB connection parameters in registry file ensembl.registry, add SIFT
			score and prediction, PolyPhen prediction
	<pre class="code">perl variant_effect_predictor.pl -database -i variants.txt -registry ensembl.registry -sift b -polyphen p</pre>
		</li>
		<li>
			Connect to Ensembl Genomes db server for A.thaliana
	<pre class="code">perl variant_effect_predictor.pl -database -i variants.txt -genomes -species arabidopsis_thaliana</pre>
		</li>
		<li>
			Load config from ini file, run in quiet mode
	<pre class="code">perl variant_effect_predictor.pl -config vep.ini -i variants.txt -q</pre>
		</li>
		<li>
			Use cache in /home/vep/mycache/, use gzcat instead of zcat
	<pre class="code">perl variant_effect_predictor.pl -cache -dir /home/vep/mycache/ -i variants.txt -compress gzcat</pre>
		</li>
		<li>
			Convert RefSeq-based HGVS notations to genomic coordinates in VCF format
	<pre class="code">perl variant_effect_predictor.pl -database -i hgvs.txt -o hgvs.vcf -refseq -convert vcf</pre>
		</li>
		<li>
			Filter input file on consequence type to include only variants that
			cause a coding sequence change, write output in original input format
	<pre class="code">perl variant_effect_predictor.pl -cache -i variants.vcf -o variants_filtered.vcf -filter coding_change -original</pre>
		</li>
		<li>
			Add custom position-based phenotype annotation from remote BED file
	<pre class="code">perl variant_effect_predictor.pl -cache -i variants.vcf -custom ftp://ftp.myhost.org/data/phenotypes.bed.gz,phenotype</pre>
		</li>
		<li>
			Use the plugin named MyPlugin, output only the variation name, feature, consequence type and MyPluginOutput fields
	<pre class="code">perl variant_effect_predictor.pl -cache -i variants.vcf -plugin MyPlugin -fields Uploaded_variation,Feature,Consequence,MyPluginOutput</pre>
		</li>
	</ul>
	
		
	<hr/>
	<h2 id="performance">Performance</h2>
	
	<p> In optimal conditions, the VEP script is capable of processing >2
	million variants in 1 hour. Run time is dependent on various factors, and is
	especially affected by the chromosomal distribution of variants. Variants in
	exonic regions naturally take longer to process than those in intronic or
	intergenic regions. Due to the way transcript data is cached in memory, the
	VEP will, for example, process a file containing 100 variants that fall in
	one gene faster than it would a file containing 100 variants in 100
	different genes. </p>
	
	<p> The VEP is also optimised to run on input files that are sorted in
	chromosomal order. Unsorted files will still work, albeit more slowly. </p>
	
	<p> For very large files (for example those from whole-genome sequencing),
	the VEP process can be easily parallelised by dividing your file into chunks
	(e.g. by chromosome). The VEP will also work with tabix-indexed, bgzipped
	VCF files, and so the tabix utility could be used to divide the input file:
	</p>
	
	<pre class="code"> tabix -h variants.vcf.gz 12:1000000-20000000 | perl variant_effect_predictor.pl -cache -vcf </pre>
	
	<h4 id="forking"> Forking </h4>
	
	<p> Using forking enables the VEP to run multiple parallel "threads", with
	each thread processing a subset of your input. Most modern computers have
	more than one processor core, so running the VEP with forking enabled can
	give huge speed increases (3-4x faster in most cases). Even computers with a
	single core will see speed benefits due to overheads associated with using
	object-oriented code in Perl. </p>
	
    <p> To use forking, you must choose a number of forks to use with the <a
    href="#opt_fork">--fork</a> flag. Most users should use 4 forks: </p>
	
	<pre class=code>perl variant_effect_predictor.pl -i my_input.vcf -fork 4 -offline</pre>
	
	<p> but depending on various factors specific to your setup you may see
	faster performance with fewer or more forks. </p>
	
    <p> <b>NB:</b> VEP users writing <a href="vep_plugins.html">plugins</a>
    should be aware that while the VEP code attempts to preserve the state of
    any plugin-specific cached data between separate forks, there may be
    situations where data is lost. If you find this is the case, you should
    disable forking in the new() method of your plugin by deleting the "fork"
    key from the $config hash. </p>
	
	

	
</div>

</body>
</html>
